// @gooddata/sdk-ui-ext/esm/internal/utils/colors.js
import { __assign, __spreadArrays } from "tslib";
// (C) 2019-2020 GoodData Corporation
import set from "lodash/set";
import isEqual from "lodash/isEqual";
import uniqBy from "lodash/uniqBy";
import isEmpty from "lodash/isEmpty";
import cloneDeep from "lodash/cloneDeep";
import compact from "lodash/compact";
import { isAttributeDescriptor, isMeasureDescriptor, isResultAttributeHeader, } from "@gooddata/sdk-backend-spi";
import { isColorFromPalette, isRgbColor } from "@gooddata/sdk-model";
import { getMappingHeaderName } from "@gooddata/sdk-ui";
import { ColorUtils } from "@gooddata/sdk-ui-charts";
function getItemName(item) {
    return getMappingHeaderName(item.mappingHeader) || "";
}
export function getSearchedItems(inputItems, searchString) {
    if (isEmpty(searchString)) {
        return inputItems;
    }
    return inputItems.filter(function (item) {
        var name = getItemName(item);
        return name.toLowerCase().includes(searchString.toLowerCase());
    });
}
export function getColoredInputItems(colors) {
    var inputItems = [];
    if (colors && colors.colorAssignments) {
        inputItems = colors.colorAssignments.map(function (assignmentItem, index) {
            if (isColorFromPalette(assignmentItem.color)) {
                return {
                    colorItem: assignmentItem.color,
                    mappingHeader: assignmentItem.headerItem,
                    color: ColorUtils.getColorByGuid(colors.colorPalette, assignmentItem.color.value, index),
                };
            }
            else if (isRgbColor(assignmentItem.color)) {
                return {
                    colorItem: assignmentItem.color,
                    mappingHeader: assignmentItem.headerItem,
                    color: assignmentItem.color.value,
                };
            }
        });
    }
    return inputItems;
}
function getMeasureMappingIdentifier(item) {
    return item.measureHeaderItem.localIdentifier;
}
function mergeColorMappingToProperties(properties, id, color) {
    var _a, _b;
    var colorMapping = [
        {
            id: id,
            color: color,
        },
    ];
    var previousColorMapping = (_b = (_a = properties === null || properties === void 0 ? void 0 : properties.controls) === null || _a === void 0 ? void 0 : _a.colorMapping) !== null && _b !== void 0 ? _b : [];
    var mergedMapping = compact(uniqBy(__spreadArrays(colorMapping, previousColorMapping), "id"));
    var newProperties = cloneDeep(properties);
    set(newProperties, "controls.colorMapping", mergedMapping);
    return newProperties;
}
export function getProperties(properties, item, color) {
    if (isMeasureDescriptor(item)) {
        var id = getMeasureMappingIdentifier(item);
        return mergeColorMappingToProperties(properties, id, color);
    }
    else if (isResultAttributeHeader(item)) {
        return mergeColorMappingToProperties(properties, item.attributeHeaderItem.uri, color);
    }
    else if (isAttributeDescriptor(item)) {
        return mergeColorMappingToProperties(properties, item.attributeHeader.uri, color);
    }
    return {};
}
export function getValidProperties(properties, colorAssignments) {
    if (!properties || !properties.controls || !properties.controls.colorMapping) {
        return properties;
    }
    var reducedColorMapping = properties.controls.colorMapping.filter(function (mappingItem) {
        var id = mappingItem.id;
        var colorValue = mappingItem.color.value;
        var assignmentValid = colorAssignments.find(function (colorAssignment) {
            if (isMeasureDescriptor(colorAssignment.headerItem)) {
                return (colorAssignment.headerItem.measureHeaderItem.localIdentifier === id &&
                    isEqual(colorAssignment.color.value, colorValue));
            }
            else if (isResultAttributeHeader(colorAssignment.headerItem)) {
                return colorAssignment.headerItem.attributeHeaderItem.uri === id;
            }
            else if (isAttributeDescriptor(colorAssignment.headerItem)) {
                return colorAssignment.headerItem.attributeHeader.uri === id;
            }
            return false;
        });
        return assignmentValid !== undefined;
    });
    return __assign(__assign({}, properties), { controls: __assign(__assign({}, properties.controls), { colorMapping: reducedColorMapping.length ? reducedColorMapping : null }) });
}
