// @gooddata/sdk-backend-bear/esm/backend/workspace/execution/preparedExecutionByRef.js
// (C) 2019-2022 GoodData Corporation
import { __awaiter, __generator, __spreadArrays } from "tslib";
import { defFingerprint, defWithDimensions, defWithSorting, insightRef, defWithDateFormat, } from "@gooddata/sdk-model";
import isEqual from "lodash/isEqual";
import isEmpty from "lodash/isEmpty";
import { convertExecutionApiError } from "../../../utils/errorHandling";
import { BearExecutionResult } from "./executionResult";
import { convertResultSpec } from "../../../convertors/toBackend/afm/ExecutionConverter";
import { objRefToUri } from "../../../utils/api";
import { convertFilters } from "../../../convertors/toBackend/afm/FilterConverter";
var BearPreparedExecutionByRef = /** @class */ (function () {
    function BearPreparedExecutionByRef(authCall, definition, insight, filters, executionFactory) {
        if (filters === void 0) { filters = []; }
        this.authCall = authCall;
        this.definition = definition;
        this.insight = insight;
        this.filters = filters;
        this.executionFactory = executionFactory;
    }
    BearPreparedExecutionByRef.prototype.execute = function () {
        return __awaiter(this, void 0, void 0, function () {
            var execution;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.createVisualizationExecution()];
                    case 1:
                        execution = _a.sent();
                        return [2 /*return*/, this.authCall(function (sdk) {
                                return sdk.execution
                                    ._getVisExecutionResponse(_this.definition.workspace, execution)
                                    .then(function (response) {
                                    return new BearExecutionResult(_this.authCall, _this.definition, _this.executionFactory, response);
                                });
                            }, convertExecutionApiError)];
                }
            });
        });
    };
    BearPreparedExecutionByRef.prototype.explain = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                // eslint-disable-next-line no-console
                console.warn("Backend does not support explain mode");
                return [2 /*return*/, Promise.resolve()];
            });
        });
    };
    BearPreparedExecutionByRef.prototype.createVisualizationExecution = function () {
        return __awaiter(this, void 0, void 0, function () {
            var uri, resultSpec, filters;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, objRefToUri(insightRef(this.insight), this.definition.workspace, this.authCall)];
                    case 1:
                        uri = _a.sent();
                        resultSpec = convertResultSpec(this.definition);
                        filters = convertFilters(this.filters);
                        return [2 /*return*/, {
                                visualizationExecution: {
                                    reference: uri,
                                    resultSpec: resultSpec,
                                    filters: filters,
                                },
                            }];
                }
            });
        });
    };
    BearPreparedExecutionByRef.prototype.withDimensions = function () {
        var dimsOrGen = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            dimsOrGen[_i] = arguments[_i];
        }
        return this.executionFactory.forDefinition(defWithDimensions.apply(void 0, __spreadArrays([this.definition], dimsOrGen)));
    };
    BearPreparedExecutionByRef.prototype.withSorting = function () {
        var items = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            items[_i] = arguments[_i];
        }
        return this.executionFactory.forDefinition(defWithSorting(this.definition, items));
    };
    BearPreparedExecutionByRef.prototype.withDateFormat = function (dateFormat) {
        return this.executionFactory.forDefinition(defWithDateFormat(this.definition, dateFormat));
    };
    BearPreparedExecutionByRef.prototype.withExecConfig = function (config) {
        if (!isEmpty(config === null || config === void 0 ? void 0 : config.dataSamplingPercentage)) {
            // eslint-disable-next-line no-console
            console.warn("Backend does not support data sampling, result will be not affected");
        }
        return this.executionFactory.forDefinition(this.definition);
    };
    BearPreparedExecutionByRef.prototype.fingerprint = function () {
        if (!this._fingerprint) {
            this._fingerprint = defFingerprint(this.definition);
        }
        return this._fingerprint;
    };
    BearPreparedExecutionByRef.prototype.equals = function (other) {
        return isEqual(this.definition, other.definition);
    };
    return BearPreparedExecutionByRef;
}());
export { BearPreparedExecutionByRef };
