// @gooddata/sdk-ui-kit/esm/Menu/menuOpener/MenuOpener.js
import { __extends } from "tslib";
// (C) 2007-2018 GoodData Corporation
import React from "react";
import { MenuOpenedByClick } from "./MenuOpenedByClick";
import { MenuOpenedByHover } from "./MenuOpenedByHover";
var MenuOpener = /** @class */ (function (_super) {
    __extends(MenuOpener, _super);
    function MenuOpener() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getComponentByOpenAction = function () {
            switch (_this.props.openAction) {
                case "click":
                    return MenuOpenedByClick;
                case "hover":
                    return MenuOpenedByHover;
            }
        };
        return _this;
    }
    MenuOpener.prototype.render = function () {
        var Component = this.getComponentByOpenAction();
        return (React.createElement(Component, { opened: this.props.opened, onOpenedChange: this.props.onOpenedChange, topLevelMenu: this.props.topLevelMenu, alignment: this.props.alignment, spacing: this.props.spacing, offset: this.props.offset, toggler: this.props.toggler, togglerWrapperClassName: this.props.togglerWrapperClassName, portalTarget: this.props.portalTarget },
            React.createElement("div", { className: "gd-menuOpener" }, this.props.children)));
    };
    MenuOpener.defaultProps = {
        openAction: "hover",
        alignment: ["right", "bottom"],
        spacing: 0,
        offset: 0,
        portalTarget: document.querySelector("body"),
    };
    return MenuOpener;
}(React.Component));
export { MenuOpener };
