// @gooddata/sdk-model/esm/execution/measure/fingerprint.js
// (C) 2019-2020 GoodData Corporation
import { __assign } from "tslib";
import stringify from "json-stable-stringify";
import { isSimpleMeasure } from "./index";
import merge from "lodash/merge";
import { isFilterRelevantForFingerprinting } from "../filter/fingerprint";
function simpleMeasureFingerprint(measure) {
    var _a;
    var measureDefinition = measure.measure.definition.measureDefinition;
    var measureDefinitionWithSanitizedFilters = __assign(__assign({}, measureDefinition), { filters: (_a = measureDefinition.filters) === null || _a === void 0 ? void 0 : _a.filter(isFilterRelevantForFingerprinting) });
    var measureDefinitionDefaults = {
        filters: [],
        computeRatio: false,
    };
    var measureDefinitionWithDefaults = merge(measureDefinitionDefaults, measureDefinitionWithSanitizedFilters);
    return stringify({
        measure: __assign(__assign({}, measure.measure), { definition: {
                measureDefinition: measureDefinitionWithDefaults,
            } }),
    });
}
/**
 * Calculates dimension fingerprint; ensures that the optional vs default values are correctly reflected in
 * the fingerprint.
 *
 * @internal
 */
export function measureFingerprint(measure) {
    if (isSimpleMeasure(measure)) {
        /*
         * Simple measure has a few optional properties, which, when not provided default to specific
         * values (aggregation, computeRatio etc).
         *
         * Fingerprinting simple measure thus requires normalization of the definition so that measure with
         * optional props not specified has same fingerprint as measure with optional props specified to default
         * values.
         */
        return simpleMeasureFingerprint(measure);
    }
    return stringify(measure);
}
