// @gooddata/sdk-ui-ext/esm/internal/dashboardEmbedding/KpiAlerts/utils/translationUtils.js
import { __assign } from "tslib";
import lowerFirst from "lodash/lowerFirst";
import { DateFilterHelpers } from "@gooddata/sdk-ui-filters";
import { absoluteDateFilterValues, isAbsoluteDateFilter, isAllTimeDateFilter, isRelativeDateFilter, relativeDateFilterValues, } from "@gooddata/sdk-model";
function filterMetadata(filter) {
    var _a, _b, _c, _d, _e, _f;
    if (isRelativeDateFilter(filter)) {
        return __assign(__assign({}, relativeDateFilterValues(filter)), { type: "relative" });
    }
    else if (isAbsoluteDateFilter(filter)) {
        return __assign(__assign({}, absoluteDateFilterValues(filter)), { type: "absolute" });
    }
    else if (filter.dateFilter.type === "relative") {
        return {
            type: "relative",
            from: Number.parseInt((_b = (_a = filter.dateFilter.from) === null || _a === void 0 ? void 0 : _a.toString()) !== null && _b !== void 0 ? _b : "0", 10),
            to: Number.parseInt((_d = (_c = filter.dateFilter.to) === null || _c === void 0 ? void 0 : _c.toString()) !== null && _d !== void 0 ? _d : "0", 10),
            granularity: filter.dateFilter.granularity,
        };
    }
    else {
        return {
            type: "absolute",
            from: (_e = filter.dateFilter.from) === null || _e === void 0 ? void 0 : _e.toString(),
            to: (_f = filter.dateFilter.to) === null || _f === void 0 ? void 0 : _f.toString(),
        };
    }
}
export function translateDateFilter(filter, intl, dateFormat) {
    var metadata = filterMetadata(filter);
    return metadata.type === "absolute"
        ? DateFilterHelpers.formatAbsoluteDateRange(metadata.from, metadata.to, dateFormat)
        : DateFilterHelpers.formatRelativeDateRange(metadata.from, metadata.to, metadata.granularity, intl);
}
function getIntlIdRoot(filter) {
    var metadata = filterMetadata(filter);
    if (metadata.type === "absolute") {
        return "filters.alertMessage.relativePreset.inPeriod";
    }
    var hasOneBoundToday = metadata.from === 0 || metadata.to === 0; // e.g. last 4 months, next 6 days
    var isLastOneX = metadata.from === -1 && metadata.to === -1; // e.g last month
    var isNextOneX = metadata.from === 1 && metadata.to === 1; // e.g. next month
    return hasOneBoundToday || isLastOneX || isNextOneX
        ? "filters.alertMessage.relativePreset"
        : "filters.alertMessage.relativePreset.inPeriod";
}
export function getKpiAlertTranslationData(filter, intl, dateFormat) {
    if (!filter || isAllTimeDateFilter(filter)) {
        return null;
    }
    return {
        intlIdRoot: getIntlIdRoot(filter),
        rangeText: lowerFirst(translateDateFilter(filter, intl, dateFormat)),
    };
}
