// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/xirr/XirrDescriptor.js
// (C) 2021 GoodData Corporation
import { PluggableXirr } from "./PluggableXirr";
import { DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT, MAX_VISUALIZATION_HEIGHT } from "../constants";
var XirrDescriptor = /** @class */ (function () {
    function XirrDescriptor() {
    }
    XirrDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableXirr(params); };
    };
    XirrDescriptor.prototype.getSizeInfo = function (_insight, layoutDescriptor, settings) {
        return {
            width: {
                default: 2,
                min: 2,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getMinHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    };
    XirrDescriptor.prototype.getDefaultHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT;
        }
        return 8;
    };
    XirrDescriptor.prototype.getMinHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT;
        }
        return 6;
    };
    XirrDescriptor.prototype.getMaxHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT;
        }
        return MAX_VISUALIZATION_HEIGHT;
    };
    XirrDescriptor.prototype.applyDrillDown = function (insight) {
        return insight;
    };
    return XirrDescriptor;
}());
export { XirrDescriptor };
