// @gooddata/api-model-bear/esm/executeAfm/GdcExecution.js
// (C) 2007-2020 GoodData Corporation
import isEmpty from "lodash/isEmpty";
/**
 * @public
 */
export var GdcExecution;
(function (GdcExecution) {
    function isAttributeHeaderItem(header) {
        return !isEmpty(header) && header.attributeHeaderItem !== undefined;
    }
    GdcExecution.isAttributeHeaderItem = isAttributeHeaderItem;
    function isMeasureHeaderItem(header) {
        return !isEmpty(header) && header.measureHeaderItem !== undefined;
    }
    GdcExecution.isMeasureHeaderItem = isMeasureHeaderItem;
    function isTotalHeaderItem(header) {
        return !isEmpty(header) && header.totalHeaderItem !== undefined;
    }
    GdcExecution.isTotalHeaderItem = isTotalHeaderItem;
    function isAttributeHeader(header) {
        return !isEmpty(header) && header.attributeHeader !== undefined;
    }
    GdcExecution.isAttributeHeader = isAttributeHeader;
    function isMeasureGroupHeader(header) {
        return !isEmpty(header) && header.measureGroupHeader !== undefined;
    }
    GdcExecution.isMeasureGroupHeader = isMeasureGroupHeader;
})(GdcExecution || (GdcExecution = {}));
