// @gooddata/sdk-model/esm/execution/attribute/factory.js
import { __assign } from "tslib";
// (C) 2019-2021 GoodData Corporation
import identity from "lodash/identity";
import cloneDeep from "lodash/cloneDeep";
import isEmpty from "lodash/isEmpty";
import { isAttribute } from "./index";
import { objRefToString, isObjRef } from "../../objRef";
import { idRef } from "../../objRef/factory";
import SparkMD5 from "spark-md5";
import { sanitizeLocalId } from "../../sanitizeLocalId";
/**
 * Builder for attributes.
 *
 * Do not instantiate this class directly. Instead use {@link newAttribute} or {@link modifyAttribute}.
 *
 * @public
 */
var AttributeBuilder = /** @class */ (function () {
    /**
     * @internal
     */
    function AttributeBuilder(input) {
        var _this = this;
        this.customLocalId = false;
        /**
         * Sets alias - alternative title - for the attribute. This value will then be used in various
         * chart-specific descriptive elements. For convenience if no alias is specified, the attribute
         * will fall back to server-defined value.
         *
         * @param alias - alias to use instead of attribute title; undefined to use server-defined value
         */
        this.alias = function (alias) {
            if (!alias) {
                return _this.noAlias();
            }
            _this.attribute.alias = alias;
            return _this;
        };
        /**
         * Resets alias - alternative title - set for the attribute. The server-defined title of the attribute
         * will be used instead.
         */
        this.noAlias = function () {
            delete _this.attribute.alias;
            return _this;
        };
        /**
         * Sets display form reference.
         *
         * @param ref - new ref to use
         */
        this.displayForm = function (ref) {
            _this.attribute.displayForm = ref;
            return _this;
        };
        /**
         * Sets local identifier (localId) for the attribute. LocalId can be used to reference the attribute
         * within the execution definition.
         *
         * Normally, builder will generate localId based on contents of the attribute definition - taking all
         * properties into account: in typical scenarios you don't have to call this function at all. The only exception
         * where you have to provide custom local id is if your execution must contain the exact same attribute twice.
         *
         * For convenience, this method also accepts 'undefined', which indicates that the default local id generation
         * logic should be used.
         *
         * @param localId - local identifier to set; if not specified, the builder will ensure local id will
         * be generated
         */
        this.localId = function (localId) {
            if (!localId || localId.trim().length === 0) {
                return _this.defaultLocalId();
            }
            _this.attribute.localIdentifier = localId;
            _this.customLocalId = true;
            return _this;
        };
        /**
         * Indicates that the attribute's localId should be generated using the default local-id generator logic.
         */
        this.defaultLocalId = function () {
            _this.attribute.localIdentifier = "";
            _this.customLocalId = false;
            return _this;
        };
        /**
         * Creates the IAttribute instance.
         */
        this.build = function () {
            var localIdentifier = _this.getOrGenerateLocalId();
            return {
                attribute: __assign(__assign({}, _this.attribute), { localIdentifier: localIdentifier }),
            };
        };
        if (isAttribute(input)) {
            this.attribute = cloneDeep(input.attribute);
            this.customLocalId = true;
        }
        else {
            var displayForm = isObjRef(input) ? input : idRef(input, "displayForm");
            this.attribute = {
                displayForm: displayForm,
                localIdentifier: "",
            };
        }
    }
    AttributeBuilder.prototype.getOrGenerateLocalId = function () {
        if (this.customLocalId && !isEmpty(this.attribute.localIdentifier)) {
            return this.attribute.localIdentifier;
        }
        return sanitizeLocalId(["a", this.calculateAliasHash(), objRefToString(this.attribute.displayForm)]
            .filter(function (part) { return !isEmpty(part); })
            .join("_"));
    };
    AttributeBuilder.prototype.calculateAliasHash = function () {
        if (!this.attribute.alias) {
            return "";
        }
        var hasher = new SparkMD5();
        hasher.append(this.attribute.alias);
        return hasher.end().substr(0, 8);
    };
    return AttributeBuilder;
}());
export { AttributeBuilder };
/**
 * Creates a new attribute with the specified display form ref and optional modifications and localIdentifier.
 * @param displayFormRefOrId - ref or identifier of the attribute display form
 * @param modifications - optional modifications (e.g. alias, etc.)
 * @public
 */
export function newAttribute(displayFormRefOrId, modifications) {
    if (modifications === void 0) { modifications = identity; }
    var builder = new AttributeBuilder(displayFormRefOrId);
    return modifications(builder).build();
}
/**
 * Allows modification of an existing attribute instance.
 *
 * The returned attribute will have the same localId as the original attribute. If you would like to assign
 * new/different local identifier to the attribute, you can do that using the modifications where you can provide
 * either new custom localId or indicate that the attribute should fall back to the auto-generated localId.
 *
 * @param attribute - attribute to modify
 * @param modifications - modification function
 * @public
 */
export function modifyAttribute(attribute, modifications) {
    if (modifications === void 0) { modifications = identity; }
    var builder = new AttributeBuilder(attribute);
    return modifications(builder).build();
}
