// @gooddata/sdk-ui-dashboard/esm/presentation/widget/kpi/DefaultDashboardKpi/KpiAlerts/utils/brokenFilterUtils.js
// (C) 2021 GoodData Corporation
import { UnexpectedError, } from "@gooddata/sdk-backend-spi";
import { areObjRefsEqual, attributeElementsCount, attributeElementsIsEmpty, isAttributeElementsByRef, objRefToString, } from "@gooddata/sdk-model";
import { translateDateFilter } from "./translationUtils";
import { isBrokenAlertAttributeFilterInfo, isBrokenAlertDateFilterInfo, } from "../../../../../../model";
/**
 * Takes basic broken alert info and adds additional information used for displaying of such filters to the user.
 *
 * @param brokenAlertFilters - the basic broken alert filters info to enrich
 * @param intl - the intl object used
 * @param dateFormat - the date format to be used
 * @param dateDataSets - all available date data sets
 * @param attributeFiltersMeta - additional information about attribute filters (see {@link IAttributeFilterMetaCollection} for details)
 */
export function enrichBrokenAlertsInfo(brokenAlertFilters, intl, dateFormat, dateDataSets, attributeFiltersMeta) {
    return brokenAlertFilters.map(function (brokenFilter) {
        if (isBrokenAlertAttributeFilterInfo(brokenFilter)) {
            return enrichBrokenAttributeFilter(brokenFilter, attributeFiltersMeta);
        }
        if (isBrokenAlertDateFilterInfo(brokenFilter)) {
            return enrichBrokenDateFilter(brokenFilter, intl, dateFormat, dateDataSets);
        }
        throw new UnexpectedError("Unknown broken alert filter type.");
    });
}
function enrichBrokenDateFilter(brokenFilter, intl, dateFormat, dateDataSets) {
    var _a;
    var alertFilter = brokenFilter.alertFilter, brokenType = brokenFilter.brokenType;
    var dateFilterTitle = translateDateFilter(alertFilter, intl, dateFormat);
    var matchingDateDataset = dateDataSets.find(function (dataset) {
        return areObjRefsEqual(dataset, alertFilter.dateFilter.dataSet);
    });
    return {
        type: "date",
        brokenType: brokenType,
        dateFilterTitle: dateFilterTitle,
        title: (_a = matchingDateDataset === null || matchingDateDataset === void 0 ? void 0 : matchingDateDataset.title) !== null && _a !== void 0 ? _a : intl.formatMessage({ id: "kpiAlertDialog.brokenAlertDefaultDateLabel" }),
    };
}
function enrichBrokenAttributeFilter(brokenFilter, attributeFiltersMeta) {
    var alertFilter = brokenFilter.alertFilter, brokenType = brokenFilter.brokenType;
    var metaKey = objRefToString(alertFilter.attributeFilter.displayForm);
    var meta = attributeFiltersMeta[metaKey];
    var isNegative = alertFilter.attributeFilter.negativeSelection;
    var totalCount = meta.totalElementsCount;
    var elements = meta.validElements.filter(function (element) {
        var isInSelected = isAttributeElementsByRef(alertFilter.attributeFilter.attributeElements)
            ? alertFilter.attributeFilter.attributeElements.uris.some(function (uri) { return uri === element.uri; })
            : alertFilter.attributeFilter.attributeElements.values.some(function (value) { return value === element.title; });
        return isInSelected !== isNegative;
    });
    var selection = elements.map(function (el) { return el.title; }).join(", ");
    var title = meta.title;
    var selectedCount = attributeElementsCount(alertFilter.attributeFilter.attributeElements);
    return {
        type: "attribute",
        brokenType: brokenType,
        isAllSelected: isNegative && attributeElementsIsEmpty(alertFilter.attributeFilter.attributeElements),
        selection: selection,
        selectionSize: isNegative ? totalCount - selectedCount : selectedCount,
        title: title,
    };
}
