// @gooddata/sdk-ui-pivot/esm/columnWidths.js
import { __assign, __spreadArrays } from "tslib";
// (C) 2007-2021 GoodData Corporation
import isEmpty from "lodash/isEmpty";
import { attributeLocalId, measureLocalId } from "@gooddata/sdk-model";
//
// types used in implementation internals
//
/**
 * @internal
 */
export var ColumnEventSourceType;
(function (ColumnEventSourceType) {
    ColumnEventSourceType["AUTOSIZE_COLUMNS"] = "autosizeColumns";
    ColumnEventSourceType["UI_DRAGGED"] = "uiColumnDragged";
    ColumnEventSourceType["FIT_GROW"] = "growToFit";
})(ColumnEventSourceType || (ColumnEventSourceType = {}));
/**
 * @internal
 */
export var UIClick;
(function (UIClick) {
    UIClick[UIClick["CLICK"] = 1] = "CLICK";
    UIClick[UIClick["DOUBLE_CLICK"] = 2] = "DOUBLE_CLICK";
})(UIClick || (UIClick = {}));
/**
 * Tests whether object is an instance of {@link IMeasureColumnLocator}
 *
 * @public
 */
export function isMeasureColumnLocator(obj) {
    return !isEmpty(obj) && obj.measureLocatorItem !== undefined;
}
/**
 * Tests whether object is an instance of {@link IAttributeColumnLocator}
 *
 * @public
 */
export function isAttributeColumnLocator(obj) {
    return !isEmpty(obj) && obj.attributeLocatorItem !== undefined;
}
/**
 * Tests whether object is an instance of {@link IAbsoluteColumnWidth}
 *
 * @public
 */
export function isAbsoluteColumnWidth(columnWidth) {
    return Number(columnWidth.value) === columnWidth.value;
}
/**
 * Tests whether object is an instance of {@link IAttributeColumnWidthItem}
 *
 * @public
 */
export function isAttributeColumnWidthItem(obj) {
    return !isEmpty(obj) && obj.attributeColumnWidthItem !== undefined;
}
/**
 * Tests whether object is an instance of {@link IMeasureColumnWidthItem}
 *
 * @public
 */
export function isMeasureColumnWidthItem(obj) {
    return (!isEmpty(obj) &&
        obj.measureColumnWidthItem !== undefined &&
        obj.measureColumnWidthItem.locators !== undefined);
}
/**
 * Tests whether object is an instance of {@link IAllMeasureColumnWidthItem}
 *
 * @public
 */
export function isAllMeasureColumnWidthItem(obj) {
    return (!isEmpty(obj) &&
        obj.measureColumnWidthItem !== undefined &&
        obj.measureColumnWidthItem.locators === undefined &&
        obj.measureColumnWidthItem.locator === undefined);
}
/**
 * Tests whether object is an instance of {@link IWeakMeasureColumnWidthItem}
 *
 * @public
 */
export function isWeakMeasureColumnWidthItem(obj) {
    return (!isEmpty(obj) &&
        obj.measureColumnWidthItem !== undefined &&
        obj.measureColumnWidthItem.locator !== undefined);
}
/**
 * @internal
 */
export function newMeasureColumnLocator(measureOrId) {
    var measureIdentifier = measureLocalId(measureOrId);
    return {
        measureLocatorItem: {
            measureIdentifier: measureIdentifier,
        },
    };
}
/**
 * Creates width item that will set width of a column which contains values of a row attribute.
 *
 * @param attributeOrId - attribute specified by value or by localId reference
 * @param width - width in pixels
 * @param allowGrowToFit - indicates whether the column is allowed to grow if the table's growToFit is enabled
 * @public
 */
export function newWidthForAttributeColumn(attributeOrId, width, allowGrowToFit) {
    var growToFitProp = allowGrowToFit !== undefined ? { allowGrowToFit: allowGrowToFit } : {};
    return {
        attributeColumnWidthItem: {
            attributeIdentifier: attributeLocalId(attributeOrId),
            width: __assign({ value: width }, growToFitProp),
        },
    };
}
/**
 * Creates width item that will set width for all measure columns in the table.
 *
 * @param width - width in pixels
 * @param allowGrowToFit - indicates whether the column is allowed to grow if the table's growToFit is enabled
 * @public
 */
export function newWidthForAllMeasureColumns(width, allowGrowToFit) {
    var growToFitProp = allowGrowToFit !== undefined ? { allowGrowToFit: allowGrowToFit } : {};
    return {
        measureColumnWidthItem: {
            width: __assign({ value: width }, growToFitProp),
        },
    };
}
/**
 * Creates width item that will set width for all columns containing values of the provided measure.
 *
 * @param measureOrId - measure specified either by value or by localId reference
 * @param width - width in pixels
 * @param allowGrowToFit - indicates whether the column is allowed to grow if the table's growToFit is enabled
 * @public
 */
export function newWidthForAllColumnsForMeasure(measureOrId, width, allowGrowToFit) {
    var locator = newMeasureColumnLocator(measureOrId);
    var growToFitProp = allowGrowToFit !== undefined ? { allowGrowToFit: allowGrowToFit } : {};
    return {
        measureColumnWidthItem: {
            width: __assign({ value: width }, growToFitProp),
            locator: locator,
        },
    };
}
/**
 * Creates width item that will set width for all columns containing values of the provided measure.
 *
 * See also {@link newAttributeColumnLocator} to learn more about the attribute column locators.
 *
 * @param measureOrId - measure specified either by value or by localId reference
 * @param locators - attribute locators to narrow down selection
 * @param width - width in pixels
 * @param allowGrowToFit - indicates whether the column is allowed to grow if the table's growToFit is enabled
 * @public
 */
export function newWidthForSelectedColumns(measureOrId, locators, width, allowGrowToFit) {
    var measureLocator = newMeasureColumnLocator(measureOrId);
    var growToFitProp = allowGrowToFit !== undefined && width !== "auto" ? { allowGrowToFit: allowGrowToFit } : {};
    // Note: beware here. The attribute locators _must_ come first for some obscure, impl dependent reason
    return {
        measureColumnWidthItem: {
            width: __assign({ value: width }, growToFitProp),
            locators: __spreadArrays(locators, [measureLocator]),
        },
    };
}
/**
 * Creates a new attribute column locator - this is used to narrow down location of measure columns in pivot table, where
 * measures are further scoped by different attribute elements - imagine pivot table with defined for measure 'Amount' and column
 * attribute 'Product'. The table will have multiple columns for the 'Amount' measure - each for different element of the
 * 'Product' attribute. In this context, identifying particular measure columns needs to be more specific.
 *
 * The attribute column locator can match either single element of particular attribute, or all elements of particular
 * attribute.
 *
 * @param attributeOrId - column attribute specified by either value or by localId reference
 * @param element - optionally specify attribute element URI or primary key; if not specified, the locator will match
 *  all elements of the attribute
 * @public
 */
export function newAttributeColumnLocator(attributeOrId, element) {
    return {
        attributeLocatorItem: {
            attributeIdentifier: attributeLocalId(attributeOrId),
            element: element,
        },
    };
}
