// @gooddata/sdk-ui-kit/esm/Menu/Menu.js
// (C) 2007-2018 GoodData Corporation
import React from "react";
import isFunction from "lodash/isFunction";
import { MenuState } from "./MenuState";
import { ControlledMenu } from "./ControlledMenu";
/**
 * @internal
 */
export var Menu = function (props) { return (React.createElement(MenuState, { opened: props.opened, defaultOpened: props.defaultOpened, onOpenedChange: props.onOpenedChange }, function (controlledProps) { return (React.createElement(ControlledMenu, { opened: controlledProps.opened, onOpenedChange: controlledProps.onOpenedChange, openAction: props.openAction, alignment: props.alignment, spacing: props.spacing, offset: props.offset, toggler: props.toggler, togglerWrapperClassName: props.togglerWrapperClassName, portalTarget: props.portalTarget, closeOnScroll: props.closeOnScroll }, isFunction(props.children)
    ? props.children({
        closeMenu: function () {
            return controlledProps.onOpenedChange({
                opened: false,
                source: "CLOSE_MENU_RENDER_PROP",
            });
        },
    })
    : props.children)); })); };
Menu.defaultProps = {
    openAction: "click",
    alignment: ["bottom", "right"],
    closeOnScroll: false,
};
