// @gooddata/sdk-ui/esm/execution/useExecution.js
import { __assign, __read } from "tslib";
import { useBackend, useResolveValuesWithPlaceholders, useWorkspace, } from "../base";
import { createExecution } from "./createExecution";
/**
 * This hook provides a simplified interface for creating executions.
 * It builds the execution on top of the backend passed to the BackendProvider, and workspace passed to the WorkspaceProvider.
 *
 * @beta
 */
export function useExecution(config) {
    var backend = useBackend(config.backend);
    var workspace = useWorkspace(config.workspace);
    var _a = __read(useResolveValuesWithPlaceholders([config.seriesBy, config.slicesBy, config.totals, config.filters, config.sortBy], config.placeholdersResolutionContext), 5), seriesBy = _a[0], slicesBy = _a[1], totals = _a[2], filters = _a[3], sortBy = _a[4];
    return createExecution(__assign(__assign({}, config), { backend: backend,
        workspace: workspace,
        seriesBy: seriesBy,
        slicesBy: slicesBy,
        totals: totals,
        filters: filters,
        sortBy: sortBy }));
}
