// @gooddata/sdk-ui-kit/esm/Dialog/ShareDialog/ShareDialogBase/useShareDialogBase.js
import { __read, __spread } from "tslib";
// (C) 2021 GoodData Corporation
import { useCallback, useMemo, useState } from "react";
import { areObjRefsEqual } from "@gooddata/sdk-model";
import { isGranteeUserInactive, } from "./types";
import { notInArrayFilter, getAppliedGrantees } from "./utils";
import { useGetAccessList } from "./backend/useGetAccessList";
import { mapShareStatusToGroupAll } from "../shareDialogMappers";
/**
 * @internal
 */
var useShareDialogState = function (isUnderLenientControl, isLocked) {
    var _a = __read(useState("ShareGrantee"), 2), dialogMode = _a[0], setDialogMode = _a[1];
    var _b = __read(useState(true), 2), isGranteesLoading = _b[0], setIsGranteesLoading = _b[1];
    var _c = __read(useState([]), 2), grantees = _c[0], setGrantees = _c[1];
    var _d = __read(useState([]), 2), granteesToAdd = _d[0], setGranteesToAdd = _d[1];
    var _e = __read(useState([]), 2), granteesToDelete = _e[0], setGranteesToDelete = _e[1];
    var _f = __read(useState(isUnderLenientControl), 2), isUnderLenientControlNow = _f[0], setUnderLenientControlNow = _f[1];
    var _g = __read(useState(isLocked), 2), isLockedNow = _g[0], setLockedNow = _g[1];
    var onSharedGranteeDelete = useCallback(function (grantee) {
        setGranteesToDelete(function (state) { return __spread(state, [grantee]); });
    }, []);
    var onAddedGranteeDelete = useCallback(function (grantee) {
        setGranteesToAdd(function (state) { return state.filter(function (g) { return !areObjRefsEqual(g.id, grantee.id); }); });
    }, []);
    var onGranteeAdd = useCallback(function (grantee) {
        setGranteesToAdd(function (state) { return __spread(state, [grantee]); });
    }, []);
    var onAddGranteeButtonClick = useCallback(function () {
        setDialogMode("AddGrantee");
    }, []);
    var onAddGranteeBackClick = useCallback(function () {
        setDialogMode("ShareGrantee");
        setGranteesToAdd([]);
    }, []);
    var onLoadGrantees = useCallback(function (grantees, groupAll) {
        if (groupAll) {
            setGrantees(__spread(grantees, [groupAll]));
        }
        else {
            setGrantees(grantees);
        }
        setIsGranteesLoading(false);
    }, []);
    var onUnderLenientControlChange = useCallback(function (isUnderLenientControl) {
        setUnderLenientControlNow(isUnderLenientControl);
    }, []);
    var onLockChange = useCallback(function (isLocked) {
        setLockedNow(isLocked);
    }, []);
    return {
        dialogMode: dialogMode,
        isGranteesLoading: isGranteesLoading,
        grantees: grantees,
        granteesToAdd: granteesToAdd,
        granteesToDelete: granteesToDelete,
        onLoadGrantees: onLoadGrantees,
        onSharedGranteeDelete: onSharedGranteeDelete,
        onAddedGranteeDelete: onAddedGranteeDelete,
        onGranteeAdd: onGranteeAdd,
        onAddGranteeButtonClick: onAddGranteeButtonClick,
        onAddGranteeBackClick: onAddGranteeBackClick,
        isUnderLenientControlNow: isUnderLenientControlNow,
        isLockedNow: isLockedNow,
        onUnderLenientControlChange: onUnderLenientControlChange,
        onLockChange: onLockChange,
    };
};
/**
 * @internal
 */
export var useShareDialogBase = function (props) {
    var sharedObject = props.sharedObject, currentUserRef = props.currentUserRef, onSubmit = props.onSubmit, onError = props.onError;
    var ref = sharedObject.ref, shareStatus = sharedObject.shareStatus, owner = sharedObject.owner, isUnderLenientControl = sharedObject.isUnderLenientControl, isLocked = sharedObject.isLocked;
    var _a = useShareDialogState(isUnderLenientControl, isLocked), dialogMode = _a.dialogMode, isGranteesLoading = _a.isGranteesLoading, isLockedNow = _a.isLockedNow, isUnderLenientControlNow = _a.isUnderLenientControlNow, grantees = _a.grantees, granteesToAdd = _a.granteesToAdd, granteesToDelete = _a.granteesToDelete, onLoadGrantees = _a.onLoadGrantees, onSharedGranteeDelete = _a.onSharedGranteeDelete, onAddedGranteeDelete = _a.onAddedGranteeDelete, onGranteeAdd = _a.onGranteeAdd, onAddGranteeButtonClick = _a.onAddGranteeButtonClick, onAddGranteeBackClick = _a.onAddGranteeBackClick, onLockChange = _a.onLockChange, onUnderLenientControlChange = _a.onUnderLenientControlChange;
    var onLoadGranteesSuccess = useCallback(function (result) {
        var groupAll = mapShareStatusToGroupAll(shareStatus);
        onLoadGrantees(result, groupAll);
    }, [onLoadGrantees, shareStatus]);
    useGetAccessList({ currentUserRef: currentUserRef, sharedObjectRef: ref, onSuccess: onLoadGranteesSuccess, onError: onError });
    var isShareDialogDirty = useMemo(function () {
        return (granteesToDelete.length !== 0 ||
            isLocked !== isLockedNow ||
            isUnderLenientControl !== isUnderLenientControlNow);
    }, [granteesToDelete, isLocked, isLockedNow, isUnderLenientControl, isUnderLenientControlNow]);
    var isAddDialogDirty = useMemo(function () {
        return granteesToAdd.length !== 0;
    }, [granteesToDelete, granteesToAdd]);
    var onSubmitShareGrantee = useCallback(function () {
        if (!isShareDialogDirty) {
            return;
        }
        onSubmit(grantees, granteesToAdd, granteesToDelete, isUnderLenientControlNow, isLockedNow);
    }, [
        grantees,
        granteesToAdd,
        granteesToDelete,
        isShareDialogDirty,
        isUnderLenientControlNow,
        isLockedNow,
        onSubmit,
    ]);
    var onSubmitAddGrantee = useCallback(function () {
        if (!isAddDialogDirty) {
            return;
        }
        onSubmit(grantees, granteesToAdd, granteesToDelete, isUnderLenientControlNow, isLockedNow);
    }, [
        grantees,
        granteesToAdd,
        granteesToDelete,
        isAddDialogDirty,
        isUnderLenientControlNow,
        isLockedNow,
        onSubmit,
    ]);
    var sharedGrantees = useMemo(function () {
        return notInArrayFilter(grantees, granteesToDelete);
    }, [grantees, granteesToDelete]);
    var appliedGranteesWithOwner = useMemo(function () {
        var appliedGrantees = getAppliedGrantees(grantees, granteesToAdd, granteesToDelete);
        if (isGranteeUserInactive(owner)) {
            return appliedGrantees;
        }
        return __spread(appliedGrantees, [owner]);
    }, [grantees, granteesToDelete, granteesToAdd]);
    return {
        onAddedGranteeDelete: onAddedGranteeDelete,
        onSharedGranteeDelete: onSharedGranteeDelete,
        onAddGranteeBackClick: onAddGranteeBackClick,
        onAddGranteeButtonClick: onAddGranteeButtonClick,
        onGranteeAdd: onGranteeAdd,
        onSubmitShareGrantee: onSubmitShareGrantee,
        onSubmitAddGrantee: onSubmitAddGrantee,
        isGranteesLoading: isGranteesLoading,
        granteesToAdd: granteesToAdd,
        dialogMode: dialogMode,
        isShareDialogDirty: isShareDialogDirty,
        isAddDialogDirty: isAddDialogDirty,
        sharedGrantees: sharedGrantees,
        appliedGranteesWithOwner: appliedGranteesWithOwner,
        onLockChange: onLockChange,
        onUnderLenientControlChange: onUnderLenientControlChange,
        isUnderLenientControlNow: isUnderLenientControlNow,
        isLockedNow: isLockedNow,
    };
};
