// @gooddata/sdk-ui-charts/esm/highcharts/adapter/plugins/dataLabelsColors.js
// (C) 2007-2021 GoodData Corporation
import flatMap from "lodash/flatMap";
// TODO remove this when IE11 support is dropped
import "svg-classlist-polyfill";
import { VisualizationTypes } from "@gooddata/sdk-ui";
import { getChartType, getVisibleSeries, isStacked, getShapeAttributes, } from "../../chartTypes/_chartCreators/helpers";
import { getDataLabelAttributes } from "../../chartTypes/_chartCreators/dataLabelsHelpers";
import { parseRGBColorCode } from "@gooddata/sdk-ui-vis-commons";
import { isOneOfTypes } from "../../chartTypes/_util/common";
var setWhiteColor = function (point) {
    point.dataLabel.element.childNodes[0].style.fill = "#fff";
    point.dataLabel.element.childNodes[0].style["text-shadow"] = "rgb(0, 0, 0) 0px 0px 1px";
    point.dataLabel.element.classList.remove("gd-contrast-label");
};
var setBlackColor = function (point) {
    point.dataLabel.element.childNodes[0].style.fill = "#000";
    point.dataLabel.element.childNodes[0].style["text-shadow"] = "none";
    point.dataLabel.element.classList.remove("gd-contrast-label");
};
var setContrastColor = function (point) {
    point.dataLabel.element.childNodes[0].style.fill = "";
    point.dataLabel.element.childNodes[0].style["text-shadow"] = "none";
    point.dataLabel.element.classList.add("gd-contrast-label");
};
var changeDataLabelsColor = function (condition, point) {
    return condition ? setWhiteColor(point) : setContrastColor(point);
};
function getVisiblePointsWithLabel(chart) {
    return flatMap(getVisibleSeries(chart), function (series) { return series.points; }).filter(function (point) { return point.dataLabel && point.graphic; });
}
function setBarDataLabelsColor(chart) {
    var points = getVisiblePointsWithLabel(chart);
    return points.forEach(function (point) {
        var labelDimensions = getDataLabelAttributes(point);
        var barDimensions = getShapeAttributes(point);
        var barRight = barDimensions.x + barDimensions.width;
        var barLeft = barDimensions.x;
        var labelLeft = labelDimensions.x;
        if (point.negative) {
            if (labelLeft > barLeft) {
                // labelRight is overlapping bar even it is outside of it
                setWhiteColor(point);
            }
            else {
                setContrastColor(point);
            }
        }
        else {
            if (labelLeft < barRight) {
                setWhiteColor(point);
            }
            else {
                setContrastColor(point);
            }
        }
    });
}
function setColumnDataLabelsColor(chart) {
    var points = getVisiblePointsWithLabel(chart);
    return points
        .filter(function (point) { return point.shapeArgs; }) // skip if shapeArgs missing (such as line points in line/column combo chart)
        .forEach(function (point) {
        var labelDimensions = getDataLabelAttributes(point);
        var columnDimensions = getShapeAttributes(point);
        var columnTop = columnDimensions.y + columnDimensions.height;
        var columnDown = columnDimensions.y;
        var labelDown = labelDimensions.y;
        if (point.negative) {
            changeDataLabelsColor(labelDown < columnDown, point);
        }
        else if (!isStacked(chart)) {
            changeDataLabelsColor(labelDown > columnTop, point);
        }
        else {
            changeDataLabelsColor(labelDown < columnTop, point);
        }
    });
}
export function isWhiteNotContrastEnough(color) {
    // to keep first 17 colors from our default palette with white labels
    var HIGHCHARTS_CONTRAST_THRESHOLD = 530;
    var _a = parseRGBColorCode(color), R = _a.R, G = _a.G, B = _a.B;
    var lightnessHCH = R + G + B;
    return lightnessHCH > HIGHCHARTS_CONTRAST_THRESHOLD;
}
function setContrastLabelsColor(chart) {
    var points = getVisiblePointsWithLabel(chart);
    return points.forEach(function (point) {
        if (isWhiteNotContrastEnough(point.color)) {
            setBlackColor(point);
        }
        else {
            setWhiteColor(point);
        }
    });
}
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function extendDataLabelColors(Highcharts) {
    Highcharts.Chart.prototype.callbacks.push(function (chart) {
        var type = getChartType(chart);
        var changeLabelColor = function () {
            if (type === VisualizationTypes.BAR) {
                setTimeout(function () {
                    setBarDataLabelsColor(chart);
                }, 500);
            }
            else if (isOneOfTypes(type, [VisualizationTypes.COLUMN, VisualizationTypes.PIE])) {
                setTimeout(function () {
                    setColumnDataLabelsColor(chart);
                }, 500);
            }
            else if (isOneOfTypes(type, [VisualizationTypes.HEATMAP, VisualizationTypes.TREEMAP])) {
                setContrastLabelsColor(chart);
            }
        };
        changeLabelColor();
        Highcharts.addEvent(chart, "redraw", changeLabelColor);
    });
}
