// @gooddata/sdk-ui-kit/esm/Bubble/BubbleHoverTrigger.js
import { __extends } from "tslib";
// (C) 2007-2020 GoodData Corporation
import { BubbleTrigger } from "./BubbleTrigger";
export var SHOW_DELAY = 425;
export var HIDE_DELAY = 200;
/**
 * @internal
 */
var BubbleHoverTrigger = /** @class */ (function (_super) {
    __extends(BubbleHoverTrigger, _super);
    function BubbleHoverTrigger() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BubbleHoverTrigger.prototype.componentWillUnmount = function () {
        this.cancelBubbleVisibilityChange();
    };
    BubbleHoverTrigger.prototype.cancelBubbleVisibilityChange = function () {
        if (this.scheduleId) {
            window.clearTimeout(this.scheduleId);
        }
    };
    BubbleHoverTrigger.prototype.scheduleBubbleVisibilityChange = function (visible, delay) {
        var _this = this;
        if (delay === void 0) { delay = 0; }
        this.cancelBubbleVisibilityChange();
        this.scheduleId = window.setTimeout(function () {
            _this.changeBubbleVisibility(visible);
            var hoverHideDelay = _this.props.hoverHideDelay;
            if (visible && hoverHideDelay) {
                _this.scheduleBubbleVisibilityChange(false, hoverHideDelay);
            }
        }, delay);
    };
    BubbleHoverTrigger.prototype.eventListeners = function () {
        return {
            onMouseEnter: this.scheduleBubbleVisibilityChange.bind(this, true, this.props.showDelay),
            onMouseLeave: this.scheduleBubbleVisibilityChange.bind(this, false, this.props.hideDelay),
        };
    };
    BubbleHoverTrigger.defaultProps = {
        showDelay: SHOW_DELAY,
        hideDelay: HIDE_DELAY,
        hoverHideDelay: 0,
        eventsOnBubble: false,
        tagName: "span",
    };
    return BubbleHoverTrigger;
}(BubbleTrigger));
export { BubbleHoverTrigger };
