// @gooddata/sdk-ui-dashboard/esm/_staging/metadata/objRefMap.js
import { __assign, __read } from "tslib";
import { insightId, insightRef, insightUri, isIdentifierRef, } from "@gooddata/sdk-model";
import values from "lodash/values";
/**
 * Utility class that assists with type-agnostic mapping of metadata objects by ObjRef.
 *
 * Problem
 * =======
 *
 * The challenges with ObjRef's start in context of backend that supports both uri and id ref (e.g. bear) and the client.
 *
 * Backend according to contract creates one type of ref - uri ref - so that is fine. However when instances of `ref` are created
 * by the client code and are passed in through the public API (as is the case with the dashboard component) - problems start.
 *
 * For clients it is often more convenient to use ID refs.. because they are transferable across workspaces and because
 * they appear in the catalog export.
 *
 * Doing strict ref-to-ref matching between user input and the data stored in state will result in no matches because
 * the types of ref's do not match.
 *
 * ---
 *
 * This class addresses the problem by having the `get` method check the type of ObjRef first and then perform
 * lookups into either id to item or uri to item mapping.
 *
 * @alpha
 */
var ObjRefMap = /** @class */ (function () {
    function ObjRefMap(config) {
        var _this = this;
        this.config = config;
        this[Symbol.toStringTag] = "ObjRefMap";
        this.size = 0;
        this.items = [];
        this.itemsByIdentifier = {};
        this.itemsByUri = {};
        this.idRefToKey = function (identifier, type) {
            return !_this.config.strictTypeCheck || !type ? identifier : identifier + "#" + type;
        };
        this.addItem = function (item) {
            var _a = _this.config, refExtract = _a.refExtract, uriExtract = _a.uriExtract, idExtract = _a.idExtract;
            var uri = uriExtract(item);
            var identifier = idExtract(item);
            _this.itemsByUri[uri] = item;
            _this.itemsByIdentifier[_this.idRefToKey(identifier, _this.config.type)] = item;
            _this.items.push([refExtract(item), item]);
            _this.size++;
        };
        this.cleanupUnmappedItems = function () {
            var allItems = values(_this.itemsByUri).concat(values(_this.itemsByIdentifier));
            _this.items = _this.items.filter(function (_a) {
                var _b = __read(_a, 2), _ = _b[0], item = _b[1];
                return allItems.find(function (mappedItem) { return mappedItem === item; });
            });
        };
        this.fromItems = function (items) {
            items.forEach(_this.addItem);
            _this.cleanupUnmappedItems();
            return _this;
        };
    }
    ObjRefMap.prototype[Symbol.iterator] = function () {
        return this.items[Symbol.iterator]();
    };
    ObjRefMap.prototype.entries = function () {
        return this.items[Symbol.iterator]();
    };
    ObjRefMap.prototype.get = function (key) {
        if (isIdentifierRef(key)) {
            var strictMatch = this.itemsByIdentifier[this.idRefToKey(key.identifier, key.type)];
            if (!strictMatch && !key.type && this.config.strictTypeCheck) {
                // eslint-disable-next-line no-console
                console.warn("You are working with an analytical backend which can only match entities by idRefs that contain both identifier and type. However, while trying to find match an object by ref your code supplied just the identifier without type.");
            }
            return strictMatch;
        }
        return this.itemsByUri[key.uri];
    };
    ObjRefMap.prototype.has = function (key) {
        return this.get(key) !== undefined;
    };
    ObjRefMap.prototype.keys = function () {
        return this.items.map(function (i) { return i[0]; })[Symbol.iterator]();
    };
    ObjRefMap.prototype.values = function () {
        return this.items.map(function (i) { return i[1]; })[Symbol.iterator]();
    };
    return ObjRefMap;
}());
export { ObjRefMap };
var metadataObjectExtractors = {
    idExtract: function (i) { return i.id; },
    uriExtract: function (i) { return i.uri; },
    refExtract: function (i) { return i.ref; },
};
/**
 * Creates {@link ObjRefMap} for catalog date datasets. Either normal attributes or catalog date attributes.
 *
 * @param items - items to add into mapping
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newCatalogDateDatasetMap(items, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap({
        type: "dataSet",
        strictTypeCheck: strictTypeCheck,
        idExtract: function (i) { return i.dataSet.id; },
        uriExtract: function (i) { return i.dataSet.uri; },
        refExtract: function (i) { return i.dataSet.ref; },
    });
    return map.fromItems(items);
}
/**
 * Creates {@link ObjRefMap} for catalog attribute items. Either normal attributes or catalog date attributes.
 *
 * @param items - items to add into mapping
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newCatalogAttributeMap(items, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap({
        type: "attribute",
        strictTypeCheck: strictTypeCheck,
        idExtract: function (i) { return i.attribute.id; },
        uriExtract: function (i) { return i.attribute.uri; },
        refExtract: function (i) { return i.attribute.ref; },
    });
    return map.fromItems(items);
}
/**
 * Creates {@link ObjRefMap} for catalog measure items.
 *
 * @param items - items to add into mapping
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newCatalogMeasureMap(items, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap({
        type: "measure",
        strictTypeCheck: strictTypeCheck,
        idExtract: function (i) { return i.measure.id; },
        uriExtract: function (i) { return i.measure.uri; },
        refExtract: function (i) { return i.measure.ref; },
    });
    return map.fromItems(items);
}
/**
 * Creates {@link ObjRefMap} for attribute display form metadata objects.
 *
 * @param items - items to add into map
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newDisplayFormMap(items, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap(__assign({ type: "displayForm", strictTypeCheck: strictTypeCheck }, metadataObjectExtractors));
    return map.fromItems(items);
}
/**
 * Creates {@link ObjRefMap} for attribute metadata objects.
 *
 * @param items - items to add into map
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newAttributeMap(items, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap(__assign({ type: "attribute", strictTypeCheck: strictTypeCheck }, metadataObjectExtractors));
    return map.fromItems(items);
}
/**
 * Creates {@link ObjRefMap} for any object of any type granted that it contains `ref` property with ObjRef.
 *
 * The map will be setup so that it extracts id and uri from the `ref` - thus ensuring that objects that are
 * both uri and identifier refs are indexed properly.
 *
 * Storing objects whose `ref` is of both types then allows lookup of those objects externally using either id
 * or uri.
 *
 * @param items - items to insert
 * @param type - type of objects, may be undefined
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newMapForObjectWithIdentity(items, type, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap({
        type: type,
        strictTypeCheck: strictTypeCheck,
        idExtract: function (i) { return i.identifier; },
        uriExtract: function (i) { return i.uri; },
        refExtract: function (i) { return i.ref; },
    });
    return map.fromItems(items);
}
/**
 * Creates {@link ObjRefMap} for insights.
 *
 * @param items - items to add into mapping
 * @param strictTypeCheck - whether to do strict type checking when getting by identifierRef
 * @alpha
 */
export function newInsightMap(items, strictTypeCheck) {
    if (strictTypeCheck === void 0) { strictTypeCheck = false; }
    var map = new ObjRefMap({
        type: "insight",
        strictTypeCheck: strictTypeCheck,
        idExtract: insightId,
        uriExtract: insightUri,
        refExtract: insightRef,
    });
    return map.fromItems(items);
}
