// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/comboChart/PluggableComboChart.js
import { __assign, __extends, __spreadArrays } from "tslib";
// (C) 2019 GoodData Corporation
import React from "react";
import { render } from "react-dom";
import cloneDeep from "lodash/cloneDeep";
import set from "lodash/set";
import without from "lodash/without";
import { BucketNames, VisualizationTypes } from "@gooddata/sdk-ui";
import { isAreaChart, isLineChart } from "@gooddata/sdk-ui-charts";
import { AXIS, AXIS_NAME } from "../../../constants/axis";
import { BUCKETS, METRIC } from "../../../constants/bucket";
import { PROPERTY_CONTROLS_DUAL_AXIS } from "../../../constants/properties";
import { COMBO_CHART_SUPPORTED_PROPERTIES } from "../../../constants/supportedProperties";
import { COMBO_CHART_UICONFIG } from "../../../constants/uiConfig";
import { configureOverTimeComparison, configurePercent } from "../../../utils/bucketConfig";
import { applyUiConfig, findBucket, getAllAttributeItemsWithPreference, getAllMeasuresShowOnSecondaryAxis, getBucketItemsByType, getBucketItemsWithExcludeByType, getMeasureItems, hasBucket, sanitizeFilters, setMeasuresShowOnSecondaryAxis, } from "../../../utils/bucketHelper";
import { getMasterMeasuresCount } from "../../../utils/bucketRules";
import { getReferencePointWithSupportedProperties, isDualAxisOrSomeSecondaryAxisMeasure, setSecondaryMeasures, } from "../../../utils/propertiesHelper";
import { removeSort } from "../../../utils/sort";
import { setComboChartUiConfig } from "../../../utils/uiConfigHelpers/comboChartUiConfigHelper";
import LineChartBasedConfigurationPanel from "../../configurationPanels/LineChartBasedConfigurationPanel";
import { PluggableBaseChart } from "../baseChart/PluggableBaseChart";
import { insightBuckets, bucketsIsEmpty } from "@gooddata/sdk-model";
var PluggableComboChart = /** @class */ (function (_super) {
    __extends(PluggableComboChart, _super);
    function PluggableComboChart(props) {
        var _a, _b, _c;
        var _this = _super.call(this, props) || this;
        _this.primaryChartType = VisualizationTypes.COLUMN;
        _this.secondaryChartType = VisualizationTypes.COLUMN;
        _this.type = VisualizationTypes.COMBO;
        _this.axis = AXIS.DUAL;
        _this.secondaryAxis = AXIS_NAME.SECONDARY_Y;
        var propertiesControls = (_a = props.visualizationProperties) === null || _a === void 0 ? void 0 : _a.controls;
        _this.primaryChartType = (_b = propertiesControls === null || propertiesControls === void 0 ? void 0 : propertiesControls.primaryChartType) !== null && _b !== void 0 ? _b : VisualizationTypes.COLUMN;
        _this.secondaryChartType = (_c = propertiesControls === null || propertiesControls === void 0 ? void 0 : propertiesControls.secondaryChartType) !== null && _c !== void 0 ? _c : VisualizationTypes.COLUMN;
        _this.supportedPropertiesList = _this.getSupportedPropertiesList();
        _this.defaultControlsProperties = {
            stackMeasures: _this.isStackMeasuresByDefault(),
        };
        _this.initializeProperties(props.visualizationProperties);
        return _this;
    }
    PluggableComboChart.prototype.getSupportedPropertiesList = function () {
        return COMBO_CHART_SUPPORTED_PROPERTIES[this.axis] || [];
    };
    PluggableComboChart.prototype.getUiConfig = function () {
        return cloneDeep(__assign(__assign({}, COMBO_CHART_UICONFIG), { optionalStacking: {
                supported: true,
                disabled: isLineChart(this.primaryChartType),
                stackMeasures: this.isStackMeasuresByDefault(),
            } }));
    };
    PluggableComboChart.prototype.getExtendedReferencePoint = function (referencePoint) {
        var _a, _b, _c, _d, _e, _f;
        var clonedReferencePoint = cloneDeep(referencePoint);
        var properties = this.configureChartTypes(clonedReferencePoint);
        this.primaryChartType = (_b = (_a = properties === null || properties === void 0 ? void 0 : properties.controls) === null || _a === void 0 ? void 0 : _a.primaryChartType) !== null && _b !== void 0 ? _b : VisualizationTypes.COLUMN;
        this.secondaryChartType = (_d = (_c = properties === null || properties === void 0 ? void 0 : properties.controls) === null || _c === void 0 ? void 0 : _c.secondaryChartType) !== null && _d !== void 0 ? _d : VisualizationTypes.COLUMN;
        var newReferencePoint = __assign(__assign({}, clonedReferencePoint), { properties: properties, uiConfig: this.getUiConfig() });
        this.defaultControlsProperties = {
            stackMeasures: this.isStackMeasuresByDefault(),
        };
        this.configureBuckets(newReferencePoint);
        newReferencePoint = setSecondaryMeasures(newReferencePoint, this.secondaryAxis);
        this.axis = (_f = (_e = newReferencePoint === null || newReferencePoint === void 0 ? void 0 : newReferencePoint.uiConfig) === null || _e === void 0 ? void 0 : _e.axis) !== null && _f !== void 0 ? _f : AXIS.PRIMARY;
        this.supportedPropertiesList = this.getSupportedPropertiesList();
        newReferencePoint = setComboChartUiConfig(newReferencePoint, this.intl, this.type);
        newReferencePoint = configurePercent(newReferencePoint, this.isPercentDisabled(newReferencePoint));
        newReferencePoint = configureOverTimeComparison(newReferencePoint, !!this.featureFlags["enableWeekFilters"]);
        newReferencePoint = getReferencePointWithSupportedProperties(newReferencePoint, this.supportedPropertiesList);
        newReferencePoint = applyUiConfig(newReferencePoint);
        newReferencePoint = removeSort(newReferencePoint);
        return Promise.resolve(sanitizeFilters(newReferencePoint));
    };
    PluggableComboChart.prototype.isStackMeasuresByDefault = function () {
        return isAreaChart(this.primaryChartType);
    };
    PluggableComboChart.prototype.configureBuckets = function (extReferencePoint) {
        var _a, _b, _c, _d, _e, _f, _g;
        var buckets = (_a = extReferencePoint === null || extReferencePoint === void 0 ? void 0 : extReferencePoint.buckets) !== null && _a !== void 0 ? _a : [];
        var attributes = getAllAttributeItemsWithPreference(buckets, [
            BucketNames.TREND,
            BucketNames.VIEW,
        ]).slice(0, 1);
        var measures = [];
        var secondaryMeasures = [];
        // ref. point has both my buckets -> reuse them fully
        if (hasBucket(buckets, BucketNames.MEASURES) && hasBucket(buckets, BucketNames.SECONDARY_MEASURES)) {
            measures = getBucketItemsByType(buckets, BucketNames.MEASURES, [METRIC]);
            secondaryMeasures = getBucketItemsByType(buckets, BucketNames.SECONDARY_MEASURES, [METRIC]);
            var restMeasures = getBucketItemsWithExcludeByType(buckets, [BucketNames.MEASURES, BucketNames.SECONDARY_MEASURES], [METRIC]);
            secondaryMeasures = secondaryMeasures.concat(restMeasures);
        }
        else {
            // transform from dual axis chart to combo chart
            var allMeasures = getMeasureItems(buckets);
            secondaryMeasures = getAllMeasuresShowOnSecondaryAxis(buckets);
            measures = without.apply(void 0, __spreadArrays([allMeasures], secondaryMeasures));
        }
        var isDualAxisEnabled = isDualAxisOrSomeSecondaryAxisMeasure(extReferencePoint, secondaryMeasures);
        set(extReferencePoint, PROPERTY_CONTROLS_DUAL_AXIS, isDualAxisEnabled);
        var primaryChartType = (_d = (_c = (_b = extReferencePoint === null || extReferencePoint === void 0 ? void 0 : extReferencePoint.properties) === null || _b === void 0 ? void 0 : _b.controls) === null || _c === void 0 ? void 0 : _c.primaryChartType) !== null && _d !== void 0 ? _d : VisualizationTypes.COLUMN;
        var secondaryChartType = (_g = (_f = (_e = extReferencePoint === null || extReferencePoint === void 0 ? void 0 : extReferencePoint.properties) === null || _e === void 0 ? void 0 : _e.controls) === null || _f === void 0 ? void 0 : _f.secondaryChartType) !== null && _g !== void 0 ? _g : VisualizationTypes.LINE;
        set(extReferencePoint, BUCKETS, [
            {
                localIdentifier: BucketNames.MEASURES,
                items: setMeasuresShowOnSecondaryAxis(measures, false),
                chartType: primaryChartType,
            },
            {
                localIdentifier: BucketNames.SECONDARY_MEASURES,
                items: setMeasuresShowOnSecondaryAxis(secondaryMeasures, isDualAxisEnabled),
                chartType: secondaryChartType,
            },
            {
                localIdentifier: BucketNames.VIEW,
                items: attributes,
            },
        ]);
    };
    PluggableComboChart.prototype.configureChartTypes = function (referencePoint) {
        var _a, _b, _c, _d, _e, _f, _g;
        var buckets = (_a = referencePoint === null || referencePoint === void 0 ? void 0 : referencePoint.buckets) !== null && _a !== void 0 ? _a : [];
        var controls = (_c = (_b = referencePoint === null || referencePoint === void 0 ? void 0 : referencePoint.properties) === null || _b === void 0 ? void 0 : _b.controls) !== null && _c !== void 0 ? _c : {};
        var primaryChartType = (_e = (_d = findBucket(buckets, BucketNames.MEASURES)) === null || _d === void 0 ? void 0 : _d.chartType) !== null && _e !== void 0 ? _e : (controls.primaryChartType || VisualizationTypes.COLUMN);
        var secondaryChartType = (_g = (_f = findBucket(buckets, BucketNames.SECONDARY_MEASURES)) === null || _f === void 0 ? void 0 : _f.chartType) !== null && _g !== void 0 ? _g : (controls.secondaryChartType || VisualizationTypes.LINE);
        if (primaryChartType || secondaryChartType) {
            return __assign(__assign({}, referencePoint.properties), { controls: __assign(__assign({}, controls), { primaryChartType: primaryChartType,
                    secondaryChartType: secondaryChartType }) });
        }
        return referencePoint.properties;
    };
    PluggableComboChart.prototype.isPercentDisabled = function (extReferencePoint) {
        var _a;
        if (this.axis === AXIS.DUAL) {
            return false;
        }
        var buckets = (_a = extReferencePoint === null || extReferencePoint === void 0 ? void 0 : extReferencePoint.buckets) !== null && _a !== void 0 ? _a : [];
        var primaryMasterMeasures = getMasterMeasuresCount(buckets, BucketNames.MEASURES);
        var secondaryMasterMeasures = getMasterMeasuresCount(buckets, BucketNames.SECONDARY_MEASURES);
        // disable percent if there is more than one measure on primary/secondary y-axis
        return primaryMasterMeasures + secondaryMasterMeasures > 1;
    };
    PluggableComboChart.prototype.isDataPointsControlDisabled = function (insight) {
        var measureBucketsOfNonColumnCharts = [
            [this.primaryChartType, BucketNames.MEASURES],
            [this.secondaryChartType, BucketNames.SECONDARY_MEASURES],
        ]
            .filter(function (_a) {
            var chartType = _a[0];
            return chartType !== VisualizationTypes.COLUMN;
        })
            .map(function (_a) {
            var bucketId = _a[1];
            return insightBuckets(insight, bucketId);
        });
        return (measureBucketsOfNonColumnCharts.length === 0 ||
            measureBucketsOfNonColumnCharts.every(function (bucket) { return bucketsIsEmpty(bucket); }));
    };
    PluggableComboChart.prototype.renderConfigurationPanel = function (insight) {
        if (document.querySelector(this.configPanelElement)) {
            render(React.createElement(LineChartBasedConfigurationPanel, { locale: this.locale, references: this.references, properties: this.visualizationProperties, propertiesMeta: this.propertiesMeta, insight: insight, colors: this.colors, pushData: this.handlePushData, type: this.type, isError: this.getIsError(), isLoading: this.isLoading, featureFlags: this.featureFlags, axis: this.axis, panelConfig: {
                    isDataPointsControlDisabled: this.isDataPointsControlDisabled(insight),
                }, dataLabelDefaultValue: "auto" }), document.querySelector(this.configPanelElement));
        }
    };
    return PluggableComboChart;
}(PluggableBaseChart));
export { PluggableComboChart };
