// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/scatterPlot/scatterPlotSeries.js
// (C) 2020-2021 GoodData Corporation
import { BucketNames } from "@gooddata/sdk-ui";
import { parseValue } from "../_util/common";
export function getScatterPlotSeries(dv, 
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
stackByAttribute, colorStrategy) {
    var primaryMeasuresBucketEmpty = dv.def().isBucketEmpty(BucketNames.MEASURES);
    var secondaryMeasuresBucketEmpty = dv.def().isBucketEmpty(BucketNames.SECONDARY_MEASURES);
    var data = dv
        .rawData()
        .twoDimData()
        .map(function (seriesItem, seriesIndex) {
        var values = seriesItem.map(function (value) {
            return parseValue(value);
        });
        return {
            x: !primaryMeasuresBucketEmpty ? values[0] : 0,
            y: !secondaryMeasuresBucketEmpty ? (primaryMeasuresBucketEmpty ? values[0] : values[1]) : 0,
            name: stackByAttribute ? stackByAttribute.items[seriesIndex].attributeHeaderItem.name : "",
        };
    });
    return [
        {
            turboThreshold: 0,
            color: colorStrategy.getColorByIndex(0),
            legendIndex: 0,
            data: data,
            seriesIndex: 0,
        },
    ];
}
