// @gooddata/api-client-bear/esm/dashboard/dashboard.js
import { __awaiter, __generator } from "tslib";
// (C) 2019-2020 GoodData Corporation
import { sanitizeFiltersForExport } from "@gooddata/api-model-bear";
import { handleHeadPolling } from "../util";
import { isExportFinished } from "../utils/export";
var DashboardModule = /** @class */ (function () {
    function DashboardModule(xhr) {
        this.xhr = xhr;
    }
    DashboardModule.prototype.exportToPdf = function (projectId, dashboardUri, filters, pollingOptions) {
        if (filters === void 0) { filters = []; }
        if (pollingOptions === void 0) { pollingOptions = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var sanitizedFilters, payload, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        sanitizedFilters = sanitizeFiltersForExport(filters);
                        payload = this.getDashboardExportPayload(dashboardUri, sanitizedFilters);
                        return [4 /*yield*/, this.xhr.post("/gdc/internal/projects/" + projectId + "/exportDashboard", { body: payload })];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, this.pollPdfFile(response, pollingOptions)];
                }
            });
        });
    };
    DashboardModule.prototype.pollPdfFile = function (response, pollingOptions) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            return __generator(this, function (_a) {
                data = response.getData();
                return [2 /*return*/, handleHeadPolling(this.xhr.head.bind(this.xhr), data.uri, isExportFinished, pollingOptions)];
            });
        });
    };
    DashboardModule.prototype.getDashboardExportPayload = function (dashboardUri, filters) {
        if (filters.length) {
            return {
                dashboardExport: {
                    dashboardUri: dashboardUri,
                    filters: filters,
                },
            };
        }
        // minimize payload
        return {
            dashboardExport: {
                dashboardUri: dashboardUri,
            },
        };
    };
    return DashboardModule;
}());
export { DashboardModule };
