// @gooddata/sdk-ui-dashboard/esm/presentation/widget/insight/DefaultDashboardInsight/InsightDrillDialog/useDrillDialogInsightDrills.js
import { __assign, __read } from "tslib";
// (C) 2020-2021 GoodData Corporation
import { useCallback, useState } from "react";
import isEqual from "lodash/isEqual";
import { useDashboardSelector, selectImplicitDrillsByAvailableDrillTargets, selectDrillableItemsByAvailableDrillTargets, } from "../../../../../model";
import { DataViewFacade, isSomeHeaderPredicateMatched, } from "@gooddata/sdk-ui";
/**
 * @internal
 */
export var useDrillDialogInsightDrills = function (_a) {
    var widget = _a.widget, insight = _a.insight, onDrillFn = _a.onDrill;
    // Drilling
    var _b = __read(useState(), 2), drillTargets = _b[0], setDrillTargets = _b[1];
    var onPushData = useCallback(function (data) {
        if ((data === null || data === void 0 ? void 0 : data.availableDrillTargets) && !isEqual(drillTargets, data.availableDrillTargets)) {
            setDrillTargets(data.availableDrillTargets);
        }
    }, [drillTargets]);
    var implicitDrillDefinitions = useDashboardSelector(selectImplicitDrillsByAvailableDrillTargets(drillTargets));
    var drillableItems = useDashboardSelector(selectDrillableItemsByAvailableDrillTargets(drillTargets));
    var onDrill = onDrillFn
        ? function (event) {
            var facade = DataViewFacade.for(event.dataView);
            var matchingImplicitDrillDefinitions = implicitDrillDefinitions.filter(function (info) {
                var _a;
                return (_a = event.drillContext.intersection) === null || _a === void 0 ? void 0 : _a.some(function (intersection) {
                    return isSomeHeaderPredicateMatched(info.predicates, intersection.header, facade);
                });
            });
            var drillEvent = __assign(__assign({}, event), { widgetRef: widget.ref, drillDefinitions: matchingImplicitDrillDefinitions.map(function (info) { return info.drillDefinition; }) });
            return (typeof onDrillFn === "function" &&
                onDrillFn(drillEvent, {
                    insight: insight,
                    widget: widget,
                }));
        }
        : undefined;
    return {
        drillableItems: drillableItems,
        onPushData: onPushData,
        onDrill: onDrill,
    };
};
