// @gooddata/sdk-backend-bear/esm/backend/workspace/accessControl/index.js
import { __awaiter, __generator } from "tslib";
import { objRefToUri } from "../../../utils/api";
import { convertGranteeEntry } from "../../../convertors/fromBackend/GranteeEntryConverter";
var BearWorkspaceAccessControlService = /** @class */ (function () {
    function BearWorkspaceAccessControlService(authCall, workspace) {
        this.authCall = authCall;
        this.workspace = workspace;
    }
    BearWorkspaceAccessControlService.prototype.getAccessList = function (sharedObject) {
        return __awaiter(this, void 0, void 0, function () {
            var objectUri, granteesList, items;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, objRefToUri(sharedObject, this.workspace, this.authCall)];
                    case 1:
                        objectUri = _a.sent();
                        return [4 /*yield*/, this.authCall(function (sdk) { return sdk.project.getGranteesInfo(objectUri, {}); })];
                    case 2:
                        granteesList = _a.sent();
                        items = granteesList.grantees.items;
                        return [2 /*return*/, items.map(convertGranteeEntry)];
                }
            });
        });
    };
    BearWorkspaceAccessControlService.prototype.grantAccess = function (sharedObject, grantees) {
        return __awaiter(this, void 0, void 0, function () {
            var objectUri, granteeUris;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, objRefToUri(sharedObject, this.workspace, this.authCall)];
                    case 1:
                        objectUri = _a.sent();
                        return [4 /*yield*/, Promise.all(grantees.map(function (grantee) { return objRefToUri(grantee.granteeRef, _this.workspace, _this.authCall); }))];
                    case 2:
                        granteeUris = _a.sent();
                        return [2 /*return*/, this.authCall(function (sdk) { return sdk.project.addGrantees(objectUri, granteeUris); })];
                }
            });
        });
    };
    BearWorkspaceAccessControlService.prototype.revokeAccess = function (sharedObject, grantees) {
        return __awaiter(this, void 0, void 0, function () {
            var objectUri, granteeUris;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, objRefToUri(sharedObject, this.workspace, this.authCall)];
                    case 1:
                        objectUri = _a.sent();
                        return [4 /*yield*/, Promise.all(grantees.map(function (grantee) { return objRefToUri(grantee.granteeRef, _this.workspace, _this.authCall); }))];
                    case 2:
                        granteeUris = _a.sent();
                        return [2 /*return*/, this.authCall(function (sdk) { return sdk.project.removeGrantees(objectUri, granteeUris); })];
                }
            });
        });
    };
    return BearWorkspaceAccessControlService;
}());
export { BearWorkspaceAccessControlService };
