// @gooddata/sdk-ui-ext/esm/internal/dashboardEmbedding/DashboardLayout/utils/legacy.js
import { DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT_PX } from "../../../components/pluggableVisualizations/constants";
import { DASHBOARD_LAYOUT_GRID_COLUMNS_COUNT, DASHBOARD_LAYOUT_VIS_TYPE, } from "../../DashboardLayout/constants";
var RATIO_16_9 = 9 / 16;
var HEADER_HEIGHT = 330;
var PADDING_BOTTOM = 55;
var HEADLINE_OUTER_HEIGHT = 48 + 20; // item-headline-outer + padding-top/bottom
export function calculateGeoPushpinWidgetHeight(windowHeight, visualizationItemWidth) {
    var widgetHeight = Math.round(visualizationItemWidth * RATIO_16_9);
    var widgetMaxHeight = windowHeight - HEADER_HEIGHT - PADDING_BOTTOM;
    var defaultVisualizationHeight = DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT_PX - HEADLINE_OUTER_HEIGHT;
    var maxHeight = widgetMaxHeight > defaultVisualizationHeight ? widgetMaxHeight : defaultVisualizationHeight;
    return {
        height: widgetHeight,
        maxHeight: maxHeight,
    };
}
export function isGeoPushpin(visType) {
    return visType === DASHBOARD_LAYOUT_VIS_TYPE.pushpin;
}
export function isFullWidthGeoPushpin(currentColumnWidth, visType) {
    return isGeoPushpin(visType) && currentColumnWidth === DASHBOARD_LAYOUT_GRID_COLUMNS_COUNT;
}
export function getGeoPushpinWidgetStyle(visType, visualizationItemWidth, currentColumnWidth, windowHeight, enableCustomHeight) {
    if (isFullWidthGeoPushpin(currentColumnWidth, visType) && !enableCustomHeight) {
        var _a = calculateGeoPushpinWidgetHeight(windowHeight, visualizationItemWidth), height = _a.height, maxHeight = _a.maxHeight;
        return {
            height: height,
            maxHeight: maxHeight,
        };
    }
    return null;
}
