// @gooddata/sdk-ui-dashboard/esm/presentation/shareDialog/DefaultShareDialog.js
// (C) 2020-2022 GoodData Corporation
import React, { useMemo } from "react";
import { ShareDialog } from "@gooddata/sdk-ui-kit";
import { useIntl } from "react-intl";
import { useDashboardSelector, selectLocale } from "../../model";
/**
 * @alpha
 */
export var DefaultShareDialog = function (props) {
    var workspace = props.workspace, backend = props.backend, isVisible = props.isVisible, sharedObject = props.sharedObject, currentUserRef = props.currentUserRef, isLockingSupported = props.isLockingSupported, onApply = props.onApply, onCancel = props.onCancel, onError = props.onError;
    var locale = useDashboardSelector(selectLocale);
    var intl = useIntl();
    var labels = useMemo(function () { return ({
        accessTypeLabel: intl.formatMessage({ id: "dashboard.shareDialog.lock.label" }),
        accessRegimeLabel: intl.formatMessage({
            id: "dashboard.shareDialog.underLenientControl.label",
        }),
        removeAccessCreatorTooltip: intl.formatMessage({
            id: "dashboard.shareDialog.removeAccessCreatorTooltip",
        }),
        removeAccessGranteeTooltip: intl.formatMessage({
            id: "dashboard.shareDialog.removeAccessGranteeTooltip",
        }),
    }); }, [intl]);
    if (!isVisible) {
        return null;
    }
    return (React.createElement(ShareDialog, { locale: locale, backend: backend, workspace: workspace, sharedObject: sharedObject, currentUserRef: currentUserRef, onApply: onApply, onCancel: onCancel, onError: onError, isLockingSupported: isLockingSupported, labels: labels }));
};
