// @gooddata/sdk-ui-dashboard/esm/model/events/alerts.js
import { eventGuard } from "./util";
export function alertCreated(ctx, alert, correlationId) {
    return {
        type: "GDC.DASH/EVT.ALERT.CREATED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            alert: alert,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAlertCreated}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAlertCreated = eventGuard("GDC.DASH/EVT.ALERT.CREATED");
export function alertsRemoved(ctx, alerts, correlationId) {
    return {
        type: "GDC.DASH/EVT.ALERTS.REMOVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            alerts: alerts,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAlertsRemoved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAlertsRemoved = eventGuard("GDC.DASH/EVT.ALERTS.REMOVED");
export function alertUpdated(ctx, updated, correlationId) {
    return {
        type: "GDC.DASH/EVT.ALERT.UPDATED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            updated: updated,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAlertUpdated}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAlertUpdated = eventGuard("GDC.DASH/EVT.ALERT.UPDATED");
