// @gooddata/sdk-ui-dashboard/esm/model/events/drill.js
import { eventGuard } from "./util";
/**
 * @alpha
 */
export function drillRequested(ctx, drillEvent, drillContext, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillContext: drillContext,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillRequested = eventGuard("GDC.DASH/EVT.DRILL.REQUESTED");
/**
 * @alpha
 */
export function drillResolved(ctx, drillEvent, drillContext, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillContext: drillContext,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillResolved = eventGuard("GDC.DASH/EVT.DRILL.RESOLVED");
/**
 * @alpha
 */
export function drillDownRequested(ctx, insight, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_DOWN.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            insight: insight,
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillDownRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillDownRequested = eventGuard("GDC.DASH/EVT.DRILL.DRILL_DOWN.REQUESTED");
/**
 * @alpha
 */
export function drillDownResolved(ctx, insight, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_DOWN.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            insight: insight,
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillDownResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillDownResolved = eventGuard("GDC.DASH/EVT.DRILL.DRILL_DOWN.RESOLVED");
/**
 * @alpha
 */
export function drillToInsightRequested(ctx, insight, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_INSIGHT.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            insight: insight,
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToInsightRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToInsightRequested = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_INSIGHT.REQUESTED");
/**
 * @alpha
 */
export function drillToInsightResolved(ctx, insight, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_INSIGHT.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            insight: insight,
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToInsightResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToInsightResolved = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_INSIGHT.RESOLVED");
/**
 * @alpha
 */
export function drillToDashboardRequested(ctx, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_DASHBOARD.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillDefinition: drillDefinition,
            drillEvent: drillEvent,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToDashboardRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToDashboardRequested = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_DASHBOARD.REQUESTED");
/**
 * @alpha
 */
export function drillToDashboardResolved(ctx, filters, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_DASHBOARD.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            filters: filters,
            drillDefinition: drillDefinition,
            drillEvent: drillEvent,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToDashboardResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToDashboardResolved = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_DASHBOARD.RESOLVED");
/**
 * @alpha
 */
export function drillToCustomUrlRequested(ctx, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_CUSTOM_URL.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToCustomUrlRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToCustomUrlRequested = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_CUSTOM_URL.REQUESTED");
/**
 * @alpha
 */
export function drillToCustomUrlResolved(ctx, url, drillDefinition, drillEvent, filtersInfo, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_CUSTOM_URL.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            url: url,
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
            filtersInfo: filtersInfo,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToCustomUrlResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToCustomUrlResolved = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_CUSTOM_URL.RESOLVED");
/**
 * @alpha
 */
export function drillToAttributeUrlRequested(ctx, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_ATTRIBUTE_URL.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToAttributeUrlRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToAttributeUrlRequested = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_ATTRIBUTE_URL.REQUESTED");
/**
 * @alpha
 */
export function drillToAttributeUrlResolved(ctx, url, drillDefinition, drillEvent, filtersInfo, isImplicit, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_ATTRIBUTE_URL.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
            url: url,
            filtersInfo: filtersInfo,
            isImplicit: isImplicit,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToAttributeUrlResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToAttributeUrlResolved = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_ATTRIBUTE_URL.RESOLVED");
/**
 * @alpha
 */
export function drillToLegacyDashboardRequested(ctx, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_LEGACY_DASHBOARD.REQUESTED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToLegacyDashboardRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToLegacyDashboardRequested = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_LEGACY_DASHBOARD.REQUESTED");
/**
 * @alpha
 */
export function drillToLegacyDashboardResolved(ctx, drillDefinition, drillEvent, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILL_TO_LEGACY_DASHBOARD.RESOLVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillEvent: drillEvent,
            drillDefinition: drillDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillToLegacyDashboardResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillToLegacyDashboardResolved = eventGuard("GDC.DASH/EVT.DRILL.DRILL_TO_LEGACY_DASHBOARD.RESOLVED");
/**
 * @alpha
 */
export function drillableItemsChanged(ctx, drillableItems, correlationId) {
    return {
        type: "GDC.DASH/EVT.DRILL.DRILLABLE_ITEMS.CHANGED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            drillableItems: drillableItems,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDrillableItemsChanged}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardDrillableItemsChanged = eventGuard("GDC.DASH/EVT.DRILL.DRILLABLE_ITEMS.CHANGED");
