// @gooddata/sdk-ui-dashboard/esm/model/events/filters.js
import { eventGuard } from "./util";
export function dateFilterChanged(ctx, filter, dateFilterOptionLocalId, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.SELECTION_CHANGED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            filter: filter,
            dateFilterOptionLocalId: dateFilterOptionLocalId,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardDateFilterSelectionChanged}.
 *
 * @param obj - object to test
 * @public
 */
export var isDashboardDateFilterSelectionChanged = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.DATE_FILTER.SELECTION_CHANGED");
export function attributeFilterAdded(ctx, added, index, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.ADDED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            added: added,
            index: index,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAttributeFilterAdded}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAttributeFilterAdded = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.ADDED");
export function attributeFilterRemoved(ctx, removed, children, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.REMOVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            removed: removed,
            children: children,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAttributeFilterRemoved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAttributeFilterRemoved = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.REMOVED");
export function attributeFilterMoved(ctx, moved, fromIndex, toIndex, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.MOVED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            moved: moved,
            fromIndex: fromIndex,
            toIndex: toIndex,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAttributeFilterMoved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAttributeFilterMoved = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.MOVED");
export function attributeFilterSelectionChanged(ctx, filter, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.SELECTION_CHANGED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            filter: filter,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAttributeFilterSelectionChanged}.
 *
 * @param obj - object to test
 * @public
 */
export var isDashboardAttributeFilterSelectionChanged = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.SELECTION_CHANGED");
export function attributeFilterParentChanged(ctx, filter, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.PARENT_CHANGED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            filter: filter,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAttributeFilterParentChanged}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAttributeFilterParentChanged = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.ATTRIBUTE_FILTER.PARENT_CHANGED");
export function filterContextChanged(ctx, filterContext, correlationId) {
    return {
        type: "GDC.DASH/EVT.FILTER_CONTEXT.CHANGED",
        ctx: ctx,
        correlationId: correlationId,
        payload: {
            filterContext: filterContext,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardFilterContextChanged}.
 *
 * @param obj - object to test
 * @public
 */
export var isDashboardFilterContextChanged = eventGuard("GDC.DASH/EVT.FILTER_CONTEXT.CHANGED");
