// @gooddata/sdk-ui-dashboard/esm/model/events/render.js
import { eventGuard } from "./util";
/**
 * @alpha
 */
export function renderRequested(ctx, correlationId) {
    return {
        type: "GDC.DASH/EVT.RENDER.REQUESTED",
        correlationId: correlationId,
        ctx: ctx,
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardRenderRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardRenderRequested = eventGuard("GDC.DASH/EVT.RENDER.REQUESTED");
/**
 * @alpha
 */
export function asyncRenderRequested(id, ctx, correlationId) {
    return {
        type: "GDC.DASH/EVT.RENDER.ASYNC.REQUESTED",
        correlationId: correlationId,
        ctx: ctx,
        payload: {
            id: id,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAsyncRenderRequested}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAsyncRenderRequested = eventGuard("GDC.DASH/EVT.RENDER.ASYNC.REQUESTED");
/**
 * @alpha
 */
export function asyncRenderResolved(id, ctx, correlationId) {
    return {
        type: "GDC.DASH/EVT.RENDER.ASYNC.RESOLVED",
        correlationId: correlationId,
        ctx: ctx,
        payload: {
            id: id,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardAsyncRenderResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardAsyncRenderResolved = eventGuard("GDC.DASH/EVT.RENDER.ASYNC.RESOLVED");
/**
 * This event is emitted as soon as the dashboard component is fully rendered,
 * and asynchronous rendering of each component is complete.
 *
 * @alpha
 */
export function renderResolved(ctx, correlationId) {
    return {
        type: "GDC.DASH/EVT.RENDER.RESOLVED",
        correlationId: correlationId,
        ctx: ctx,
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardRenderResolved}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardRenderResolved = eventGuard("GDC.DASH/EVT.RENDER.RESOLVED");
