// @gooddata/sdk-ui-dashboard/esm/model/events/widget.js
// (C) 2021-2022 GoodData Corporation
import { eventGuard } from "./util";
/**
 * @alpha
 */
export function widgetExecutionStarted(widgetRef, executionDefinition, correlationId) {
    return {
        type: "GDC.DASH/EVT.WIDGET.EXECUTION_STARTED",
        correlationId: correlationId,
        payload: {
            widgetRef: widgetRef,
            executionDefinition: executionDefinition,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardWidgetExecutionStarted}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardWidgetExecutionStarted = eventGuard("GDC.DASH/EVT.WIDGET.EXECUTION_STARTED");
/**
 * @alpha
 */
export function widgetExecutionFailed(widgetRef, error, correlationId) {
    return {
        type: "GDC.DASH/EVT.WIDGET.EXECUTION_FAILED",
        correlationId: correlationId,
        payload: {
            widgetRef: widgetRef,
            error: error,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardWidgetExecutionFailed}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardWidgetExecutionFailed = eventGuard("GDC.DASH/EVT.WIDGET.EXECUTION_FAILED");
/**
 * @alpha
 */
export function widgetExecutionSucceeded(widgetRef, dataView, correlationId) {
    return {
        type: "GDC.DASH/EVT.WIDGET.EXECUTION_SUCCEEDED",
        correlationId: correlationId,
        payload: {
            dataView: dataView,
            widgetRef: widgetRef,
        },
    };
}
/**
 * Tests whether the provided object is an instance of {@link DashboardWidgetExecutionSucceeded}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isDashboardWidgetExecutionSucceeded = eventGuard("GDC.DASH/EVT.WIDGET.EXECUTION_SUCCEEDED");
