// @gooddata/api-model-bear/esm/visualizationWidget/GdcVisualizationWidget.js
// (C) 2020-2021 GoodData Corporation
import isEmpty from "lodash/fp/isEmpty";
import has from "lodash/has";
/**
 * @public
 */
export var GdcVisualizationWidget;
(function (GdcVisualizationWidget) {
    function isDrillToVisualization(obj) {
        return !isEmpty(obj) && !!obj.drillToVisualization;
    }
    GdcVisualizationWidget.isDrillToVisualization = isDrillToVisualization;
    function isDrillToDashboard(obj) {
        return !isEmpty(obj) && !!obj.drillToDashboard;
    }
    GdcVisualizationWidget.isDrillToDashboard = isDrillToDashboard;
    function isDrillToCustomUrl(obj) {
        return !isEmpty(obj) && !!obj.drillToCustomUrl;
    }
    GdcVisualizationWidget.isDrillToCustomUrl = isDrillToCustomUrl;
    function isDrillToAttributeUrl(obj) {
        return !isEmpty(obj) && !!obj.drillToAttributeUrl;
    }
    GdcVisualizationWidget.isDrillToAttributeUrl = isDrillToAttributeUrl;
    function isDrillFromMeasure(obj) {
        return !isEmpty(obj) && !!obj.drillFromMeasure;
    }
    GdcVisualizationWidget.isDrillFromMeasure = isDrillFromMeasure;
    function isDrillFromAttribute(obj) {
        return !isEmpty(obj) && !!obj.drillFromAttribute;
    }
    GdcVisualizationWidget.isDrillFromAttribute = isDrillFromAttribute;
    function isVisualizationWidget(obj) {
        return !isEmpty(obj) && obj.meta.category === "visualizationWidget";
    }
    GdcVisualizationWidget.isVisualizationWidget = isVisualizationWidget;
    function isWrappedVisualizationWidget(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "visualizationWidget");
    }
    GdcVisualizationWidget.isWrappedVisualizationWidget = isWrappedVisualizationWidget;
})(GdcVisualizationWidget || (GdcVisualizationWidget = {}));
