// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/bulletChart/bulletConfiguration.js
import { __assign } from "tslib";
// (C) 2007-2021 GoodData Corporation
import cloneDeep from "lodash/cloneDeep";
import { MAX_POINT_WIDTH } from "../_chartCreators/commonConfiguration";
import { getCommonResponsiveConfig } from "../_chartCreators/responsive";
var BULLET_TEMPLATE = {
    chart: {
        type: "bar",
    },
    plotOptions: {
        bar: {
            maxPointWidth: MAX_POINT_WIDTH,
            dataLabels: {
                enabled: false,
            },
        },
        bullet: {
            maxPointWidth: MAX_POINT_WIDTH,
            tooltip: {
                followPointer: true,
            },
            targetOptions: {
                borderWidth: 0,
            },
        },
        series: {
            states: {
                hover: {
                    enabled: false,
                },
            },
            grouping: false,
            borderWidth: 0,
        },
    },
    yAxis: [
        {
            stackLabels: {
                enabled: false,
            },
        },
    ],
};
export function getBulletConfiguration(config) {
    var bulletConfiguration = cloneDeep(BULLET_TEMPLATE);
    if ((config === null || config === void 0 ? void 0 : config.enableCompactSize) && !(config === null || config === void 0 ? void 0 : config.zoomInsight)) {
        var reversed = true;
        return __assign(__assign({}, bulletConfiguration), { responsive: getCommonResponsiveConfig(reversed) });
    }
    return bulletConfiguration;
}
