// @gooddata/sdk-backend-tiger/esm/convertors/toBackend/AnalyticalDashboardConverter.js
import { __spreadArrays } from "tslib";
import { isDrillToCustomUrl, isInsightWidget, isInsightWidgetDefinition, walkLayout, } from "@gooddata/sdk-backend-spi";
import omit from "lodash/omit";
import updateWith from "lodash/updateWith";
import { cloneWithSanitizedIds } from "./IdSanitization";
import isEmpty from "lodash/isEmpty";
import update from "lodash/fp/update";
import { splitDrillUrlParts } from "@gooddata/sdk-backend-spi/dist/workspace/dashboards/drillUrls";
function removeIdentifiers(widget) {
    return omit(widget, ["ref", "uri", "identifier"]);
}
function removeWidgetIdentifiersInLayout(layout) {
    if (!layout) {
        return;
    }
    var widgetsPaths = [];
    walkLayout(layout, {
        widgetCallback: function (_, widgetPath) { return widgetsPaths.push(widgetPath); },
    });
    return widgetsPaths.reduce(function (layout, widgetPath) {
        return updateWith(layout, widgetPath, removeIdentifiers);
    }, layout);
}
export function convertAnalyticalDashboard(dashboard, filterContextRef) {
    var _a;
    var layout = convertDrillToCustomUrlInLayoutToBackend(removeWidgetIdentifiersInLayout(dashboard.layout));
    return {
        dateFilterConfig: cloneWithSanitizedIds(dashboard.dateFilterConfig),
        filterContextRef: cloneWithSanitizedIds(filterContextRef),
        layout: cloneWithSanitizedIds(layout),
        plugins: (_a = dashboard.plugins) === null || _a === void 0 ? void 0 : _a.map(convertDashboardPluginLinkToBackend),
        version: "2",
    };
}
export function convertFilterContextToBackend(filterContext) {
    return {
        filters: cloneWithSanitizedIds(filterContext.filters),
        version: "2",
    };
}
export function convertDashboardPluginToBackend(plugin) {
    return {
        url: plugin.url,
        version: "2",
    };
}
export function convertDashboardPluginLinkToBackend(pluginLink) {
    return {
        plugin: cloneWithSanitizedIds(pluginLink.plugin),
        parameters: pluginLink.parameters,
        version: "2",
    };
}
export function getDrillToCustomUrlPaths(layout) {
    var paths = [];
    walkLayout(layout, {
        widgetCallback: function (widget, widgetPath) {
            if (!isInsightWidget(widget) && !isInsightWidgetDefinition(widget)) {
                return;
            }
            widget.drills.forEach(function (drill, drillIndex) {
                if (!isDrillToCustomUrl(drill)) {
                    return;
                }
                paths.push(__spreadArrays(widgetPath, ["drills", drillIndex]));
            });
        },
    });
    return paths;
}
function convertTargetUrlToParts(drill) {
    return update(["target", "url"], splitDrillUrlParts, drill);
}
export function convertDrillToCustomUrlInLayoutToBackend(layout) {
    if (!layout) {
        return;
    }
    var paths = getDrillToCustomUrlPaths(layout);
    if (isEmpty(paths)) {
        return layout;
    }
    return paths.reduce(function (layout, path) {
        return update(path, convertTargetUrlToParts, layout);
    }, layout);
}
