// @gooddata/sdk-ui-dashboard/esm/presentation/drill/hooks/useDrills.js
import { useDrill } from "../hooks/useDrill";
import { useDrillDown } from "../hooks/useDrillDown";
import { useDrillToInsight } from "../hooks/useDrillToInsight";
import { useDrillToDashboard } from "../hooks/useDrillToDashboard";
import { useDrillToAttributeUrl } from "../hooks/useDrillToAttributeUrl";
import { useDrillToCustomUrl } from "../hooks/useDrillToCustomUrl";
import { useDrillToLegacyDashboard } from "./useDrillToLegacyDashboard";
/**
 * @internal
 */
export function useDrills(_a) {
    var onDrill = _a.onDrill, onDrillSuccess = _a.onDrillSuccess, onDrillError = _a.onDrillError, 
    //
    onDrillDown = _a.onDrillDown, onDrillDownSuccess = _a.onDrillDownSuccess, onDrillDownError = _a.onDrillDownError, 
    //
    onDrillToInsight = _a.onDrillToInsight, onDrillToInsightSuccess = _a.onDrillToInsightSuccess, onDrillToInsightError = _a.onDrillToInsightError, 
    //
    onDrillToDashboard = _a.onDrillToDashboard, onDrillToDashboardSuccess = _a.onDrillToDashboardSuccess, onDrillToDashboardError = _a.onDrillToDashboardError, 
    //
    onDrillToAttributeUrl = _a.onDrillToAttributeUrl, onDrillToAttributeUrlSuccess = _a.onDrillToAttributeUrlSuccess, onDrillToAttributeUrlError = _a.onDrillToAttributeUrlError, 
    //
    onDrillToCustomUrl = _a.onDrillToCustomUrl, onDrillToCustomUrlSuccess = _a.onDrillToCustomUrlSuccess, onDrillToCustomUrlError = _a.onDrillToCustomUrlError, 
    //
    onDrillToLegacyDashboard = _a.onDrillToLegacyDashboard, onDrillToLegacyDashboardSuccess = _a.onDrillToLegacyDashboardSuccess, onDrillToLegacyDashboardError = _a.onDrillToLegacyDashboardError, 
    //
    onError = _a.onError;
    var drill = useDrill({
        onBeforeRun: onDrill,
        onSuccess: onDrillSuccess,
        onError: onDrillError !== null && onDrillError !== void 0 ? onDrillError : onError,
    });
    var drillDown = useDrillDown({
        onBeforeRun: onDrillDown,
        onSuccess: onDrillDownSuccess,
        onError: onDrillDownError !== null && onDrillDownError !== void 0 ? onDrillDownError : onError,
    });
    var drillToInsight = useDrillToInsight({
        onBeforeRun: onDrillToInsight,
        onSuccess: onDrillToInsightSuccess,
        onError: onDrillToInsightError !== null && onDrillToInsightError !== void 0 ? onDrillToInsightError : onError,
    });
    var drillToDashboard = useDrillToDashboard({
        onBeforeRun: onDrillToDashboard,
        onSuccess: onDrillToDashboardSuccess,
        onError: onDrillToDashboardError !== null && onDrillToDashboardError !== void 0 ? onDrillToDashboardError : onError,
    });
    var drillToAttributeUrl = useDrillToAttributeUrl({
        onBeforeRun: onDrillToAttributeUrl,
        onSuccess: onDrillToAttributeUrlSuccess,
        onError: onDrillToAttributeUrlError !== null && onDrillToAttributeUrlError !== void 0 ? onDrillToAttributeUrlError : onError,
    });
    var drillToCustomUrl = useDrillToCustomUrl({
        onBeforeRun: onDrillToCustomUrl,
        onSuccess: onDrillToCustomUrlSuccess,
        onError: onDrillToCustomUrlError !== null && onDrillToCustomUrlError !== void 0 ? onDrillToCustomUrlError : onError,
    });
    var drillToLegacyDashboard = useDrillToLegacyDashboard({
        onBeforeRun: onDrillToLegacyDashboard,
        onSuccess: onDrillToLegacyDashboardSuccess,
        onError: onDrillToLegacyDashboardError !== null && onDrillToLegacyDashboardError !== void 0 ? onDrillToLegacyDashboardError : onError,
    });
    return {
        drill: drill,
        drillDown: drillDown,
        drillToInsight: drillToInsight,
        drillToDashboard: drillToDashboard,
        drillToAttributeUrl: drillToAttributeUrl,
        drillToCustomUrl: drillToCustomUrl,
        drillToLegacyDashboard: drillToLegacyDashboard,
    };
}
