// @gooddata/sdk-backend-tiger/esm/backend/workspaces/index.js
import { __awaiter, __extends, __generator } from "tslib";
import { OrganizationUtilities } from "@gooddata/api-client-tiger";
import { workspaceConverter } from "../../convertors/fromBackend/WorkspaceConverter";
import { InMemoryPaging } from "@gooddata/sdk-backend-base";
import { TigerWorkspace } from "../workspace";
var TigerWorkspaceQueryFactory = /** @class */ (function () {
    function TigerWorkspaceQueryFactory(authCall, dateFormatter) {
        this.authCall = authCall;
        this.dateFormatter = dateFormatter;
    }
    TigerWorkspaceQueryFactory.prototype.forUser = function (userId) {
        return new TigerWorkspaceQuery(this.authCall, this.dateFormatter, userId);
    };
    TigerWorkspaceQueryFactory.prototype.forCurrentUser = function () {
        return new TigerWorkspaceQuery(this.authCall, this.dateFormatter);
    };
    return TigerWorkspaceQueryFactory;
}());
export { TigerWorkspaceQueryFactory };
var TigerWorkspaceQuery = /** @class */ (function () {
    function TigerWorkspaceQuery(authCall, dateFormatter, 
    // @ts-expect-error Keeping this for now for future use
    userId) {
        var _this = this;
        this.authCall = authCall;
        this.dateFormatter = dateFormatter;
        this.userId = userId;
        this.limit = 100;
        this.offset = 0;
        this.search = undefined;
        this.resultToWorkspaceDescriptors = function (result) {
            return result.data.map(workspaceConverter);
        };
        this.searchWorkspaceDescriptors = function (search) {
            return function (results) {
                if (search) {
                    var lowercaseSearch_1 = search.toLocaleLowerCase();
                    return results.filter(function (workspace) {
                        var title = workspace.title;
                        return title && title.toLowerCase().indexOf(lowercaseSearch_1) > -1;
                    });
                }
                return results;
            };
        };
        this.descriptorToAnalyticalWorkspace = function (descriptor) {
            return new TigerWorkspace(_this.authCall, descriptor.id, _this.dateFormatter, descriptor);
        };
        this.descriptorsToAnalyticalWorkspaces = function (descriptors) { return descriptors.map(_this.descriptorToAnalyticalWorkspace); };
    }
    TigerWorkspaceQuery.prototype.withLimit = function (limit) {
        this.limit = limit;
        return this;
    };
    TigerWorkspaceQuery.prototype.withOffset = function (offset) {
        this.offset = offset;
        return this;
    };
    TigerWorkspaceQuery.prototype.withSearch = function (search) {
        this.search = search;
        return this;
    };
    TigerWorkspaceQuery.prototype.query = function () {
        return this.queryWorker(this.offset, this.limit, this.search);
    };
    TigerWorkspaceQuery.prototype.queryWorker = function (offset, limit, search) {
        return __awaiter(this, void 0, void 0, function () {
            var allWorkspaces;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.authCall(function (client) {
                            return OrganizationUtilities.getAllPagesOf(client, client.organizationObjects.getAllEntitiesWorkspaces, { sort: ["name"], include: ["workspaces"] })
                                .then(OrganizationUtilities.mergeEntitiesResults)
                                .then(_this.resultToWorkspaceDescriptors)
                                .then(_this.searchWorkspaceDescriptors(search))
                                .then(_this.descriptorsToAnalyticalWorkspaces);
                        })];
                    case 1:
                        allWorkspaces = _a.sent();
                        return [2 /*return*/, new WorkspacesInMemoryPaging(allWorkspaces, limit !== null && limit !== void 0 ? limit : 50, offset !== null && offset !== void 0 ? offset : 0, search)];
                }
            });
        });
    };
    return TigerWorkspaceQuery;
}());
var WorkspacesInMemoryPaging = /** @class */ (function (_super) {
    __extends(WorkspacesInMemoryPaging, _super);
    function WorkspacesInMemoryPaging(allItems, limit, offset, search) {
        if (limit === void 0) { limit = 50; }
        if (offset === void 0) { offset = 0; }
        if (search === void 0) { search = undefined; }
        var _this = _super.call(this, allItems, limit, offset) || this;
        _this.search = search;
        return _this;
    }
    WorkspacesInMemoryPaging.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (this.items.length === 0) {
                    return [2 /*return*/, this];
                }
                return [2 /*return*/, new WorkspacesInMemoryPaging(this.allItems, this.limit, this.offset + this.items.length, this.search)];
            });
        });
    };
    return WorkspacesInMemoryPaging;
}(InMemoryPaging));
