// @gooddata/sdk-backend-tiger/esm/backend/workspace/attributes/elements/date.js
import addYears from "date-fns/addYears";
import addMonths from "date-fns/addMonths";
import addDays from "date-fns/addDays";
import setDate from "date-fns/setDate";
import setMonth from "date-fns/setMonth";
import getMonth from "date-fns/getMonth";
import format from "date-fns/format";
var DATE_GRANULARITY;
(function (DATE_GRANULARITY) {
    DATE_GRANULARITY["DATE"] = "GDC.time.date";
    DATE_GRANULARITY["WEEK"] = "GDC.time.week_us";
    DATE_GRANULARITY["MONTH"] = "GDC.time.month";
    DATE_GRANULARITY["QUARTER"] = "GDC.time.quarter";
    DATE_GRANULARITY["YEAR"] = "GDC.time.year";
})(DATE_GRANULARITY || (DATE_GRANULARITY = {}));
var iterateDates = function (start, end) {
    var wholeDates = [];
    while (start <= end) {
        wholeDates.push(format(start, "MM/dd/yyyy"));
        var newDate = addDays(start, 1);
        start = newDate;
    }
    return wholeDates;
};
export var getRelativeDateFilterShiftedValues = function (currentDate, relativeDateFilter) {
    var _a = relativeDateFilter.relativeDateFilter, from = _a.from, to = _a.to, granularity = _a.granularity;
    var startDate = currentDate;
    switch (granularity) {
        case DATE_GRANULARITY.YEAR: {
            startDate = addYears(startDate, from);
            startDate = setMonth(startDate, 0);
            startDate = setDate(startDate, 1);
            var endDate = currentDate;
            endDate = addYears(endDate, to);
            endDate = setMonth(endDate, 12);
            endDate = setDate(endDate, 0);
            return iterateDates(startDate, endDate);
        }
        case DATE_GRANULARITY.MONTH: {
            startDate = currentDate;
            startDate = addMonths(startDate, from);
            startDate = setDate(startDate, 1);
            var endDate = currentDate;
            endDate = addMonths(endDate, to + 1);
            endDate = setDate(endDate, 0);
            return iterateDates(startDate, endDate);
        }
        case DATE_GRANULARITY.DATE: {
            startDate = addDays(startDate, from);
            var endDate = currentDate;
            endDate = addDays(endDate, to);
            return iterateDates(startDate, endDate);
        }
        case DATE_GRANULARITY.QUARTER: {
            startDate = getFirstDayQuarter(currentDate, from);
            var endDate = getLastDayQuarter(currentDate, to);
            return iterateDates(startDate, endDate);
        }
        default: {
            return [];
        }
    }
};
var getFirstDayQuarter = function (currentDate, from) {
    var _a = getQuarterDateAndMonth(currentDate, from), date = _a.date, month = _a.month;
    var result;
    if (month <= 3) {
        result = setDate(setMonth(date, 0), 1);
    }
    else if (month <= 6) {
        result = setDate(setMonth(date, 3), 1);
    }
    else if (month <= 9) {
        result = setDate(setMonth(date, 6), 1);
    }
    else {
        result = setDate(setMonth(date, 9), 1);
    }
    return result;
};
var getLastDayQuarter = function (currentDate, to) {
    var _a = getQuarterDateAndMonth(currentDate, to), date = _a.date, month = _a.month;
    var result;
    if (month <= 3) {
        result = setDate(setMonth(date, 3), 0);
    }
    else if (month <= 6) {
        result = setDate(setMonth(date, 6), 0);
    }
    else if (month <= 9) {
        result = setDate(setMonth(date, 9), 0);
    }
    else {
        result = setDate(setMonth(date, 12), 0);
    }
    return result;
};
var getQuarterDateAndMonth = function (currentDate, value) {
    var date = addMonths(currentDate, value * 3);
    return {
        date: date,
        month: getMonth(date),
    };
};
