// @gooddata/sdk-backend-tiger/esm/backend/workspace/catalog/catalogWithAvailableItems.js
// (C) 2019-2020 GoodData Corporation
import { isCatalogAttribute, isCatalogFact, isCatalogMeasure, isCatalogDateDataset, } from "@gooddata/sdk-backend-spi";
var TigerWorkspaceCatalogWithAvailableItems = /** @class */ (function () {
    function TigerWorkspaceCatalogWithAvailableItems(catalogGroups, items, availableItems) {
        this.catalogGroups = catalogGroups;
        this.items = items;
        this.availableItems = availableItems;
    }
    TigerWorkspaceCatalogWithAvailableItems.prototype.groups = function () {
        return this.catalogGroups;
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.allItems = function () {
        return this.items;
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.attributes = function () {
        return this.items.filter(isCatalogAttribute);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.measures = function () {
        return this.items.filter(isCatalogMeasure);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.facts = function () {
        return this.items.filter(isCatalogFact);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.dateDatasets = function () {
        return this.items.filter(isCatalogDateDataset);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.allAvailableItems = function () {
        return this.availableItems;
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.availableAttributes = function () {
        return this.availableItems.filter(isCatalogAttribute);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.availableMeasures = function () {
        return this.availableItems.filter(isCatalogMeasure);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.availableFacts = function () {
        return this.availableItems.filter(isCatalogFact);
    };
    TigerWorkspaceCatalogWithAvailableItems.prototype.availableDateDatasets = function () {
        return this.availableItems.filter(isCatalogDateDataset);
    };
    return TigerWorkspaceCatalogWithAvailableItems;
}());
export { TigerWorkspaceCatalogWithAvailableItems };
