// @gooddata/sdk-backend-tiger/esm/backend/workspace/catalog/catalog.js
// (C) 2019-2020 GoodData Corporation
import { isCatalogAttribute, isCatalogFact, isCatalogMeasure, isCatalogDateDataset, } from "@gooddata/sdk-backend-spi";
import { TigerWorkspaceCatalogAvailableItemsFactory } from "./availableItemsFactory";
var TigerWorkspaceCatalog = /** @class */ (function () {
    function TigerWorkspaceCatalog(authCall, workspace, catalogGroups, items, options) {
        this.authCall = authCall;
        this.workspace = workspace;
        this.catalogGroups = catalogGroups;
        this.items = items;
        this.options = options;
    }
    TigerWorkspaceCatalog.prototype.groups = function () {
        return this.catalogGroups;
    };
    TigerWorkspaceCatalog.prototype.allItems = function () {
        return this.items;
    };
    TigerWorkspaceCatalog.prototype.attributes = function () {
        return this.items.filter(isCatalogAttribute);
    };
    TigerWorkspaceCatalog.prototype.measures = function () {
        return this.items.filter(isCatalogMeasure);
    };
    TigerWorkspaceCatalog.prototype.facts = function () {
        return this.items.filter(isCatalogFact);
    };
    TigerWorkspaceCatalog.prototype.dateDatasets = function () {
        return this.items.filter(isCatalogDateDataset);
    };
    TigerWorkspaceCatalog.prototype.availableItems = function () {
        return new TigerWorkspaceCatalogAvailableItemsFactory(this.authCall, this.workspace, this.catalogGroups, this.items, this.options);
    };
    return TigerWorkspaceCatalog;
}());
export { TigerWorkspaceCatalog };
