// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/barChart/barConfiguration.js
import { __assign } from "tslib";
// (C) 2007-2021 GoodData Corporation
import cloneDeep from "lodash/cloneDeep";
import { MAX_POINT_WIDTH } from "../_chartCreators/commonConfiguration";
import { getCommonResponsiveConfig } from "../_chartCreators/responsive";
import { getAxesCounts } from "../_util/common";
var BAR_TEMPLATE = {
    chart: {
        type: "bar",
    },
    plotOptions: {
        bar: {
            maxPointWidth: MAX_POINT_WIDTH,
            dataLabels: {
                enabled: true,
                padding: 2,
            },
        },
        series: {
            states: {
                hover: {
                    enabled: false,
                },
            },
        },
    },
    yAxis: [
        {
            stackLabels: {
                enabled: false,
            },
        },
    ],
};
export function getBarConfiguration(config) {
    var barConfiguration = cloneDeep(BAR_TEMPLATE);
    if ((config === null || config === void 0 ? void 0 : config.enableCompactSize) && !(config === null || config === void 0 ? void 0 : config.zoomInsight)) {
        var reversed = true;
        var _a = getAxesCounts(config), xAxesCount = _a[0], yAxesCount = _a[1];
        return __assign(__assign({}, barConfiguration), { responsive: getCommonResponsiveConfig(reversed, xAxesCount, yAxesCount) });
    }
    return barConfiguration;
}
