// @gooddata/sdk-ui-dashboard/esm/model/store/config/configSelectors.js
// (C) 2021 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import invariant from "ts-invariant";
var selectSelf = createSelector(function (state) { return state; }, function (state) { return state.config; });
/**
 * Returns dashboard's config. It is expected that the selector is called only after the config state
 * is correctly initialized. Invocations before initialization lead to invariant errors.
 *
 * @public
 */
export var selectConfig = createSelector(selectSelf, function (configState) {
    invariant(configState.config, "attempting to access uninitialized config state");
    return configState.config;
});
/**
 * Returns workspace-level configuration for the of the date filter options and presets.
 *
 * Note: this configuration SHOULD be further augmented by the dashboard-level overrides to obtain
 * the effective date filter configuration.
 *
 * @public
 */
export var selectDateFilterConfig = createSelector(selectConfig, function (state) {
    return state.dateFilterConfig;
});
/**
 * Returns settings that are in effect for the current dashboard.
 *
 * @public
 */
export var selectSettings = createSelector(selectConfig, function (state) {
    return state.settings;
});
/**
 * Returns locale to use for internationalization of the dashboard.
 *
 * @public
 */
export var selectLocale = createSelector(selectConfig, function (state) {
    return state.locale;
});
/**
 * Returns number separators to use when rendering numeric values on charts or KPIs.
 *
 * @public
 */
export var selectSeparators = createSelector(selectConfig, function (state) {
    return state.separators;
});
/**
 * Returns the color palette for dashboard charts.
 *
 * @public
 */
export var selectColorPalette = createSelector(selectConfig, function (state) {
    return state.colorPalette;
});
/**
 * Returns the object availability configuration for this dashboard. Only objects that match the availability
 * criteria can appear in selections where user has pick an object to use for some purpose (for instance metric for
 * KPI or date dataset to filter by).
 *
 * @public
 */
export var selectObjectAvailabilityConfig = createSelector(selectConfig, function (state) {
    return state.objectAvailability;
});
/**
 * Returns Mapbox token.
 *
 * @internal
 */
export var selectMapboxToken = createSelector(selectConfig, function (state) {
    return state.mapboxToken;
});
/**
 * Returns whether the Dashboard is executed in read-only mode.
 * Read-only mode disables any interactions that can alter the backend data.
 *
 * @public
 */
export var selectIsReadOnly = createSelector(selectConfig, function (state) {
    return state.isReadOnly;
});
/**
 * Returns whether the Dashboard is executed in embedded context.
 * In embedded mode, some interactions may be disabled.
 *
 * @public
 */
export var selectIsEmbedded = createSelector(selectConfig, function (state) {
    return state.isEmbedded;
});
/**
 * Returns whether the Dashboard is rendered in the export mode.
 * In export mode, some components can be hidden, or rendered differently.
 *
 * @internal
 */
export var selectIsExport = createSelector(selectConfig, function (state) {
    return state.isExport;
});
/**
 * Returns whether the default drills configured on the widgets or implicit drills (eg. drill down) are disabled.
 * This option does not affect drilling enabled by drillableItems.
 *
 * @public
 */
export var selectDisableDefaultDrills = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.disableDefaultDrills) !== null && _a !== void 0 ? _a : false;
});
/**
 * Returns whether filter values in drill events should be resolved.
 *
 * @public
 */
export var selectEnableFilterValuesResolutionInDrillEvents = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.enableFilterValuesResolutionInDrillEvents) !== null && _a !== void 0 ? _a : false;
});
//
// FEATURE FLAGS
//
/**
 * Returns date format.
 *
 * @public
 */
export var selectDateFormat = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.settings) === null || _a === void 0 ? void 0 : _a.responsiveUiDateFormat;
});
/**
 * Returns whether the current user can schedule emails.
 *
 * @public
 */
export var selectEnableKPIDashboardSchedule = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardSchedule;
});
/**
 * Returns whether the current user can share scheduled email to other recipients.
 *
 * @public
 */
export var selectEnableKPIDashboardScheduleRecipients = createSelector(selectConfig, function (state) {
    var _a;
    return (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardScheduleRecipients;
});
/**
 * Returns current platform edition.
 *
 * @public
 */
export var selectPlatformEdition = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.platformEdition) !== null && _b !== void 0 ? _b : "enterprise";
});
/**
 * Returns whether company logo should be visible in embedded dashboard.
 *
 * @public
 */
export var selectEnableCompanyLogoInEmbeddedUI = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableCompanyLogoInEmbeddedUI) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the export to pdf is enabled.
 *
 * @public
 */
export var selectEnableKPIDashboardExportPDF = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardExportPDF) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the drill to dashboard is enabled.
 *
 * @public
 */
export var selectEnableKPIDashboardDrillToDashboard = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDrillToDashboard) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether the save as new dashboard functionality is enabled.
 *
 * @public
 */
export var selectEnableKPIDashboardSaveAsNew = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardSaveAsNew) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether implicit drill to attributes url enabled
 *
 * @public
 */
export var selectEnableClickableAttributeURL = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableClickableAttributeURL) !== null && _b !== void 0 ? _b : true;
});
/**
 * Returns whether drill to url is enabled
 *
 * @public
 */
export var selectEnableKPIDashboardDrillToURL = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDrillToURL) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether drill to insight is enabled
 *
 * @public
 */
export var selectEnableKPIDashboardDrillToInsight = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardDrillToInsight) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether implicit drill to attributes url enabled
 *
 * @public
 */
export var selectEnableKPIDashboardImplicitDrillDown = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.enableKPIDashboardImplicitDrillDown) !== null && _b !== void 0 ? _b : false;
});
/**
 * Returns whether Kpi drills in embedded mode are disabled.
 *
 * @public
 */
export var selectHideKpiDrillInEmbedded = createSelector(selectConfig, function (state) {
    var _a, _b;
    return (_b = (_a = state.settings) === null || _a === void 0 ? void 0 : _a.hideKpiDrillInEmbedded) !== null && _b !== void 0 ? _b : false;
});
