// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/columnChart/columnConfiguration.js
import { __assign } from "tslib";
import { getCommonResponsiveConfig } from "../_chartCreators/responsive";
import { MAX_POINT_WIDTH } from "../_chartCreators/commonConfiguration";
import { getAxesCounts } from "../_util/common";
export function getColumnConfiguration(config, _definition, theme) {
    var _a, _b, _c;
    var columnConfiguration = {
        chart: {
            type: "column",
            spacingTop: 20,
        },
        plotOptions: {
            column: {
                dataLabels: {
                    enabled: true,
                    padding: 2,
                },
                maxPointWidth: MAX_POINT_WIDTH,
            },
            series: {
                states: {
                    hover: {
                        enabled: false,
                    },
                },
            },
        },
        yAxis: [
            {
                stackLabels: __assign({ enabled: true, allowOverlap: false }, (((_a = theme === null || theme === void 0 ? void 0 : theme.palette) === null || _a === void 0 ? void 0 : _a.complementary) && {
                    style: {
                        color: (_c = (_b = theme === null || theme === void 0 ? void 0 : theme.palette) === null || _b === void 0 ? void 0 : _b.complementary) === null || _c === void 0 ? void 0 : _c.c9,
                        textOutline: "none",
                    },
                })),
            },
        ],
    };
    if ((config === null || config === void 0 ? void 0 : config.enableCompactSize) && !(config === null || config === void 0 ? void 0 : config.zoomInsight)) {
        var _d = getAxesCounts(config), xAxesCount = _d[0], yAxesCount = _d[1];
        return __assign(__assign({}, columnConfiguration), { responsive: getCommonResponsiveConfig(false, xAxesCount, yAxesCount) });
    }
    return columnConfiguration;
}
