// @gooddata/sdk-ui-kit/esm/Form/Checkbox.js
import { __extends } from "tslib";
// (C) 2019-2022 GoodData Corporation
import React from "react";
import cx from "classnames";
import noop from "lodash/noop";
/**
 * @internal
 */
var Checkbox = /** @class */ (function (_super) {
    __extends(Checkbox, _super);
    function Checkbox() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onChange = function (e) {
            _this.props.onChange(e.target.checked);
        };
        return _this;
    }
    Checkbox.prototype.render = function () {
        var _a = this.props, disabled = _a.disabled, name = _a.name, text = _a.text, title = _a.title, value = _a.value, labelSize = _a.labelSize;
        var labelClasses = cx("input-label-text", {
            "gd-label-small gd-checkbox-label-small": labelSize === "small",
            "gd-label gd-checkbox-label": labelSize === "normal",
        });
        return (React.createElement(React.Fragment, null,
            title && React.createElement("h6", null, title),
            React.createElement("label", { className: "input-checkbox-label" },
                React.createElement("input", { type: "checkbox", className: "input-checkbox", name: name, checked: value, disabled: disabled, onChange: this.onChange }),
                React.createElement("span", { className: labelClasses }, text))));
    };
    Checkbox.defaultProps = {
        disabled: false,
        name: "",
        text: "",
        title: "",
        value: false,
        labelSize: "small",
        onChange: noop,
    };
    return Checkbox;
}(React.PureComponent));
export { Checkbox };
