// @gooddata/sdk-ui-dashboard/esm/presentation/widget/insightMenu/DefaultDashboardInsightMenu/getDefaultInsightMenuItems.js
/**
 * @internal
 */
export function getDefaultInsightMenuItems(intl, config) {
    var exportCSVDisabled = config.exportCSVDisabled, exportXLSXDisabled = config.exportXLSXDisabled, onExportCSV = config.onExportCSV, onExportXLSX = config.onExportXLSX, tooltipMessage = config.tooltipMessage;
    return [
        {
            type: "button",
            itemId: "ExportXLSXBubble",
            itemName: intl.formatMessage({ id: "widget.options.menu.exportToXLSX" }),
            onClick: onExportXLSX,
            disabled: exportXLSXDisabled,
            tooltip: tooltipMessage,
            icon: "gd-icon-download",
            className: "s-options-menu-export-xlsx",
        },
        {
            type: "button",
            itemId: "ExportCSVBubble",
            itemName: intl.formatMessage({ id: "widget.options.menu.exportToCSV" }),
            onClick: onExportCSV,
            disabled: exportCSVDisabled,
            tooltip: tooltipMessage,
            icon: "gd-icon-download",
            className: "s-options-menu-export-csv",
        },
    ];
}
