// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/baseChart/BaseChartDescriptor.js
// (C) 2021 GoodData Corporation
import { DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT, MAX_VISUALIZATION_HEIGHT, MIN_VISUALIZATION_HEIGHT, MIDDLE_VISUALIZATION_HEIGHT, } from "../constants";
import { addIntersectionFiltersToInsight, modifyBucketsAttributesForDrillDown } from "../drillDownUtil";
var BaseChartDescriptor = /** @class */ (function () {
    function BaseChartDescriptor() {
    }
    BaseChartDescriptor.prototype.getSizeInfo = function (_insight, layoutDescriptor, settings) {
        return {
            width: {
                default: 6,
                min: 4,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getMinHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    };
    BaseChartDescriptor.prototype.getDefaultHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MIDDLE_VISUALIZATION_HEIGHT;
    };
    BaseChartDescriptor.prototype.getMinHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MIN_VISUALIZATION_HEIGHT;
    };
    BaseChartDescriptor.prototype.getMaxHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MAX_VISUALIZATION_HEIGHT;
    };
    BaseChartDescriptor.prototype.applyDrillDown = function (insight, drillDownContext) {
        var intersection = drillDownContext.event.drillContext.intersection;
        var withFilters = addIntersectionFiltersToInsight(insight, intersection);
        return modifyBucketsAttributesForDrillDown(withFilters, drillDownContext.drillDefinition);
    };
    return BaseChartDescriptor;
}());
export { BaseChartDescriptor };
