// @gooddata/sdk-ui/esm/base/localization/intlUtils.js
import { __assign, __extends } from "tslib";
// (C) 2007-2018 GoodData Corporation
import React from "react";
import { IntlProvider, createIntl } from "react-intl";
import { messagesMap } from "./IntlWrapper";
import { DefaultLocale } from "./Locale";
import { wrapDisplayName } from "../react/wrapDisplayName";
/**
 * @internal
 */
export function createIntlMock(customMessages, locale) {
    if (customMessages === void 0) { customMessages = {}; }
    if (locale === void 0) { locale = "en-US"; }
    return createIntl({
        locale: locale,
        messages: __assign(__assign({}, messagesMap[locale]), customMessages),
    });
}
/**
 * @internal
 */
export function withIntl(WrappedComponent, customLocale, customMessages) {
    var WithIntl = /** @class */ (function (_super) {
        __extends(WithIntl, _super);
        function WithIntl() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WithIntl.prototype.render = function () {
            var locale = customLocale ? customLocale : DefaultLocale;
            var messages = customMessages ? customMessages : messagesMap[locale];
            return (React.createElement(IntlProvider, { locale: locale, messages: messages },
                React.createElement(WrappedComponent, __assign({}, this.props))));
        };
        return WithIntl;
    }(React.Component));
    return wrapDisplayName("withIntl", WrappedComponent)(WithIntl);
}
