// @gooddata/sdk-ui/esm/base/localization/TranslationsProvider.js
import { __extends } from "tslib";
// (C) 2007-2018 GoodData Corporation
import React from "react";
import { injectIntl } from "react-intl";
var getNumericSymbols = function (intl) {
    return [
        "visualization.numericValues.k",
        "visualization.numericValues.m",
        "visualization.numericValues.g",
        "visualization.numericValues.t",
        "visualization.numericValues.p",
        "visualization.numericValues.e",
    ].map(function (id) { return intl.formatMessage({ id: id }); });
};
/**
 * @internal
 */
var TranslationsProvider = /** @class */ (function (_super) {
    __extends(TranslationsProvider, _super);
    function TranslationsProvider() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TranslationsProvider.prototype.render = function () {
        var props = {
            numericSymbols: getNumericSymbols(this.props.intl),
            emptyHeaderString: this.getEmptyHeaderString(),
            intl: this.props.intl,
        };
        return this.props.children(props);
    };
    TranslationsProvider.prototype.getEmptyHeaderString = function () {
        var emptyValueTranslation = this.props.intl.formatMessage({ id: "visualization.emptyValue" });
        return "(" + emptyValueTranslation + ")";
    };
    return TranslationsProvider;
}(React.PureComponent));
export { TranslationsProvider };
/**
 * @internal
 */
export var IntlTranslationsProvider = injectIntl(TranslationsProvider);
