// @gooddata/api-client-tiger/esm/client.js
// (C) 2019-2021 GoodData Corporation
import { tigerExecutionClientFactory } from "./execution";
import { tigerExecutionResultClientFactory } from "./executionResult";
import { LabelElementsConfiguration, LabelElementsBaseApi, tigerLabelElementsClientFactory, } from "./labelElements";
import { MetadataConfiguration, MetadataBaseApi, tigerWorkspaceObjectsClientFactory, } from "./workspaceObjects";
import { tigerValidObjectsClientFactory } from "./validObjects";
import { tigerOrganizationObjectsClientFactory } from "./organizationObjects";
import { setAxiosAuthorizationToken } from "./axios";
import { tigerLayoutClientFactory } from "./layout";
import { tigerAfmExplainClientFactory } from "./explain";
export { tigerWorkspaceObjectsClientFactory, tigerExecutionClientFactory, tigerExecutionResultClientFactory, tigerLabelElementsClientFactory, tigerValidObjectsClientFactory, tigerOrganizationObjectsClientFactory, tigerLayoutClientFactory, tigerAfmExplainClientFactory, MetadataConfiguration, MetadataBaseApi, LabelElementsConfiguration, LabelElementsBaseApi, };
/**
 * Tiger execution client
 *
 */
export var tigerClientFactory = function (axios) {
    var execution = tigerExecutionClientFactory(axios);
    var executionResult = tigerExecutionResultClientFactory(axios);
    var labelElements = tigerLabelElementsClientFactory(axios);
    var workspaceObjects = tigerWorkspaceObjectsClientFactory(axios);
    var validObjects = tigerValidObjectsClientFactory(axios);
    var organizationObjects = tigerOrganizationObjectsClientFactory(axios);
    var declarativeLayout = tigerLayoutClientFactory(axios);
    var explain = tigerAfmExplainClientFactory(axios);
    return {
        axios: axios,
        execution: execution,
        executionResult: executionResult,
        labelElements: labelElements,
        workspaceObjects: workspaceObjects,
        validObjects: validObjects,
        organizationObjects: organizationObjects,
        declarativeLayout: declarativeLayout,
        explain: explain,
        setApiToken: function (token) {
            setAxiosAuthorizationToken(axios, token);
        },
    };
};
