// @gooddata/api-client-tiger/esm/index.js
// (C) 2019-2022 GoodData Corporation
/**
 * This package provides low-level functions for communication with GoodData.CN.
 *
 * @remarks
 * The package is used by `@gooddata/sdk-backend-tiger`, which you should use instead of directly using
 * `@gooddata/api-client-tiger` whenever possible.
 * For the similar package for the GoodData platform, see `@gooddata/api-client-bear`.
 *
 * @packageDocumentation
 */
import { tigerClientFactory } from "./client";
import { axios as defaultAxios, newAxios, setAxiosAuthorizationToken, setGlobalAuthorizationToken, } from "./axios";
export { VisualizationObjectModelV1 } from "./gd-tiger-model/VisualizationObjectModelV1";
export { VisualizationObjectModelV2 } from "./gd-tiger-model/VisualizationObjectModelV2";
export { AnalyticalDashboardModelV1 } from "./gd-tiger-model/AnalyticalDashboardModelV1";
export { AnalyticalDashboardModelV2 } from "./gd-tiger-model/AnalyticalDashboardModelV2";
export { isAttributeHeader, isAfmObjectIdentifier, isResultAttributeHeader, isVisualizationObjectsItem, isFilterContextData, isDashboardPluginsItem, } from "./gd-tiger-model/typeGuards";
export { newAxios, setAxiosAuthorizationToken, setGlobalAuthorizationToken };
export { AfmValidObjectsQueryTypesEnum, ArithmeticMeasureDefinitionArithmeticMeasureOperatorEnum, AttributeHeaderAttributeHeaderGranularityEnum, ComparisonMeasureValueFilterBodyOperatorEnum, ComparisonMeasureValueFilterBodyAllOfOperatorEnum, ElementsRequestFilterByLabelTypeEnum, ElementsRequestSortOrderEnum, RangeMeasureValueFilterBodyOperatorEnum, RangeMeasureValueFilterBodyAllOfOperatorEnum, RankingFilterBodyOperatorEnum, RankingFilterBodyAllOfOperatorEnum, RelativeDateFilterBodyGranularityEnum, RelativeDateFilterBodyAllOfGranularityEnum, SimpleMeasureDefinitionMeasureAggregationEnum, SortDirection, TotalFunction, ActionsApiAxiosParamCreator as AfmActionsApiAxiosParamCreator, ActionsApiFp as AfmActionsApiFp, ActionsApiFactory as AfmActionsApiFactory, ActionsApi as AfmActionsApi, } from "./generated/afm-rest-api/api";
export * from "./generated/metadata-json-api/api";
export * from "./client";
export { jsonApiHeaders, JSON_API_HEADER_VALUE, ValidateRelationsHeader } from "./constants";
export { MetadataUtilities, } from "./metadataUtilities";
export { OrganizationUtilities, } from "./organizationUtilities";
var defaultTigerClient = tigerClientFactory(defaultAxios);
export default defaultTigerClient;
