// @gooddata/sdk-ui-ext/esm/internal/components/configurationPanels/BubbleChartConfigurationPanel.js
import { __extends } from "tslib";
// (C) 2019 GoodData Corporation
import React from "react";
import { FormattedMessage } from "react-intl";
import { Bubble, BubbleHoverTrigger } from "@gooddata/sdk-ui-kit";
import cx from "classnames";
import NameSubsection from "../configurationControls/axis/NameSubsection";
import { countItemsOnAxes } from "../pluggableVisualizations/baseChart/insightIntrospection";
import ConfigurationPanelContent from "./ConfigurationPanelContent";
import LabelSubsection from "../configurationControls/axis/LabelSubsection";
import ConfigSection from "../configurationControls/ConfigSection";
import DataLabelsControl from "../configurationControls/DataLabelsControl";
import CheckboxControl from "../configurationControls/CheckboxControl";
import MinMaxControl from "../configurationControls//MinMaxControl";
import { SHOW_DELAY_DEFAULT, HIDE_DELAY_DEFAULT, BUBBLE_ARROW_OFFSET_X, BUBBLE_ARROW_OFFSET_Y, } from "../../constants/bubble";
import { bucketsIsEmpty, insightBuckets } from "@gooddata/sdk-model";
import { BucketNames } from "@gooddata/sdk-ui";
var BubbleChartConfigurationPanel = /** @class */ (function (_super) {
    __extends(BubbleChartConfigurationPanel, _super);
    function BubbleChartConfigurationPanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BubbleChartConfigurationPanel.prototype.renderConfigurationPanel = function () {
        var _a, _b, _c, _d, _e, _f;
        var _g = this.props, featureFlags = _g.featureFlags, propertiesMeta = _g.propertiesMeta, properties = _g.properties, pushData = _g.pushData, type = _g.type, insight = _g.insight;
        var controls = properties && properties.controls;
        var xAxisVisible = (_b = (_a = controls === null || controls === void 0 ? void 0 : controls.xaxis) === null || _a === void 0 ? void 0 : _a.visible) !== null && _b !== void 0 ? _b : true;
        var yAxisVisible = (_d = (_c = controls === null || controls === void 0 ? void 0 : controls.yaxis) === null || _c === void 0 ? void 0 : _c.visible) !== null && _d !== void 0 ? _d : true;
        var gridEnabled = (_f = (_e = controls === null || controls === void 0 ? void 0 : controls.grid) === null || _e === void 0 ? void 0 : _e.enabled) !== null && _f !== void 0 ? _f : true;
        var controlsDisabled = this.isControlDisabled();
        var _h = countItemsOnAxes(type, controls, insight), itemsOnXAxis = _h.xaxis, itemsOnYAxis = _h.yaxis;
        var xAxisNameSectionDisabled = controlsDisabled || itemsOnXAxis !== 1;
        var yAxisNameSectionDisabled = controlsDisabled || itemsOnYAxis !== 1;
        var isNameSubsectionVisible = featureFlags.enableAxisNameConfiguration;
        var isAxisLabelsFormatEnabled = featureFlags.enableAxisLabelFormat;
        return (React.createElement(BubbleHoverTrigger, { showDelay: SHOW_DELAY_DEFAULT, hideDelay: HIDE_DELAY_DEFAULT },
            React.createElement("div", null,
                this.renderColorSection(),
                React.createElement(ConfigSection, { id: "xaxis_section", title: "properties.xaxis.title", valuePath: "xaxis.visible", canBeToggled: true, toggledOn: xAxisVisible, toggleDisabled: controlsDisabled, propertiesMeta: propertiesMeta, properties: properties, pushData: pushData },
                    isNameSubsectionVisible && (React.createElement(NameSubsection, { disabled: xAxisNameSectionDisabled, configPanelDisabled: controlsDisabled, axis: "xaxis", properties: properties, pushData: pushData })),
                    React.createElement(LabelSubsection, { disabled: controlsDisabled, configPanelDisabled: controlsDisabled, axis: "xaxis", properties: properties, pushData: pushData, showFormat: isAxisLabelsFormatEnabled }),
                    this.renderMinMax("xaxis")),
                React.createElement(ConfigSection, { id: "yaxis_section", title: "properties.yaxis.title", valuePath: "yaxis.visible", canBeToggled: true, toggledOn: yAxisVisible, toggleDisabled: controlsDisabled, propertiesMeta: propertiesMeta, properties: properties, pushData: pushData },
                    isNameSubsectionVisible && (React.createElement(NameSubsection, { disabled: yAxisNameSectionDisabled, configPanelDisabled: controlsDisabled, axis: "yaxis", properties: properties, pushData: pushData })),
                    React.createElement(LabelSubsection, { disabled: controlsDisabled, configPanelDisabled: controlsDisabled, axis: "yaxis", properties: properties, pushData: pushData, showFormat: isAxisLabelsFormatEnabled }),
                    this.renderMinMax("yaxis")),
                this.renderLegendSection(),
                React.createElement(ConfigSection, { id: "canvas_section", title: "properties.canvas.title", propertiesMeta: propertiesMeta, properties: properties, pushData: pushData },
                    React.createElement(DataLabelsControl, { pushData: pushData, properties: properties, isDisabled: this.areDataLabelsDisabled(), defaultValue: false, showDisabledMessage: this.isDataLabelsWarningShown() }),
                    React.createElement(CheckboxControl, { valuePath: "grid.enabled", labelText: "properties.canvas.gridline", properties: properties, checked: gridEnabled, disabled: controlsDisabled, pushData: pushData }))),
            React.createElement(Bubble, { className: this.getBubbleClassNames(), arrowOffsets: { "tc bc": [BUBBLE_ARROW_OFFSET_X, BUBBLE_ARROW_OFFSET_Y] }, alignPoints: [{ align: "tc bc" }] },
                React.createElement(FormattedMessage, { id: "properties.config.not_applicable" }))));
    };
    BubbleChartConfigurationPanel.prototype.renderMinMax = function (basePath) {
        var _a = this.props, pushData = _a.pushData, properties = _a.properties, propertiesMeta = _a.propertiesMeta;
        return (React.createElement(MinMaxControl, { isDisabled: this.isControlDisabled(), basePath: basePath, pushData: pushData, properties: properties, propertiesMeta: propertiesMeta }));
    };
    BubbleChartConfigurationPanel.prototype.areDataLabelsDisabled = function () {
        var isDisabled = _super.prototype.isControlDisabled.call(this);
        return isDisabled || !hasTertiaryMeasures(this.props.insight);
    };
    BubbleChartConfigurationPanel.prototype.isDataLabelsWarningShown = function () {
        var isDisabled = _super.prototype.isControlDisabled.call(this);
        return !isDisabled && !hasTertiaryMeasures(this.props.insight);
    };
    BubbleChartConfigurationPanel.prototype.getBubbleClassNames = function () {
        return cx("bubble-primary", {
            invisible: !this.isControlDisabled(),
        });
    };
    return BubbleChartConfigurationPanel;
}(ConfigurationPanelContent));
export default BubbleChartConfigurationPanel;
function hasTertiaryMeasures(insight) {
    return !bucketsIsEmpty(insightBuckets(insight, BucketNames.TERTIARY_MEASURES));
}
