// @gooddata/sdk-ui-ext/esm/internal/components/configurationPanels/UnsupportedConfigurationPanel.js
import { __extends } from "tslib";
// (C) 2019 GoodData Corporation
import React from "react";
import omit from "lodash/omit";
import ConfigurationPanelContent from "./ConfigurationPanelContent";
import UnsupportedProperties from "../configurationControls/UnsupportedProperties";
var UnsupportedConfigurationPanel = /** @class */ (function (_super) {
    __extends(UnsupportedConfigurationPanel, _super);
    function UnsupportedConfigurationPanel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    UnsupportedConfigurationPanel.prototype.componentDidMount = function () {
        this.props.pushData({
            properties: omit(this.props.properties, "controls"),
            references: null,
            ignoreUndoRedo: true,
        });
    };
    UnsupportedConfigurationPanel.prototype.isControlDisabled = function () {
        return true;
    };
    UnsupportedConfigurationPanel.prototype.renderConfigurationPanel = function () {
        return React.createElement(UnsupportedProperties, null);
    };
    return UnsupportedConfigurationPanel;
}(ConfigurationPanelContent));
export default UnsupportedConfigurationPanel;
