// @gooddata/sdk-ui-kit/esm/ColorPicker/components/ColorPickerMatrix.js
import { __extends } from "tslib";
// (C) 2007-2020 GoodData Corporation
import React, { PureComponent } from "react";
import cx from "classnames";
import { getColorStyle, SATURATION_ARRAY } from "../utils";
var ColorPickerMatrix = /** @class */ (function (_super) {
    __extends(ColorPickerMatrix, _super);
    function ColorPickerMatrix() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ColorPickerMatrix.prototype.render = function () {
        return (React.createElement("div", { className: "color-picker-component" },
            React.createElement(ColorPickerRow, { lightness: 0.7, initColor: this.props.initColor, onColorSelected: this.props.onColorSelected }),
            React.createElement(ColorPickerRow, { lightness: 0.5, initColor: this.props.initColor, onColorSelected: this.props.onColorSelected }),
            React.createElement(ColorPickerRow, { lightness: 0.3, initColor: this.props.initColor, onColorSelected: this.props.onColorSelected })));
    };
    return ColorPickerMatrix;
}(PureComponent));
export { ColorPickerMatrix };
var ColorPickerRow = /** @class */ (function (_super) {
    __extends(ColorPickerRow, _super);
    function ColorPickerRow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ColorPickerRow.prototype.getCellClassNames = function (hslColor) {
        return cx("color-picker-cell", "s-color-" + Math.floor(hslColor.h + (hslColor.s + hslColor.l) * 100));
    };
    ColorPickerRow.prototype.render = function () {
        var _this = this;
        return (React.createElement("div", { className: "color-picker-row" }, SATURATION_ARRAY.map(function (saturation) {
            var newColor = {
                h: _this.props.initColor.h,
                s: saturation,
                l: _this.props.lightness,
            };
            return (React.createElement("div", { tabIndex: -1, key: _this.props.lightness + saturation, className: _this.getCellClassNames(newColor), style: getColorStyle(newColor), onClick: function () { return _this.props.onColorSelected(newColor); } }));
        })));
    };
    return ColorPickerRow;
}(PureComponent));
