// @gooddata/sdk-backend-bear/esm/convertors/toBackend/MeasureConverter.js
import { __assign } from "tslib";
import { isSimpleMeasure, measureAlias, measureFormat, measureLocalId, measureTitle, measureAggregation, measureDoesComputeRatio, measureIdentifier, measureUri, isArithmeticMeasure, measureArithmeticOperands, measureArithmeticOperator, isPoPMeasure, measureMasterIdentifier, measurePopAttribute, isPreviousPeriodMeasure, measurePreviousPeriodDateDataSets, measureFilters, } from "@gooddata/sdk-model";
import isEmpty from "lodash/isEmpty";
import { toBearRef } from "./ObjRefConverter";
import { convertFilter } from "./FilterConverter";
import { convertAggregation } from "./afm/MeasureConverter";
var convertPreviousPeriodDataSet = function (dataSet) {
    return {
        dataSet: toBearRef(dataSet.dataSet),
        periodsAgo: dataSet.periodsAgo,
    };
};
var convertPreviousPeriodMeasureDefinition = function (measure) {
    return {
        previousPeriodMeasure: {
            measureIdentifier: measureMasterIdentifier(measure),
            dateDataSets: measurePreviousPeriodDateDataSets(measure).map(convertPreviousPeriodDataSet),
        },
    };
};
var convertPoPMeasureDefinition = function (measure) {
    return {
        popMeasureDefinition: {
            measureIdentifier: measureMasterIdentifier(measure),
            popAttribute: toBearRef(measurePopAttribute(measure)),
        },
    };
};
var convertArithmeticMeasureDefinition = function (measure) {
    return {
        arithmeticMeasure: {
            measureIdentifiers: measureArithmeticOperands(measure),
            operator: measureArithmeticOperator(measure),
        },
    };
};
var convertSimpleMeasureDefinition = function (measure) {
    var identifier = measureIdentifier(measure);
    var uri = measureUri(measure);
    if (!identifier && !uri) {
        throw new Error("Measure has neither uri nor identifier.");
    }
    var aggregation = convertAggregation(measureAggregation(measure));
    var computeRatio = measureDoesComputeRatio(measure);
    var filters = (measureFilters(measure) || []).map(convertFilter);
    return {
        measureDefinition: __assign(__assign(__assign({ item: identifier ? { identifier: identifier } : { uri: uri } }, (aggregation && { aggregation: aggregation })), (computeRatio && { computeRatio: computeRatio })), (!isEmpty(filters) && { filters: filters })),
    };
};
var convertMeasureDefinition = function (measure) {
    if (isSimpleMeasure(measure)) {
        return convertSimpleMeasureDefinition(measure);
    }
    else if (isArithmeticMeasure(measure)) {
        return convertArithmeticMeasureDefinition(measure);
    }
    else if (isPoPMeasure(measure)) {
        return convertPoPMeasureDefinition(measure);
    }
    else if (isPreviousPeriodMeasure(measure)) {
        return convertPreviousPeriodMeasureDefinition(measure);
    }
    throw new Error("Unknown measure type");
};
export var convertMeasure = function (measure) {
    var alias = measureAlias(measure);
    var format = measureFormat(measure);
    var title = measureTitle(measure);
    return {
        measure: __assign(__assign(__assign({ definition: convertMeasureDefinition(measure), localIdentifier: measureLocalId(measure) }, (alias && { alias: alias })), (format && { format: format })), (title && { title: title })),
    };
};
