// @gooddata/sdk-ui-dashboard/esm/model/store/storeAccessors/DashboardStoreAccessor.js
// (C) 2022 GoodData Corporation
import { invariant } from "ts-invariant";
/**
 * This class serves the selector and the dispatcher properties of the dashboard component state.
 *
 * The {@link Dashboard} component has an optional property {@link IDashboardProps#onStateChange} through which
 * you can handle set the values for {@link DashboardDispatch} and {@link DashboardSelectorEvaluator}.
 *
 *
 * In your component using {@link Dashboard} you can create and instance of the {@link DashboardStoreAccessor} object
 * and use it like in the example below. The example shows the accessor's usage as well. There is a need to check
 * the select and dispatch object existence.
 *
 * See {@link DashboardStoreAccessorRepository} on possible way how to use the store accessor.
 *
 * To get latest properties, use static member function {@link DashboardStoreAccessor#getInstance}. If there is already an instance
 * created, it will return this instance and will return new instance of the {@link DashboardStoreAccessor} otherwise.
 *
 * @public
 */
var DashboardStoreAccessor = /** @class */ (function () {
    function DashboardStoreAccessor(selector, dispatch) {
        var _this = this;
        /**
         * Returns current selector for the dashboard's component state.
         */
        this.getDashboardSelect = function () {
            invariant(_this.selectorEvaluator, "DashboardStoreAccessor selectorEvaluator is not initialized");
            return _this.selectorEvaluator;
        };
        /**
         * Setter for the dashboard's component state selector.
         * @param selectorEvaluator - dashboardSelectorEvaluator
         */
        this.setSelector = function (selectorEvaluator) {
            _this.selectorEvaluator = selectorEvaluator;
        };
        /**
         * Returns current dispatch object for the dashboard component state.
         */
        this.getDashboardDispatch = function () {
            invariant(_this.dispatch, "DashboardStoreAccessor dispatch is not initialized");
            return _this.dispatch;
        };
        /**
         * Setter for the dashboard's component state dispatch.
         * @param dispatch - dashboardDispatch
         */
        this.setDispatch = function (dispatch) {
            _this.dispatch = dispatch;
        };
        /**
         * Checks if {@link DashboardStoreAccessor} is fully initialized.
         */
        this.isDashboardStoreAccessorInitialized = function () {
            return !!_this.selectorEvaluator && !!_this.dispatch;
        };
        /**
         * Callback to be passed as {@link Dashboard} component {@link Dashboard#onStateChange} property to set
         * {@link DashboardStoreAccessor#selectorEvaluator} and {@link DashboardStoreAccessor#dispatch} to handle Dashboard component state from outside of the
         * component.
         *
         * @param state - {@link DashboardState} object.
         * @param dispatch - {@link DashboardDispatch} object.
         */
        this.setSelectAndDispatch = function (state, dispatch) {
            var dashboardSelect = function (select) { return select(state); };
            _this.setSelector(dashboardSelect);
            _this.setDispatch(dispatch);
        };
        this.selectorEvaluator = selector;
        this.dispatch = dispatch;
    }
    return DashboardStoreAccessor;
}());
export { DashboardStoreAccessor };
