// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/pieChart/pieConfiguration.js
import { __assign } from "tslib";
import { alignChart } from "../_chartCreators/helpers";
import { getPieResponsiveConfig } from "../_chartCreators/responsive";
export function getPieConfiguration(config) {
    var pieConfiguration = {
        chart: {
            type: "pie",
            events: {
                load: function () {
                    var _a, _b, _c, _d, _e;
                    var distance = -(((_d = (_c = (_b = (_a = this.series[0].points) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.shapeArgs) === null || _c === void 0 ? void 0 : _c.r) !== null && _d !== void 0 ? _d : 30) / 3);
                    var options = {
                        type: "pie",
                        dataLabels: {
                            distance: distance,
                        },
                    };
                    this.series[0].update(options);
                    alignChart(this, (_e = config.chart) === null || _e === void 0 ? void 0 : _e.verticalAlign);
                },
            },
        },
        plotOptions: {
            pie: {
                size: "100%",
                allowPointSelect: false,
                dataLabels: {
                    enabled: false,
                },
                showInLegend: true,
            },
        },
        legend: {
            enabled: false,
        },
    };
    if (config === null || config === void 0 ? void 0 : config.enableCompactSize) {
        return __assign(__assign({}, pieConfiguration), { responsive: getPieResponsiveConfig() });
    }
    return pieConfiguration;
}
