// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/_util/color.js
// (C) 2007-2021 GoodData Corporation
import { DefaultColorPalette } from "@gooddata/sdk-ui";
import { getColorPaletteFromColors } from "@gooddata/sdk-ui-vis-commons";
import isEmpty from "lodash/isEmpty";
export var WHITE = "rgb(255, 255, 255)";
export var BLACK = "rgb(0, 0, 0)";
export var GRAY = "rgb(201, 213, 223)";
export var AXIS_LINE_COLOR = "#d5d5d5";
export var TRANSPARENT = "transparent";
export var HEATMAP_BLUE_COLOR_PALETTE = [
    "rgb(255,255,255)",
    "rgb(197,236,248)",
    "rgb(138,217,241)",
    "rgb(79,198,234)",
    "rgb(20,178,226)",
    "rgb(22,151,192)",
    "rgb(0,110,145)",
];
export var DEFAULT_HEATMAP_BLUE_COLOR = { r: 0, g: 110, b: 145 };
export var DEFAULT_HEATMAP_BLUE_BASE_COLOR = { r: 20, g: 178, b: 226 };
export var DEFAULT_BULLET_GRAY_COLOR = {
    r: 217,
    g: 220,
    b: 226,
};
/**
 * @internal
 */
export function getValidColorPalette(config) {
    return isEmpty(config.colorPalette)
        ? isEmpty(config.colors)
            ? DefaultColorPalette
            : getColorPaletteFromColors(config.colors)
        : config.colorPalette;
}
