// @gooddata/sdk-model/esm/insight/factory.js
// (C) 2020 GoodData Corporation
import identity from "lodash/identity";
/*
 * Factory & builder for insight instances. Keeping it in test infrastructure for now, will see later on
 * whether we should move it to prod code and expose on public API.
 */
/**
 * Creates new, empty insight definition, optionally modifying its content.
 *
 * @param visualizationUrl - visualization URL (e.g. local:bar, local:table..)
 * @param modifications - modification function which will be called with builder to update the insight
 * @internal
 */
export function newInsightDefinition(visualizationUrl, modifications) {
    if (modifications === void 0) { modifications = identity; }
    var builder = new InsightDefinitionBuilder(visualizationUrl);
    return modifications(builder).build();
}
/**
 * Insight definition builder can be used to set various properties of the insight using fluent API.
 *
 * @internal
 */
var InsightDefinitionBuilder = /** @class */ (function () {
    function InsightDefinitionBuilder(visualizationUrl) {
        var _this = this;
        this.title = function (title) {
            _this.insight.title = title;
            return _this;
        };
        this.buckets = function (buckets) {
            _this.insight.buckets = buckets;
            return _this;
        };
        this.filters = function (filters) {
            _this.insight.filters = filters;
            return _this;
        };
        this.sorts = function (sorts) {
            _this.insight.sorts = sorts;
            return _this;
        };
        this.properties = function (properties) {
            _this.insight.properties = properties;
            return _this;
        };
        this.build = function () {
            return {
                insight: _this.insight,
            };
        };
        this.insight = {
            visualizationUrl: visualizationUrl,
            title: "Untitled",
            buckets: [],
            filters: [],
            sorts: [],
            properties: {},
        };
    }
    return InsightDefinitionBuilder;
}());
export { InsightDefinitionBuilder };
