// @gooddata/sdk-ui-vis-commons/esm/index.js
// (C) 2019-2021 GoodData Corporation
/**
 * This package provides functions commonly used when building visualizations.
 *
 * @remarks
 * This package is mainly used internally by other `@gooddata/sdk-ui-*` packages, and we do not recommend using
 * it directly outside of GoodData because its API can change at any time.
 *
 * @packageDocumentation
 */
export { AttributeColorStrategy } from "./coloring/attribute";
export { ColorStrategy, getAtributeColorAssignment, isValidMappedColor, } from "./coloring/base";
export { getColorByGuid, getColorFromMapping, getColorMappingPredicate, getColorPaletteFromColors, getLighterColor, getLighterColorFromRGB, getRgbString, getRgbStringFromRGB, isCustomPalette, normalizeColorToRGB, parseRGBColorCode, getValidColorPalette, ColorUtils, parseRGBString, } from "./coloring/color";
export { Legend } from "./legend/Legend";
export { StaticLegend } from "./legend/StaticLegend";
export { PopUpLegend } from "./legend/PopUpLegend/PopUpLegend";
export { FluidLegend } from "./legend/FluidLegend";
export { HeatmapLegend } from "./legend/HeatmapLegend";
export { ColorLegend } from "./legend/ColorLegend";
export { Paging } from "./legend/Paging";
export { HeadlinePagination } from "./compactSize/HeadlinePagination";
export { formatLegendLabel, shouldShowFluid, FLUID_LEGEND_THRESHOLD } from "./legend/helpers";
export { DEFAULT_LEGEND_CONFIG, LegendPosition, SupportedLegendPositions, } from "./legend/types";
export { fixEmptyHeaderItems } from "./utils/fixEmptyHeaderItems";
export { shouldRenderPagination, calculateHeadlineHeightFontSize } from "./utils/calculateCustomHeight";
export { getHeadlineResponsiveClassName } from "./utils/headlineResponsiveClassName";
export { getLegendDetails } from "./legend/PopUpLegend/helpers";
