// @gooddata/sdk-ui-kit/esm/Dialog/HubspotConversionTouchPointDialogBase.js
import { __read } from "tslib";
// (C) 2021 GoodData Corporation
import React, { useEffect, useState } from "react";
import { useIntl } from "react-intl";
import { useHubspotForm } from "@aaronhayes/react-use-hubspot-form";
import { LoadingComponent } from "@gooddata/sdk-ui";
import { Button } from "../Button";
/**
 * @internal
 */
export var HubspotConversionTouchPointDialogBase = function (props) {
    var targetId = props.targetId, hubspotPortalId = props.hubspotPortalId, hubspotFormId = props.hubspotFormId, selectedValue = props.selectedValue, cancelButtonText = props.cancelButtonText, dialogTitle = props.dialogTitle, showCancelButton = props.showCancelButton, submitButtonClass = props.submitButtonClass, _a = props.values, values = _a === void 0 ? {} : _a, onClose = props.onClose, onFormSubmitted = props.onFormSubmitted;
    var intl = useIntl();
    var _b = __read(useState(false), 2), isFormReady = _b[0], setIsFormReady = _b[1];
    var hubspotFormTargetId = targetId || "conversion-touch-point-hubspot";
    useEffect(function () {
        var _a;
        // the Hubspot form requires jQuery object, we need to create a fake jQuery object to bypass this.
        window.jQuery =
            window.jQuery ||
                function (nodeOrSelector) {
                    if (typeof nodeOrSelector == "string") {
                        return document.querySelector(nodeOrSelector);
                    }
                    return nodeOrSelector;
                };
        if ((_a = window.hbspt) === null || _a === void 0 ? void 0 : _a.isSuccessMessageShow) {
            window.hbspt.isSuccessMessageShow = false;
        }
    }, []);
    var onHubspotFormSubmitted = function () {
        if (!window.hbspt.isSuccessMessageShow) {
            onFormSubmitted === null || onFormSubmitted === void 0 ? void 0 : onFormSubmitted();
            onClose();
            window.hbspt.isSuccessMessageShow = true;
        }
    };
    var formCreated = useHubspotForm({
        portalId: hubspotPortalId,
        formId: hubspotFormId,
        target: "#" + hubspotFormTargetId,
        locale: intl.locale.split("-").shift(),
        submitButtonClass: submitButtonClass,
        onFormSubmitted: onHubspotFormSubmitted,
        onFormReady: function ($form) {
            var _a;
            setIsFormReady(true);
            var fields = $form;
            if (fields["jquery"] && ((_a = fields[0]) === null || _a === void 0 ? void 0 : _a.length) > 0) {
                fields = fields[0];
            }
            // populating the values for hidden fields
            for (var i = 0; i < fields.length; i += 1) {
                var inputField = fields[i];
                if (!inputField) {
                    continue;
                }
                var populatedValue = values[inputField.name];
                if (populatedValue) {
                    inputField.value = populatedValue;
                }
                if (selectedValue && inputField.value.indexOf(selectedValue) !== -1) {
                    // Call click event to check the radio button on Hubspot form
                    inputField.click();
                }
            }
        },
    }).formCreated;
    return (React.createElement(React.Fragment, null,
        !formCreated && React.createElement(LoadingComponent, null),
        formCreated && (React.createElement("div", { className: "conversion-touch-point-dialog-wrapper" },
            isFormReady && dialogTitle && React.createElement("h3", null, dialogTitle),
            React.createElement("div", { id: hubspotFormTargetId, className: "hubspot-form" }),
            isFormReady && showCancelButton && (React.createElement(Button, { className: "gd-button-secondary", onClick: onClose, value: cancelButtonText }))))));
};
