// @gooddata/sdk-ui-dashboard/esm/model/commands/alerts.js
// (C) 2021-2022 GoodData Corporation
/**
 * Creates the SaveAlert command. Dispatching this command will result in the creating Kpi alert on the backend.
 *
 * @param alert - specify alert to create.
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing
 * @alpha
 */
export function createAlert(alert, correlationId) {
    return {
        type: "GDC.DASH/CMD.ALERT.CREATE",
        correlationId: correlationId,
        payload: {
            alert: alert,
        },
    };
}
/**
 * Creates the UpdateAlert command. Dispatching this command will result in the updating Kpi alert on the backend.
 *
 * @param alert - specify alert to update.
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @alpha
 */
export function updateAlert(alert, correlationId) {
    return {
        type: "GDC.DASH/CMD.ALERT.UPDATE",
        correlationId: correlationId,
        payload: {
            alert: alert,
        },
    };
}
/**
 * Creates the RemoveAlerts command. Dispatching this command will result in the removing Kpi alerts on the backend.
 *
 * @param refs - specify ObjRef of the alerts to remove
 * @param correlationId - optionally specify correlation id to use for this command. this will be included in all
 *  events that will be emitted during the command processing

 * @alpha
 */
export function removeAlerts(refs, correlationId) {
    return {
        type: "GDC.DASH/CMD.ALERTS.REMOVE",
        correlationId: correlationId,
        payload: {
            refs: refs,
        },
    };
}
