// @gooddata/sdk-ui-dashboard/esm/model/commands/executionResults.js
// (C) 2021 GoodData Corporation
import { __assign } from "tslib";
import { serializeObjRef } from "@gooddata/sdk-model";
import isString from "lodash/isString";
function upsertExecutionResult(id, envelopeData, correlationId) {
    return {
        type: "GDC.DASH/CMD.EXECUTION_RESULT.UPSERT",
        correlationId: correlationId,
        payload: __assign(__assign({}, envelopeData), { id: isString(id) ? id : serializeObjRef(id) }),
    };
}
/**
 * Creates an {@link UpsertExecutionResult} command that makes the relevant execution result indicate it is loading.
 *
 * @alpha
 */
export function setExecutionResultLoading(id, correlationId) {
    return upsertExecutionResult(id, {
        isLoading: true,
        executionResult: undefined,
        error: undefined,
    }, correlationId);
}
/**
 * Creates an {@link UpsertExecutionResult} command that makes the relevant execution result indicate an error and stop loading.
 *
 * @alpha
 */
export function setExecutionResultError(id, error, correlationId) {
    return upsertExecutionResult(id, {
        isLoading: false,
        error: error,
        executionResult: undefined,
    }, correlationId);
}
/**
 * Creates an {@link UpsertExecutionResult} command that makes the relevant execution result set new result data and stop loading.
 *
 * @alpha
 */
export function setExecutionResultData(id, executionResult, correlationId) {
    return upsertExecutionResult(id, {
        isLoading: false,
        error: undefined,
        executionResult: executionResult,
    }, correlationId);
}
