// @gooddata/sdk-ui-dashboard/esm/model/commands/render.js
/**
 * Notify the dashboard about async rendering of the component.
 *
 * Mechanism is following:
 * - You must request async rendering for at least 1 component within 2 seconds of the {@link DashboardInitialized} event.
 *   (If you do not register any asynchronous rendering, after 2 seconds the dashboard will announce that it is rendered.)
 * - You can request async rendering for any number of components. Requests are valid if the first rule is met
 *   and not all asynchronous renderings have been resolved and the maximum timeout (60s by default) has not elapsed.
 * - The component may again request asynchronous rendering within 2 seconds of resolution. Maximum 3x.
 *   (this is necessary to cover re-renders caused by data received from the component - eg pushData)
 *
 * - Each component on the dashboard that is rendered asynchronously should fire this command.
 * - Once the component is rendered, it should notify the dashboard by dispatching {@link resolveAsyncRender} command
 *   with the corresponding identifier.
 *
 * In this way, the dashboard is able to recognize and notify that it is fully rendered.
 * (which is useful, for example, when exporting)
 *
 * @alpha
 * @param id - async render identifier
 * @param correlationId - optionally specify correlation id to use for this command.
 *                        this will be included in all events that will be emitted during the command processing
 * @returns
 */
export function requestAsyncRender(id, correlationId) {
    return {
        type: "GDC.DASH/CMD.RENDER.ASYNC.REQUEST",
        correlationId: correlationId,
        payload: {
            id: id,
        },
    };
}
/**
 * Notify the dashboard about resolved async rendering of the component.
 *
 * - Each component on the dashboard that is rendered asynchronously should fire this command.
 * - This command should only be dispatched if a {@link requestAsyncRender} command with the corresponding identifier
 *   has already been dispatched.
 *
 * @alpha
 * @param id - async render identifier
 * @param correlationId - optionally specify correlation id to use for this command.
 *                        this will be included in all events that will be emitted during the command processing
 * @returns
 */
export function resolveAsyncRender(id, correlationId) {
    return {
        type: "GDC.DASH/CMD.RENDER.ASYNC.RESOLVE",
        correlationId: correlationId,
        payload: {
            id: id,
        },
    };
}
