// @gooddata/sdk-ui-charts/esm/charts/barChart/BarChart.js
import { __assign, __spreadArrays } from "tslib";
// (C) 2019 GoodData Corporation
import React from "react";
import { applyRatioRule, newBucket, } from "@gooddata/sdk-model";
import { BucketNames, useResolveValuesWithPlaceholders, } from "@gooddata/sdk-ui";
import { ViewByAttributesLimit } from "../../interfaces";
import { truncate } from "../_commons/truncate";
import { CoreBarChart } from "./CoreBarChart";
import { stackedChartDimensions } from "../_commons/dimensions";
import { withChart } from "../_base/withChart";
import { sanitizeConfig } from "../_commons/sanitizeStacking";
//
// Internals
//
var barChartDefinition = {
    chartName: "BarChart",
    bucketPropsKeys: ["measures", "viewBy", "stackBy", "filters", "sortBy"],
    bucketsFactory: function (props) {
        var measures = applyRatioRule(props.measures);
        var viewBy = truncate(props.viewBy, ViewByAttributesLimit);
        return [
            newBucket.apply(void 0, __spreadArrays([BucketNames.MEASURES], measures)),
            newBucket.apply(void 0, __spreadArrays([BucketNames.VIEW], viewBy)),
            newBucket(BucketNames.STACK, props.stackBy),
        ];
    },
    executionFactory: function (props, buckets) {
        var _a;
        var backend = props.backend, workspace = props.workspace, execConfig = props.execConfig;
        return (_a = backend
            .withTelemetry("BarChart", props)
            .workspace(workspace)
            .execution()
            .forBuckets(buckets, props.filters))
            .withSorting.apply(_a, props.sortBy).withDimensions(stackedChartDimensions)
            .withExecConfig(execConfig);
    },
    propOverridesFactory: function (props, buckets) {
        return {
            config: sanitizeConfig(buckets, props.config),
        };
    },
};
var WrappedBarChart = withChart(barChartDefinition)(CoreBarChart);
/**
 * [BarChart](http://sdk.gooddata.com/gooddata-ui/docs/bar_chart_component.html)
 *
 * Bar chart shows data in horizontal bars. Bar charts can display one or multiple metrics side by side divided by
 * attribute values or a single measure stacked by attribute values.
 *
 * @remarks See {@link IBarChartProps} to learn how to configure the AreaChart
 * @public
 */
export var BarChart = function (props) {
    var _a = useResolveValuesWithPlaceholders([props.measures, props.viewBy, props.stackBy, props.filters, props.sortBy], props.placeholdersResolutionContext), measures = _a[0], viewBy = _a[1], stackBy = _a[2], filters = _a[3], sortBy = _a[4];
    return (React.createElement(WrappedBarChart, __assign({}, props, {
        measures: measures,
        viewBy: viewBy,
        stackBy: stackBy,
        filters: filters,
        sortBy: sortBy,
    })));
};
