// @gooddata/sdk-ui-dashboard/esm/presentation/presentationComponents/DashboardItems/DashboardItemHeadlineContainer.js
import { __assign } from "tslib";
// (C) 2020 GoodData Corporation
import React, { memo } from "react";
var SMALLEST_WIDGET_HEIGHT = 120;
var SMALL_WIDGET_HEIGHT = 140;
var SMALL_HEIGHT = 38;
var SMALL_LINE_HEIGHT = 36;
var INNER_STYLE_SMALL = { fontSize: "15px" };
var CUSTOM_MARGIN_STYLE = { marginTop: 0 };
var CUSTOM_OUTER_STYLE = {
    height: SMALL_HEIGHT + "px",
    lineHeight: SMALL_LINE_HEIGHT + "px",
};
var CUSTOM_OUTER_STYLE_WITH_CUSTOM_MARGIN = __assign(__assign({}, CUSTOM_OUTER_STYLE), { CUSTOM_MARGIN_STYLE: CUSTOM_MARGIN_STYLE });
function isSmallCustomHeight(clientHeight) {
    return clientHeight !== undefined ? clientHeight < SMALLEST_WIDGET_HEIGHT : false;
}
// compose the styles "statically" so that the result's referential equality is retained where possible
function getOuterStyle(hasCustomMargin, isCustomHeightSmall) {
    if (isCustomHeightSmall) {
        return hasCustomMargin ? CUSTOM_OUTER_STYLE_WITH_CUSTOM_MARGIN : CUSTOM_OUTER_STYLE;
    }
    if (hasCustomMargin) {
        return CUSTOM_MARGIN_STYLE;
    }
}
// inner utility component for better caching: caches according to two booleans, not a clientHeight number
var DashboardItemHeadlineContainerInner = memo(function DashboardItemHeadlineContainerInner(_a) {
    var hasCustomMargin = _a.hasCustomMargin, isCustomHeightSmall = _a.isCustomHeightSmall, children = _a.children;
    return (React.createElement("div", { className: "item-headline-outer", style: getOuterStyle(hasCustomMargin, isCustomHeightSmall) },
        React.createElement("div", { className: "item-headline", style: isCustomHeightSmall ? INNER_STYLE_SMALL : undefined }, children)));
});
export var DashboardItemHeadlineContainer = function (_a) {
    var children = _a.children, clientHeight = _a.clientHeight;
    return (React.createElement(DashboardItemHeadlineContainerInner, { hasCustomMargin: clientHeight !== undefined ? clientHeight <= SMALL_WIDGET_HEIGHT : false, isCustomHeightSmall: isSmallCustomHeight(clientHeight) }, children));
};
