// @gooddata/sdk-backend-base/esm/ldmFactories/metadata/factory.js
// (C) 2019-2021 GoodData Corporation
import { __extends } from "tslib";
import { Builder } from "../builder";
/**
 * Metadata object builder
 * See {@link Builder}
 *
 * @beta
 */
var MetadataObjectBuilder = /** @class */ (function (_super) {
    __extends(MetadataObjectBuilder, _super);
    function MetadataObjectBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MetadataObjectBuilder.prototype.title = function (title) {
        this.item.title = title;
        return this;
    };
    MetadataObjectBuilder.prototype.description = function (description) {
        this.item.description = description;
        return this;
    };
    MetadataObjectBuilder.prototype.id = function (identifier) {
        this.item.id = identifier;
        return this;
    };
    MetadataObjectBuilder.prototype.uri = function (uri) {
        this.item.uri = uri;
        return this;
    };
    MetadataObjectBuilder.prototype.unlisted = function (value) {
        this.item.unlisted = value;
        return this;
    };
    MetadataObjectBuilder.prototype.production = function (isProduction) {
        this.item.production = isProduction;
        return this;
    };
    MetadataObjectBuilder.prototype.deprecated = function (isDeprecated) {
        this.item.deprecated = isDeprecated;
        return this;
    };
    return MetadataObjectBuilder;
}(Builder));
export { MetadataObjectBuilder };
