// @gooddata/sdk-ui-ext/esm/internal/components/configurationControls/ConfigSection.js
import { __extends } from "tslib";
// (C) 2019 GoodData Corporation
import React from "react";
import { injectIntl } from "react-intl";
import cx from "classnames";
import cloneDeep from "lodash/cloneDeep";
import noop from "lodash/noop";
import set from "lodash/set";
import DisabledBubbleMessage from "../DisabledBubbleMessage";
import { getTranslation } from "../../utils/translations";
var ConfigSection = /** @class */ (function (_super) {
    __extends(ConfigSection, _super);
    function ConfigSection(props) {
        var _a, _b, _c;
        var _this = _super.call(this, props) || this;
        _this.toggleCollapsed = _this.toggleCollapsed.bind(_this);
        _this.toggleValue = _this.toggleValue.bind(_this);
        var collapsed = (_c = (_b = (_a = props.propertiesMeta) === null || _a === void 0 ? void 0 : _a[props.id]) === null || _b === void 0 ? void 0 : _b.collapsed) !== null && _c !== void 0 ? _c : true;
        _this.state = {
            collapsed: collapsed,
        };
        return _this;
    }
    ConfigSection.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        var _a, _b, _c;
        // TODO: should the indexer be "nextProps.id"? Leaving as-is for now to be safe.
        var collapsed = (_c = (_b = (_a = nextProps.propertiesMeta) === null || _a === void 0 ? void 0 : _a[this.props.id]) === null || _b === void 0 ? void 0 : _b.collapsed) !== null && _c !== void 0 ? _c : true;
        this.setState({ collapsed: collapsed });
    };
    ConfigSection.prototype.render = function () {
        var collapsed = this.state.collapsed;
        var _a = this.props, title = _a.title, intl = _a.intl, subtitle = _a.subtitle, className = _a.className;
        var configSectionClassName = "adi-bucket-configuration " + className;
        return (React.createElement("div", { className: this.getSectionClassNames() },
            React.createElement("div", { className: this.getHeaderClassNames(), onClick: this.toggleCollapsed }, getTranslation(title, intl) + (subtitle ? " (" + getTranslation(subtitle, intl) + ")" : "")),
            this.renderToggleSwitch(),
            React.createElement("div", { className: configSectionClassName }, !collapsed && this.props.children)));
    };
    ConfigSection.prototype.renderToggleSwitch = function () {
        if (this.props.canBeToggled) {
            var _a = this.props, toggledOn = _a.toggledOn, toggleDisabled = _a.toggleDisabled, showDisabledMessage = _a.showDisabledMessage;
            return (React.createElement(DisabledBubbleMessage, { className: "adi-bucket-item-toggle", showDisabledMessage: showDisabledMessage },
                React.createElement("label", { className: this.getToggleLabelClassNames() },
                    React.createElement("input", { type: "checkbox", checked: toggledOn, disabled: toggleDisabled, onChange: this.toggleValue, className: "s-checkbox-toggle " + (toggleDisabled ? "s-disabled" : "s-enabled") }),
                    React.createElement("span", { className: "input-label-text" }))));
        }
        return null;
    };
    ConfigSection.prototype.getHeaderClassNames = function () {
        var collapsed = this.state.collapsed;
        return cx("adi-bucket-item-header", "adi-configuration-section-header", "s-configuration-panel-item-header", {
            expanded: !collapsed,
            "adi-bucket-item-header-has-toggle": this.props.canBeToggled,
            collapsed: collapsed,
        });
    };
    ConfigSection.prototype.getToggleLabelClassNames = function () {
        return cx("input-checkbox-toggle", "s-checkbox-toggle-label");
    };
    ConfigSection.prototype.getSectionClassNames = function () {
        return cx("adi-bucket-item", "s-config-section-" + this.props.id);
    };
    ConfigSection.prototype.toggleCollapsed = function () {
        var collapsed = this.state.collapsed;
        var propertiesMeta = {};
        propertiesMeta[this.props.id] = {
            collapsed: !collapsed,
        };
        this.setState({ collapsed: !collapsed });
        this.props.pushData({ propertiesMeta: propertiesMeta });
    };
    ConfigSection.prototype.toggleValue = function (event) {
        var _a = this.props, valuePath = _a.valuePath, properties = _a.properties, pushData = _a.pushData;
        if (valuePath && properties && pushData) {
            var clonedProperties = cloneDeep(properties);
            set(clonedProperties, "controls." + valuePath, event.target.checked);
            pushData({ properties: clonedProperties });
        }
    };
    ConfigSection.defaultProps = {
        collapsed: true,
        canBeToggled: false,
        toggleDisabled: false,
        toggledOn: true,
        disabled: false,
        pushData: noop,
        showDisabledMessage: false,
        className: "",
        properties: {},
    };
    return ConfigSection;
}(React.Component));
export { ConfigSection };
export default injectIntl(ConfigSection);
