// @gooddata/sdk-ui-ext/esm/internal/components/configurationControls/ConfigSubsection.js
import { __extends } from "tslib";
// (C) 2019 GoodData Corporation
import React from "react";
import { injectIntl } from "react-intl";
import cloneDeep from "lodash/cloneDeep";
import noop from "lodash/noop";
import set from "lodash/set";
import DisabledBubbleMessage from "../DisabledBubbleMessage";
import { getTranslation } from "../../utils/translations";
var ConfigSubsection = /** @class */ (function (_super) {
    __extends(ConfigSubsection, _super);
    function ConfigSubsection(props) {
        var _this = _super.call(this, props) || this;
        _this.toggleValue = _this.toggleValue.bind(_this);
        return _this;
    }
    ConfigSubsection.prototype.render = function () {
        var _a = this.props, title = _a.title, intl = _a.intl;
        var className = "configuration-subsection " + this.getTestClassName();
        return (React.createElement("div", { className: className },
            React.createElement("fieldset", null,
                React.createElement("legend", null,
                    React.createElement("span", { className: "legend-title" }, getTranslation(title, intl)),
                    this.renderToggleSwitch()),
                React.createElement("div", null, this.props.children))));
    };
    ConfigSubsection.prototype.renderToggleSwitch = function () {
        if (this.props.canBeToggled) {
            var _a = this.props, toggledOn = _a.toggledOn, toggleDisabled = _a.toggleDisabled, showDisabledMessage = _a.showDisabledMessage;
            return (React.createElement(DisabledBubbleMessage, { className: "input-checkbox-toggle", showDisabledMessage: showDisabledMessage },
                React.createElement("label", { className: "s-checkbox-toggle-label" },
                    React.createElement("input", { type: "checkbox", checked: toggledOn, disabled: toggleDisabled, onChange: this.toggleValue, className: "s-checkbox-toggle" }),
                    React.createElement("span", { className: "input-label-text" }))));
        }
        return null;
    };
    ConfigSubsection.prototype.toggleValue = function (event) {
        var _a = this.props, valuePath = _a.valuePath, properties = _a.properties, pushData = _a.pushData;
        if (valuePath && properties && pushData) {
            var clonedProperties = cloneDeep(properties);
            set(clonedProperties, "controls." + valuePath, event.target.checked);
            pushData({ properties: clonedProperties });
        }
    };
    ConfigSubsection.prototype.getTestClassName = function () {
        return "s-configuration-subsection-" + this.props.title.replace(/\./g, "-");
    };
    ConfigSubsection.defaultProps = {
        collapsed: true,
        canBeToggled: false,
        toggleDisabled: false,
        toggledOn: true,
        pushData: noop,
        showDisabledMessage: false,
    };
    return ConfigSubsection;
}(React.Component));
export default injectIntl(ConfigSubsection);
