// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/bubbleChart/bubbleChartSeries.js
import { BucketNames } from "@gooddata/sdk-ui";
import { parseValue, unwrap } from "../_util/common";
import last from "lodash/last";
function getCountOfEmptyBuckets(bucketEmptyFlags) {
    if (bucketEmptyFlags === void 0) { bucketEmptyFlags = []; }
    return bucketEmptyFlags.filter(function (bucketEmpyFlag) { return bucketEmpyFlag; }).length;
}
export function getBubbleChartSeries(dv, measureGroup, 
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
stackByAttribute, colorStrategy) {
    var primaryMeasuresBucketEmpty = dv.def().isBucketEmpty(BucketNames.MEASURES);
    var secondaryMeasuresBucketEmpty = dv.def().isBucketEmpty(BucketNames.SECONDARY_MEASURES);
    var legendIndex = 0;
    var seriesIndex = 0;
    return dv
        .rawData()
        .twoDimData()
        .map(function (resData, index) {
        if (resData[0] === null || resData[1] === null || resData[2] === null) {
            seriesIndex++;
            return null;
        }
        var data = [];
        var emptyBucketsCount = getCountOfEmptyBuckets([
            primaryMeasuresBucketEmpty,
            secondaryMeasuresBucketEmpty,
        ]);
        data = [
            {
                x: !primaryMeasuresBucketEmpty ? parseValue(resData[0]) : 0,
                y: !secondaryMeasuresBucketEmpty ? parseValue(resData[1 - emptyBucketsCount]) : 0,
                // we want to allow NaN on z to be able show bubble of default size when Size bucket is empty
                z: parseFloat(resData[2 - emptyBucketsCount]),
                format: unwrap(last(measureGroup.items)).format,
            },
        ];
        return {
            name: stackByAttribute ? stackByAttribute.items[index].attributeHeaderItem.name : "",
            color: colorStrategy.getColorByIndex(legendIndex),
            legendIndex: legendIndex++,
            data: data,
            seriesIndex: seriesIndex++,
        };
    })
        .filter(function (serie) { return serie !== null; });
}
