// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/heatmap/heatmapColoring.js
import { __extends, __spreadArrays } from "tslib";
// (C) 2020-2021 GoodData Corporation
import { ColorStrategy, getColorByGuid, getColorFromMapping, getRgbStringFromRGB, isCustomPalette, normalizeColorToRGB, } from "@gooddata/sdk-ui-vis-commons";
import { isColorFromPalette, isRgbColor } from "@gooddata/sdk-model";
import { isDarkTheme } from "@gooddata/sdk-ui-theme-provider";
import { findMeasureGroupInDimensions } from "../_util/executionResultHelper";
import range from "lodash/range";
import isEqual from "lodash/isEqual";
import { DEFAULT_HEATMAP_BLUE_BASE_COLOR, DEFAULT_HEATMAP_BLUE_COLOR, HEATMAP_BLUE_COLOR_PALETTE, } from "../_util/color";
import { darken, mix, saturate } from "polished";
var HeatmapColorStrategy = /** @class */ (function (_super) {
    __extends(HeatmapColorStrategy, _super);
    function HeatmapColorStrategy() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HeatmapColorStrategy.prototype.getColorByIndex = function (index) {
        return this.palette[index % this.palette.length];
    };
    HeatmapColorStrategy.prototype.createColorAssignment = function (colorPalette, colorMapping, 
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    _viewByAttribute, 
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    _stackByAttribute, dv) {
        var mappedColor;
        var colorAssignment;
        var measureGroup = findMeasureGroupInDimensions(dv.meta().dimensions());
        var headerItem = measureGroup && measureGroup.items[0];
        if (colorMapping) {
            mappedColor = getColorFromMapping(headerItem, colorMapping, dv);
            if (mappedColor) {
                colorAssignment = [
                    {
                        headerItem: headerItem,
                        color: mappedColor,
                    },
                ];
            }
        }
        colorAssignment = colorAssignment || this.getDefaultColorAssignment(colorPalette, headerItem);
        return {
            fullColorAssignment: colorAssignment,
            outputColorAssignment: colorAssignment,
        };
    };
    HeatmapColorStrategy.prototype.getThemeBackgroundColor = function () {
        var _a, _b, _c, _d, _e, _f;
        return (_c = (_b = (_a = this.theme) === null || _a === void 0 ? void 0 : _a.chart) === null || _b === void 0 ? void 0 : _b.backgroundColor) !== null && _c !== void 0 ? _c : (_f = (_e = (_d = this.theme) === null || _d === void 0 ? void 0 : _d.palette) === null || _e === void 0 ? void 0 : _e.complementary) === null || _f === void 0 ? void 0 : _f.c0;
    };
    HeatmapColorStrategy.prototype.getBackgroundColor = function () {
        var _a;
        return (_a = this.getThemeBackgroundColor()) !== null && _a !== void 0 ? _a : "#fff";
    };
    HeatmapColorStrategy.prototype.createPalette = function (colorPalette, colorAssignment) {
        var colorAssignmentColor = colorAssignment[0].color;
        if (isRgbColor(colorAssignmentColor) &&
            isEqual(colorAssignmentColor.value, DEFAULT_HEATMAP_BLUE_COLOR)) {
            return normalizeColorToRGB(this.getBackgroundColor()) === "rgb(255,255,255)"
                ? HEATMAP_BLUE_COLOR_PALETTE
                : this.getCustomHeatmapColorPalette(DEFAULT_HEATMAP_BLUE_BASE_COLOR);
        }
        if (isColorFromPalette(colorAssignmentColor)) {
            var colorFromPalette = getColorByGuid(colorPalette, colorAssignmentColor.value, 0);
            return this.getCustomHeatmapColorPalette(colorFromPalette);
        }
        return this.getCustomHeatmapColorPalette(colorAssignmentColor.value);
    };
    HeatmapColorStrategy.prototype.getCustomHeatmapColorPalette = function (baseColorRGB) {
        var themeBackgroundColor = this.getThemeBackgroundColor();
        var backgroundColor = this.getBackgroundColor();
        var baseColor = getRgbStringFromRGB(baseColorRGB);
        var baseColorLast = saturate(0.16, darken(0.2, baseColor));
        if (themeBackgroundColor && !isDarkTheme(this.theme)) {
            return __spreadArrays(this.generatePalette(baseColor, backgroundColor, 5), this.generatePalette(baseColorLast, baseColor, 3).slice(1));
        }
        else {
            return this.generatePalette(baseColor, backgroundColor, 7);
        }
    };
    HeatmapColorStrategy.prototype.generatePalette = function (colorA, colorB, steps) {
        return range(steps).map(function (step) { return normalizeColorToRGB(mix((1 / (steps - 1)) * step, colorA, colorB)); });
    };
    HeatmapColorStrategy.prototype.getDefaultColorAssignment = function (colorPalette, headerItem) {
        var hasCustomPaletteWithColors = colorPalette && isCustomPalette(colorPalette) && colorPalette[0];
        if (hasCustomPaletteWithColors) {
            return [
                {
                    headerItem: headerItem,
                    color: {
                        type: "guid",
                        value: colorPalette[0].guid,
                    },
                },
            ];
        }
        return [
            {
                headerItem: headerItem,
                color: {
                    type: "rgb",
                    value: DEFAULT_HEATMAP_BLUE_COLOR,
                },
            },
        ];
    };
    return HeatmapColorStrategy;
}(ColorStrategy));
export { HeatmapColorStrategy };
