// @gooddata/sdk-ui-dashboard/esm/_staging/dashboard/fluidLayout/facade/layout.js
// (C) 2019-2021 GoodData Corporation
import { isDashboardLayout } from "@gooddata/sdk-backend-spi";
import invariant from "ts-invariant";
import { DashboardLayoutSectionFacade } from "./section";
import { DashboardLayoutSectionsFacade } from "./sections";
/**
 * @alpha
 */
var DashboardLayoutFacade = /** @class */ (function () {
    function DashboardLayoutFacade(layout) {
        this.layout = layout;
    }
    /**
     * Creates an instance of DashboardLayoutFacade
     * @param layout - layout to wrap
     */
    DashboardLayoutFacade.for = function (layout) {
        invariant(isDashboardLayout(layout), "Provided data must be IDashboardLayout.");
        return new DashboardLayoutFacade(layout);
    };
    DashboardLayoutFacade.prototype.sections = function () {
        return DashboardLayoutSectionsFacade.for(this, this.layout.sections);
    };
    DashboardLayoutFacade.prototype.section = function (rowIndex) {
        var row = this.layout.sections[rowIndex];
        if (!row) {
            return undefined;
        }
        return DashboardLayoutSectionFacade.for(this, row, rowIndex);
    };
    DashboardLayoutFacade.prototype.size = function () {
        return this.layout.size;
    };
    DashboardLayoutFacade.prototype.raw = function () {
        return this.layout;
    };
    return DashboardLayoutFacade;
}());
export { DashboardLayoutFacade };
