// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/headline/headlineBucketHelper.js
// (C) 2019-2020 GoodData Corporation
import cloneDeep from "lodash/cloneDeep";
import every from "lodash/every";
import { BucketNames } from "@gooddata/sdk-ui";
import { METRIC, BUCKETS } from "../../../constants/bucket";
import { isDerivedBucketItem, findDerivedBucketItems, findMasterBucketItem, findMasterBucketItems, } from "../../../utils/bucketHelper";
export function findSecondMasterMeasure(allMeasures) {
    var masterBucketItems = findMasterBucketItems(allMeasures);
    return masterBucketItems.length > 1 ? masterBucketItems[1] : null;
}
export function tryToMapForeignBuckets(extendedReferencePoint) {
    var newReferencePoint = setHeadlineRefPointBuckets(extendedReferencePoint);
    var totalBuckets = extendedReferencePoint.buckets.length;
    var allMeasuresCompatible = true;
    var bucketIndex = -1;
    while (allMeasuresCompatible && ++bucketIndex < totalBuckets) {
        var sourceBucket = extendedReferencePoint.buckets[bucketIndex];
        var targetBucket = newReferencePoint[BUCKETS][bucketIndex];
        if (!targetBucket) {
            if (sourceBucket.items.length) {
                allMeasuresCompatible = false;
            }
            continue;
        }
        var targetBucketUiConfig = newReferencePoint.uiConfig.buckets[targetBucket.localIdentifier];
        if (!targetBucketUiConfig || !targetBucketUiConfig.enabled) {
            allMeasuresCompatible = false;
            continue;
        }
        var measuresFitToLimit = sourceBucket.items.length <= targetBucketUiConfig.itemsLimit;
        if (!measuresFitToLimit) {
            allMeasuresCompatible = false;
            continue;
        }
        var isCompatibleMeasureType = every(sourceBucket.items, function (item) { return item.type === METRIC; });
        if (!isCompatibleMeasureType) {
            allMeasuresCompatible = false;
            continue;
        }
        targetBucket.items = sourceBucket.items;
    }
    return allMeasuresCompatible ? newReferencePoint : null;
}
export function setHeadlineRefPointBuckets(extendedReferencePoint, primaryMeasure, secondaryMeasure) {
    var newReferencePoint = cloneDeep(extendedReferencePoint);
    newReferencePoint[BUCKETS] = [
        {
            localIdentifier: BucketNames.MEASURES,
            items: primaryMeasure ? [primaryMeasure] : [],
        },
        {
            localIdentifier: BucketNames.SECONDARY_MEASURES,
            items: secondaryMeasure ? [secondaryMeasure] : [],
        },
    ];
    return newReferencePoint;
}
export function findComplementaryOverTimeComparisonMeasure(primaryMeasure, allMeasures) {
    if (!primaryMeasure) {
        return null;
    }
    if (isDerivedBucketItem(primaryMeasure)) {
        return findMasterBucketItem(primaryMeasure, allMeasures) || null;
    }
    var derivedOfPrimaryMeasure = findDerivedBucketItems(primaryMeasure, allMeasures);
    return derivedOfPrimaryMeasure.length > 0 ? derivedOfPrimaryMeasure[0] : null;
}
