// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/headline/HeadlineDescriptor.js
import { PluggableHeadline } from "./PluggableHeadline";
import { DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT, MAX_VISUALIZATION_HEIGHT } from "../constants";
var hasSecondaryMeasure = function (insight) {
    return insight.insight.buckets.filter(function (bucket) { return bucket.items.length > 0; }).length > 1;
};
var HeadlineDescriptor = /** @class */ (function () {
    function HeadlineDescriptor() {
    }
    HeadlineDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableHeadline(params); };
    };
    HeadlineDescriptor.prototype.getSizeInfo = function (insight, layoutDescriptor, settings) {
        return {
            width: {
                default: 2,
                min: 2,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(insight, settings.enableKDWidgetCustomHeight),
                min: this.getMinHeight(insight, settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(insight, settings.enableKDWidgetCustomHeight),
            },
        };
    };
    HeadlineDescriptor.prototype.getDefaultHeight = function (insight, enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT;
        }
        return hasSecondaryMeasure(insight) ? 11 : 8;
    };
    HeadlineDescriptor.prototype.getMinHeight = function (insight, enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT;
        }
        return hasSecondaryMeasure(insight) ? 10 : 6;
    };
    HeadlineDescriptor.prototype.getMaxHeight = function (_insight, enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_KPI_HEIGHT;
        }
        return MAX_VISUALIZATION_HEIGHT;
    };
    HeadlineDescriptor.prototype.applyDrillDown = function (insight) {
        return insight;
    };
    return HeadlineDescriptor;
}());
export { HeadlineDescriptor };
