// @gooddata/sdk-ui-dashboard/esm/presentation/widget/insightMenu/LegacyDefaultDashboardInsightMenu/getDefaultLegacyInsightMenuItems.js
/**
 * @internal
 */
export function getDefaultLegacyInsightMenuItems(intl, config) {
    var exportCSVDisabled = config.exportCSVDisabled, exportXLSXDisabled = config.exportXLSXDisabled, onExportCSV = config.onExportCSV, onExportXLSX = config.onExportXLSX, tooltipMessage = config.tooltipMessage;
    return [
        {
            type: "button",
            itemId: "ExportXLSXBubble",
            itemName: intl.formatMessage({ id: "options.menu.export.XLSX" }),
            disabled: exportXLSXDisabled,
            onClick: onExportXLSX,
            tooltip: tooltipMessage,
            className: "options-menu-export-xlsx s-options-menu-export-xlsx",
        },
        {
            type: "button",
            itemId: "ExportCSVBubble",
            itemName: intl.formatMessage({ id: "options.menu.export.CSV" }),
            disabled: exportCSVDisabled,
            onClick: onExportCSV,
            tooltip: tooltipMessage,
            className: "options-menu-export-csv s-options-menu-export-csv",
        },
    ];
}
