// @gooddata/sdk-backend-tiger/esm/backend/workspace/attributes/index.js
import { __awaiter, __generator } from "tslib";
// (C) 2019-2021 GoodData Corporation
import { NotSupported, } from "@gooddata/sdk-backend-spi";
import { areObjRefsEqual, isIdentifierRef } from "@gooddata/sdk-model";
import { TigerWorkspaceElements } from "./elements";
import { jsonApiHeaders, MetadataUtilities, JsonApiDatasetOutWithLinksTypeEnum, } from "@gooddata/api-client-tiger";
import flatMap from "lodash/flatMap";
import { invariant } from "ts-invariant";
import { convertAttributesWithSideloadedLabels, convertAttributeWithSideloadedLabels, convertLabelWithSideloadedAttribute, convertDatasetWithLinks, } from "../../../convertors/fromBackend/MetadataConverter";
var TigerWorkspaceAttributes = /** @class */ (function () {
    function TigerWorkspaceAttributes(authCall, workspace) {
        var _this = this;
        this.authCall = authCall;
        this.workspace = workspace;
        this.getAttributeDisplayForm = function (ref) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                        return [2 /*return*/, loadAttributeDisplayForm(client, this.workspace, ref)];
                    }); }); })];
            });
        }); };
        this.getAttribute = function (ref) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                        return [2 /*return*/, loadAttribute(client, this.workspace, ref)];
                    }); }); })];
            });
        }); };
    }
    TigerWorkspaceAttributes.prototype.elements = function () {
        return new TigerWorkspaceElements(this.authCall, this.workspace);
    };
    TigerWorkspaceAttributes.prototype.getAttributeDisplayForms = function (refs) {
        var _this = this;
        return this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
            var allAttributes;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, loadAttributes(client, this.workspace)];
                    case 1:
                        allAttributes = _a.sent();
                        return [2 /*return*/, flatMap(allAttributes, function (attr) { return attr.displayForms; }).filter(function (df) {
                                return refs.find(function (ref) { return areObjRefsEqual(ref, df.ref); });
                            })];
                }
            });
        }); });
    };
    TigerWorkspaceAttributes.prototype.getAttributes = function (refs) {
        var _this = this;
        return this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
            var allAttributes;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, loadAttributes(client, this.workspace)];
                    case 1:
                        allAttributes = _a.sent();
                        return [2 /*return*/, allAttributes.filter(function (attr) { return refs.find(function (ref) { return areObjRefsEqual(ref, attr.ref); }); })];
                }
            });
        }); });
    };
    TigerWorkspaceAttributes.prototype.getCommonAttributes = function () {
        throw new NotSupported("not supported");
    };
    TigerWorkspaceAttributes.prototype.getCommonAttributesBatch = function () {
        throw new NotSupported("not supported");
    };
    TigerWorkspaceAttributes.prototype.getAttributeDatasetMeta = function (ref) {
        var _this = this;
        return this.authCall(function (client) {
            return loadAttributeDataset(client, _this.workspace, ref);
        });
    };
    return TigerWorkspaceAttributes;
}());
export { TigerWorkspaceAttributes };
function loadAttributeDisplayForm(client, workspaceId, ref) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return client.workspaceObjects
        .getEntityLabels({
        workspaceId: workspaceId,
        objectId: ref.identifier,
    }, {
        headers: jsonApiHeaders,
        params: {
            include: "attributes",
        },
    })
        .then(function (res) { return convertLabelWithSideloadedAttribute(res.data); });
}
function loadAttribute(client, workspaceId, ref) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return client.workspaceObjects
        .getEntityAttributes({
        workspaceId: workspaceId,
        objectId: ref.identifier,
    }, {
        headers: jsonApiHeaders,
        params: {
            include: "labels",
        },
    })
        .then(function (res) { return convertAttributeWithSideloadedLabels(res.data); });
}
function loadAttributes(client, workspaceId) {
    return MetadataUtilities.getAllPagesOf(client, client.workspaceObjects.getAllEntitiesAttributes, { workspaceId: workspaceId }, { query: { include: "labels" } })
        .then(MetadataUtilities.mergeEntitiesResults)
        .then(convertAttributesWithSideloadedLabels);
}
function loadAttributeDataset(client, workspace, ref) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return client.workspaceObjects
        .getEntityAttributes({
        workspaceId: workspace,
        objectId: ref.identifier,
    }, {
        headers: jsonApiHeaders,
        params: {
            include: "datasets",
        },
    })
        .then(function (res) {
        // if this happens then its either bad query parameterization or the backend is hosed badly
        invariant(res.data.included && res.data.included.length > 0, "server returned that attribute does not belong to any dataset");
        var datasets = res.data.included.filter(function (include) {
            return include.type === JsonApiDatasetOutWithLinksTypeEnum.Dataset;
        });
        return convertDatasetWithLinks(datasets[0]);
    });
}
