// @gooddata/api-model-bear/esm/filterContext/GdcFilterContext.js
// (C) 2019-2021 GoodData Corporation
import isEmpty from "lodash/isEmpty";
/**
 * @public
 */
export var GdcFilterContext;
(function (GdcFilterContext) {
    function isDateFilter(filter) {
        return !isEmpty(filter) && !!filter.dateFilter;
    }
    GdcFilterContext.isDateFilter = isDateFilter;
    function isAttributeFilter(filter) {
        return !isEmpty(filter) && !!filter.attributeFilter;
    }
    GdcFilterContext.isAttributeFilter = isAttributeFilter;
    function isFilterContext(obj) {
        return !isEmpty(obj) && obj.meta.category === "filterContext";
    }
    GdcFilterContext.isFilterContext = isFilterContext;
    function isWrappedFilterContext(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && obj.hasOwnProperty("filterContext");
    }
    GdcFilterContext.isWrappedFilterContext = isWrappedFilterContext;
    function isTempFilterContext(obj) {
        return !!(!isEmpty(obj) &&
            obj.created &&
            obj.uri &&
            obj.filters.every(function (x) { return isDateFilter(x) || isAttributeFilter(x); }));
    }
    GdcFilterContext.isTempFilterContext = isTempFilterContext;
    function isWrappedTempFilterContext(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && obj.hasOwnProperty("tempFilterContext");
    }
    GdcFilterContext.isWrappedTempFilterContext = isWrappedTempFilterContext;
})(GdcFilterContext || (GdcFilterContext = {}));
