// @gooddata/api-model-bear/esm/filterContext/utils.js
import { __assign } from "tslib";
// (C) 2019-2020 GoodData Corporation
import omitBy from "lodash/omitBy";
import isUndefined from "lodash/isUndefined";
import { GdcFilterContext } from "./GdcFilterContext";
var isDateFilter = GdcFilterContext.isDateFilter;
var isAttributeFilter = GdcFilterContext.isAttributeFilter;
function sanitizeDateFilter(filter) {
    var _a = filter.dateFilter, from = _a.from, to = _a.to, type = _a.type, granularity = _a.granularity, dateFilter = filter.dateFilter;
    var sanitizedProperties = omitBy(dateFilter, isUndefined);
    return {
        dateFilter: __assign(__assign(__assign({ type: type,
            granularity: granularity }, sanitizedProperties), (from === undefined ? {} : { from: String(from) })), (to === undefined ? {} : { to: String(to) }))
    };
}
function sanitizeAttributeFilter(filter) {
    var _a = filter.attributeFilter, displayForm = _a.displayForm, negativeSelection = _a.negativeSelection, attributeElements = _a.attributeElements;
    return {
        attributeFilter: {
            displayForm: displayForm,
            negativeSelection: negativeSelection,
            attributeElements: attributeElements
        }
    };
}
/**
 * @public
 */
export function sanitizeFiltersForExport(filters) {
    return filters.map(function (filter) {
        if (isDateFilter(filter)) {
            return sanitizeDateFilter(filter);
        }
        if (isAttributeFilter(filter)) {
            return sanitizeAttributeFilter(filter);
        }
        return filter;
    });
}
