// @gooddata/sdk-ui-kit/esm/measureNumberFormat/customFormatDialog/FormatInput.js
import { __extends } from "tslib";
// (C) 2020 GoodData Corporation
import React from "react";
import { injectIntl } from "react-intl";
import { FormatTemplatesDropdown } from "./formatTemplatesDropdown/FormatTemplatesDropdown";
import { SyntaxHighlightingInput } from "../../syntaxHighlightingInput/SyntaxHighlightingInput";
var formattingRules = {
    start: [
        { regex: /"(?:[^\\]|\\.)*?"/, token: "string" },
        { regex: /(?:black|blue|cyan|green|magenta|red|yellow|white)\b/i, token: "keyword" },
        {
            regex: /(backgroundColor|color)(=)([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})/i,
            token: ["variable-4", null, "keyword"],
        },
        {
            regex: /(<|>|=|>=|<=)(-?)(\d*(\.|,)?\d+|Null)/i,
            token: ["variable-5", "variable-5", "variable-5"],
        },
        { regex: /\/\/.*/, token: "comment" },
        { regex: /\/(?:[^\\]|\\.)*?\//, token: "variable-3" },
        { regex: /\/\*/, token: "comment", next: "comment" },
        { regex: /[{[(]/, indent: true, token: "variable-brackets" },
        { regex: /[}\])]/, dedent: true, token: "variable-brackets" },
        { regex: /[a-z$][\w$]*/, token: "variable" },
        { regex: /<</, token: "meta", mode: { spec: "xml", end: />>/ } },
    ],
    comment: [
        { regex: /.*?\*\//, token: "comment", next: "start" },
        { regex: /.*/, token: "comment" },
    ],
    meta: {
        dontIndentStates: ["comment"],
        lineComment: "//",
    },
};
var FormatInput = /** @class */ (function (_super) {
    __extends(FormatInput, _super);
    function FormatInput() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleInputChange = function (value) {
            _this.props.onFormatChange(value);
        };
        return _this;
    }
    FormatInput.prototype.render = function () {
        var _a = this.props, format = _a.format, onFormatChange = _a.onFormatChange, separators = _a.separators, templates = _a.templates, intl = _a.intl;
        return (React.createElement("div", { className: "gd-measure-custom-format-dialog-section" },
            React.createElement("div", { className: "gd-measure-custom-format-dialog-section-title" },
                React.createElement("span", null, intl.formatMessage({ id: "measureNumberCustomFormatDialog.definition" })),
                templates && (React.createElement(FormatTemplatesDropdown, { onChange: onFormatChange, separators: separators, templates: templates }))),
            React.createElement(SyntaxHighlightingInput, { value: format, formatting: formattingRules, onChange: this.handleInputChange, className: "s-custom-format-input" })));
    };
    return FormatInput;
}(React.PureComponent));
var FormatInputWithIntl = injectIntl(FormatInput);
export default FormatInputWithIntl;
