// @gooddata/sdk-ui-geo/esm/core/geoChart/GeoChartRenderer.js
import { __assign, __extends } from "tslib";
// (C) 2007-2020 GoodData Corporation
import React from "react";
import cx from "classnames";
import isEqual from "lodash/isEqual";
import noop from "lodash/noop";
import mapboxgl from "mapbox-gl";
import invariant from "ts-invariant";
import { createClusterLabels, createClusterPoints, createPushpinDataLayer, createUnclusterPoints, createPushpinFilter, } from "./geoChartDataLayers";
import { createPushpinDataSource } from "./geoChartDataSource";
import { DEFAULT_CLUSTER_LABELS_CONFIG, DEFAULT_CLUSTER_LAYER_NAME, DEFAULT_DATA_SOURCE_NAME, DEFAULT_LAYER_NAME, DEFAULT_MAPBOX_OPTIONS, DEFAULT_TOOLTIP_OPTIONS, INTERACTION_EVENTS, LAYER_STYLE_LABEL_PREFIX, ZOOM_CONTROLS_HEIGHT, } from "./constants/geoChart";
import { handlePushpinMouseEnter, handlePushpinMouseLeave } from "./geoChartTooltip";
import { getViewportOptions } from "./helpers/geoChart/viewport";
import { isClusteringAllowed, isPointsConfigChanged, isColorAssignmentItemChanged, } from "./helpers/geoChart/common";
import { DataViewFacade } from "@gooddata/sdk-ui";
import { handleGeoPushpinDrillEvent } from "./helpers/geoChart/drilling";
var GeoChartRenderer = /** @class */ (function (_super) {
    __extends(GeoChartRenderer, _super);
    function GeoChartRenderer(props) {
        var _this = _super.call(this, props) || this;
        _this.setChartRef = function (ref) {
            _this.chartRef = ref;
        };
        _this.createMap = function () {
            var _a = _this.props, config = _a.config, geoData = _a.geoData;
            var data = geoData.location.data;
            var _b = (config || {}).isExportMode, isExportMode = _b === void 0 ? false : _b;
            var isViewportFrozen = _this.isViewportFrozen();
            _this.chart = new mapboxgl.Map(__assign(__assign(__assign({}, DEFAULT_MAPBOX_OPTIONS), getViewportOptions(data, config)), { container: _this.chartRef, 
                // If true, the map’s canvas can be exported to a PNG using map.getCanvas().toDataURL().
                // This is false by default as a performance optimization.
                interactive: !isViewportFrozen, preserveDrawingBuffer: isExportMode }));
        };
        _this.updateMapWithConfig = function (prevConfig, prevColorStrategy) {
            if (_this.shouldResetMap(prevConfig, prevColorStrategy)) {
                // Config for clustering and pushpin size lead to change layer setting
                // Then calling resetMap here is needed
                _this.resetMap();
            }
            _this.updatePanAndZoom();
            _this.updateViewport(prevConfig);
        };
        _this.resetMap = function () {
            _this.cleanupMap();
            _this.setupMap();
        };
        _this.shouldResetMap = function (prevConfig, prevColorStrategy) {
            var _a = _this.props, colorStrategy = _a.colorStrategy, config = _a.config;
            if (isPointsConfigChanged(prevConfig.points, config.points)) {
                return true;
            }
            if (isColorAssignmentItemChanged(prevColorStrategy.getColorAssignment(), colorStrategy.getColorAssignment())) {
                return true;
            }
            return false;
        };
        _this.isViewportFrozen = function () {
            var _a, _b;
            var config = _this.props.config;
            return (_b = (_a = config === null || config === void 0 ? void 0 : config.viewport) === null || _a === void 0 ? void 0 : _a.frozen) !== null && _b !== void 0 ? _b : false;
        };
        _this.removeMapControls = function () {
            if (_this.navigationControlButton) {
                _this.chart.removeControl(_this.navigationControlButton);
                _this.navigationControlButton = null;
            }
        };
        _this.addMapControls = function () {
            if (!_this.navigationControlButton) {
                _this.navigationControlButton = new mapboxgl.NavigationControl({
                    showCompass: false
                });
                _this.chart.addControl(_this.navigationControlButton, "top-right");
            }
        };
        _this.toggleMapControls = function () {
            var isViewportFrozen = _this.isViewportFrozen();
            if (!isViewportFrozen) {
                _this.addMapControls();
            }
            else {
                _this.removeMapControls();
            }
        };
        _this.toggleInteractionEvents = function () {
            var isViewportFrozen = _this.isViewportFrozen();
            var chart = _this.chart;
            if (chart === undefined) {
                return;
            }
            var action = isViewportFrozen ? "disable" : "enable";
            INTERACTION_EVENTS.forEach(function (interactionEvent) { return chart[interactionEvent][action](); });
        };
        _this.updatePanAndZoom = function () {
            _this.toggleMapControls();
            _this.toggleInteractionEvents();
        };
        _this.updateViewport = function (prevConfig) {
            invariant(_this.chart, "illegal state - updating viewport while map not initialized");
            var _a = _this.props, config = _a.config, geoData = _a.geoData;
            var data = geoData.location.data;
            var prevViewport = prevConfig.viewport;
            var viewport = config.viewport;
            if (isEqual(prevViewport, viewport)) {
                return;
            }
            var bounds = getViewportOptions(data, config).bounds;
            if (bounds) {
                _this.chart.fitBounds(bounds, DEFAULT_MAPBOX_OPTIONS.fitBoundsOptions);
            }
        };
        _this.setFilterMap = function () {
            invariant(_this.chart, "illegal state - setting filter while map not initialized");
            var _a = _this.props.config.selectedSegmentItems, selectedSegmentItems = _a === void 0 ? [] : _a;
            if (_this.chart.getLayer(DEFAULT_LAYER_NAME)) {
                _this.chart.setFilter(DEFAULT_LAYER_NAME, createPushpinFilter(selectedSegmentItems));
            }
        };
        _this.handleMapEvent = function () {
            var chart = _this.chart;
            invariant(chart, "illegal state - setting map event handlers while map not initialized");
            chart.on("click", DEFAULT_LAYER_NAME, _this.handleMapClick);
            chart.on("idle", _this.handleMapIdle);
            chart.on("load", _this.setupMap);
            chart.on("load", _this.adjustChartHeight);
            chart.on("mouseenter", DEFAULT_LAYER_NAME, _this.handlePushpinMouseEnter);
            chart.on("mouseleave", DEFAULT_LAYER_NAME, _this.handlePushpinMouseLeave);
            chart.on("moveend", _this.handlePushpinMoveEnd);
            chart.on("zoomend", _this.handlePushpinZoomEnd);
        };
        /*
        Fired after the last frame rendered before the map enters an "idle" state:
            - No camera transitions are in progress
            - All currently requested tiles have loaded
            - All fade/transition animations have completed
        This is called one time only
        */
        _this.handleMapIdle = function () {
            var _a = _this, chart = _a.chart, afterRender = _a.props.afterRender;
            if (!chart) {
                return;
            }
            chart.off("idle", _this.handleMapIdle);
            afterRender();
        };
        _this.setupMap = function () {
            var _a = _this, chart = _a.chart, handleLayerLoaded = _a.handleLayerLoaded, props = _a.props;
            invariant(chart, "illegal state - setting up map but with no existing map instance");
            var colorStrategy = props.colorStrategy, config = props.config, geoData = props.geoData;
            var _b = config || {}, _c = _b.points, _d = (_c === void 0 ? {} : _c).groupNearbyPoints, groupNearbyPoints = _d === void 0 ? true : _d, _e = _b.showLabels, showLabels = _e === void 0 ? true : _e;
            var hasClustering = isClusteringAllowed(geoData, groupNearbyPoints);
            var dataSourceProps = {
                colorStrategy: colorStrategy,
                config: config,
                geoData: geoData,
                hasClustering: hasClustering
            };
            chart.addSource(DEFAULT_DATA_SOURCE_NAME, createPushpinDataSource(dataSourceProps));
            if (!hasClustering) {
                chart.addLayer(createPushpinDataLayer(DEFAULT_DATA_SOURCE_NAME, geoData, config), "state-label");
            }
            else {
                chart.addLayer(createClusterPoints(DEFAULT_DATA_SOURCE_NAME));
                chart.addLayer(createClusterLabels(DEFAULT_DATA_SOURCE_NAME));
                // un-clustered points will be rendered under state/county label
                chart.addLayer(createUnclusterPoints(DEFAULT_DATA_SOURCE_NAME), "state-label");
            }
            // that config is not public,
            // we only use for storybook to make it is more stable
            if (!showLabels) {
                var _f = chart.getStyle().layers, layers = _f === void 0 ? [] : _f;
                layers.forEach(function (layer) {
                    if (layer.id.includes(LAYER_STYLE_LABEL_PREFIX)) {
                        _this.removeLayer(layer.id);
                    }
                });
            }
            // keep listening to the data event until the style is loaded
            chart.on("data", handleLayerLoaded);
        };
        _this.adjustChartHeight = function () {
            var _a, _b;
            var _c = _this, chart = _c.chart, chartRef = _c.chartRef;
            if (!chartRef || !chart) {
                return;
            }
            var chartHeight = chartRef.clientHeight;
            var parentHeight = (_b = (_a = chartRef.parentElement) === null || _a === void 0 ? void 0 : _a.clientHeight) !== null && _b !== void 0 ? _b : 0;
            var shouldResize = chartHeight <= ZOOM_CONTROLS_HEIGHT && ZOOM_CONTROLS_HEIGHT <= parentHeight;
            if (shouldResize) {
                // set min height to re-position mapbox attribution and zoom control, in case there are too many top legend items
                // that take all visible height of widget and make geo chart container's height zero
                chartRef.style.minHeight = parentHeight + "px";
                chart.resize();
            }
        };
        _this.handleLayerLoaded = function () {
            var chart = _this.chart;
            if (!chart || !chart.isStyleLoaded()) {
                return;
            }
            chart.off("data", _this.handleLayerLoaded);
        };
        _this.createTooltip = function () {
            _this.tooltip = new mapboxgl.Popup(DEFAULT_TOOLTIP_OPTIONS);
        };
        _this.cleanupMap = function () {
            if (!_this.chart) {
                return;
            }
            _this.removeLayer(DEFAULT_LAYER_NAME);
            _this.removeLayer(DEFAULT_CLUSTER_LAYER_NAME);
            _this.removeLayer(DEFAULT_CLUSTER_LABELS_CONFIG.id);
            if (_this.chart.getSource(DEFAULT_DATA_SOURCE_NAME)) {
                _this.chart.removeSource(DEFAULT_DATA_SOURCE_NAME);
            }
        };
        _this.removeMap = function () {
            if (!_this.chart) {
                return;
            }
            // try catch to hide the mapbox's error message
            // TypeError: Cannot read property 'off' of undefined
            // mapbox is trying to call its function after deleted
            // https://github.com/mapbox/mapbox-gl-js/blob/master/src/ui/control/navigation_control.js#L118
            try {
                _this.chart.remove();
            }
            catch (_a) {
                return;
            }
        };
        _this.handlePushpinMoveEnd = function (e) {
            var target = e.target;
            var onCenterPositionChanged = _this.props.onCenterPositionChanged;
            var _a = target.getCenter(), lng = _a.lng, lat = _a.lat;
            var center = { lng: lng, lat: lat };
            onCenterPositionChanged(center);
        };
        _this.handlePushpinZoomEnd = function (e) {
            var target = e.target;
            var onZoomChanged = _this.props.onZoomChanged;
            var zoom = target.getZoom();
            onZoomChanged(zoom);
        };
        _this.handleMapClick = function (e) {
            var _a = _this.props, viewport = _a.config.viewport, drillableItems = _a.drillableItems, drillConfig = _a.drillConfig, dataView = _a.dataView, geoData = _a.geoData;
            var features = e.features, originalEvent = e.originalEvent;
            var _b = features[0], coordinates = _b.geometry.coordinates, properties = _b.properties;
            // Disable drilling in edit/export mode
            if (viewport === null || viewport === void 0 ? void 0 : viewport.frozen) {
                return;
            }
            return handleGeoPushpinDrillEvent(drillableItems, drillConfig, DataViewFacade["for"](dataView), geoData, properties, coordinates, originalEvent.target);
        };
        _this.handlePushpinMouseEnter = function (e) {
            var _a = _this, chart = _a.chart, props = _a.props, tooltip = _a.tooltip;
            var config = props.config, drillableItems = props.drillableItems, intl = props.intl;
            return handlePushpinMouseEnter(e, chart, tooltip, config, drillableItems, intl);
        };
        _this.handlePushpinMouseLeave = function (e) {
            var _a = _this, chart = _a.chart, props = _a.props, tooltip = _a.tooltip;
            var config = props.config;
            return handlePushpinMouseLeave(e, chart, tooltip, config);
        };
        mapboxgl.accessToken = props.config.mapboxToken;
        _this.navigationControlButton = null;
        _this.chartRef = null;
        return _this;
    }
    GeoChartRenderer.prototype.componentDidUpdate = function (prevProps) {
        var _a = this.props, selectedSegmentItems = _a.config.selectedSegmentItems, colorStrategy = _a.colorStrategy;
        var prevConfig = prevProps.config, prevColorStrategy = prevProps.colorStrategy;
        var _b = (prevConfig || {}).selectedSegmentItems, prevSelectedSegmentItems = _b === void 0 ? [] : _b;
        if (!this.chart) {
            return;
        }
        // resize map when component is updated
        // for example: toggle legend, change position of legend
        this.chart.resize();
        // only update map when style is ready
        // work around for ticket SD-898
        // avoid refresh whole map will be fixed in ticket SD-899
        if (!this.chart.isStyleLoaded()) {
            return;
        }
        var isColorChanged = isColorAssignmentItemChanged(prevColorStrategy.getColorAssignment(), colorStrategy.getColorAssignment());
        var selectedSegmentItemsChanged = selectedSegmentItems && !isEqual(selectedSegmentItems, prevSelectedSegmentItems);
        if (!isColorChanged && selectedSegmentItemsChanged) {
            return this.setFilterMap();
        }
        this.updateMapWithConfig(prevConfig, prevColorStrategy);
    };
    GeoChartRenderer.prototype.componentDidMount = function () {
        this.createTooltip();
        this.createMap();
        this.createMapControls();
        this.handleMapEvent();
    };
    GeoChartRenderer.prototype.componentWillUnmount = function () {
        this.removeMap();
    };
    GeoChartRenderer.prototype.render = function () {
        var _a = this.props.config.isExportMode, isExportMode = _a === void 0 ? false : _a;
        var classNames = cx("s-gd-geo-chart-renderer", "mapbox-container", {
            isExportMode: isExportMode,
            "s-isExportMode": isExportMode
        });
        return React.createElement("div", { className: classNames, ref: this.setChartRef });
    };
    GeoChartRenderer.prototype.createMapControls = function () {
        invariant(this.chart, "illegal state - creating map controls while map not initialized");
        var isViewportFrozen = this.isViewportFrozen();
        this.chart.addControl(new mapboxgl.AttributionControl({
            compact: true
        }));
        if (!isViewportFrozen) {
            this.addMapControls();
        }
    };
    GeoChartRenderer.prototype.removeLayer = function (layerName) {
        if (!this.chart) {
            return;
        }
        if (this.chart.getLayer(layerName)) {
            this.chart.removeLayer(layerName);
        }
    };
    GeoChartRenderer.defaultProps = {
        config: {
            mapboxToken: ""
        },
        afterRender: noop,
        onZoomChanged: noop,
        onCenterPositionChanged: noop
    };
    return GeoChartRenderer;
}(React.Component));
export default GeoChartRenderer;
