// @gooddata/sdk-ui-dashboard/esm/model/types/layoutTypes.js
// (C) 2021 GoodData Corporation
import { __assign } from "tslib";
import { isWidget, } from "@gooddata/sdk-backend-spi";
import { idRef } from "@gooddata/sdk-model";
import cloneDeep from "lodash/cloneDeep";
import isEmpty from "lodash/isEmpty";
/**
 * Creates a new custom widget.
 *
 * @param identifier - identifier for custom widget; once added onto a dashboard, widget will be referencable using this identifier
 * @param customType - custom widget type
 * @param extras - optionally provide extra data to include on the custom widget; the content of this argument can be an
 *  arbitrary plain object. note: the factory will make a copy of all the extra data. at this moment it is not possible
 *  to modify the data once the widget is added onto a dashboard.
 * @public
 */
export function newCustomWidget(identifier, customType, extras) {
    var extrasCopy = extras ? cloneDeep(extras) : {};
    return __assign({ type: "customWidget", customType: customType,
        identifier: identifier, uri: "_custom_widget_uri/" + identifier, ref: idRef(identifier) }, extrasCopy);
}
/**
 * Type-guard that tests whether an object is an instance of {@link ICustomWidget}.
 *
 * @param obj - object to test
 * @public
 */
export function isCustomWidget(obj) {
    var w = obj;
    return !isEmpty(w) && w.type === "customWidget" && w.customType !== undefined && w.ref !== undefined;
}
/**
 * Type-guard that tests whether an object is an instance of {@link ICustomWidgetDefinition}.
 *
 * @param obj - object to test
 * @public
 */
export function isCustomWidgetDefinition(obj) {
    var w = obj;
    return !isEmpty(w) && w.type === "customWidget" && w.customType !== undefined && w.ref === undefined;
}
/**
 * Dumps debug information about a widget into a string.
 *
 * @param widget - widget to dump info from
 * @internal
 */
export function extendedWidgetDebugStr(widget) {
    var widgetId = "" + widget.identifier;
    var widgetType = "unknown widget type";
    if (isWidget(widget)) {
        widgetType = widget.type;
    }
    else if (isCustomWidget(widget)) {
        widgetType = widget.type + "/" + widget.customType;
    }
    return widgetId + "(" + widgetType + ")";
}
/**
 * Creates a new dashboard item containing the provided custom widget.
 *
 * @param widget - custom widget to include
 * @param sizeOrColSize - item size specification; for convenience you can specify the size as number which will be
 *  interpreted as number of columns in a 12-col grid that the item should use when rendered on an XL screen.
 * @public
 */
export function newDashboardItem(widget, sizeOrColSize) {
    var size = typeof sizeOrColSize === "number" ? { xl: { gridWidth: sizeOrColSize } } : sizeOrColSize;
    return {
        type: "IDashboardLayoutItem",
        size: size,
        widget: cloneDeep(widget),
    };
}
function getOrCreateSectionHeader(titleOrHeader) {
    if (!titleOrHeader) {
        return undefined;
    }
    if (typeof titleOrHeader === "string") {
        if (isEmpty(titleOrHeader)) {
            return undefined;
        }
        return {
            title: titleOrHeader,
        };
    }
    return titleOrHeader;
}
/**
 * Creates a new dashboard section.
 *
 * @param titleOrHeader - header to use for this section (if any); for convenience, you may provide just string containing the title instead
 * of specifying full header. if you specify empty string for title, then there will be no header.
 * @param items - dashboard items to include in the section; note: a deep copy of each item will be used on the new section
 *
 * @public
 */
export function newDashboardSection(titleOrHeader) {
    var items = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        items[_i - 1] = arguments[_i];
    }
    var header = getOrCreateSectionHeader(titleOrHeader);
    var itemsClone = cloneDeep(items);
    return {
        type: "IDashboardLayoutSection",
        items: itemsClone,
        header: header,
    };
}
/**
 * Tests whether object is an instance of {@link StashedDashboardItemsId};
 *
 * @param obj - object to test
 * @alpha
 */
export function isStashedDashboardItemsId(obj) {
    return typeof obj === "string";
}
