// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/BigChartDescriptor.js
// (C) 2021 GoodData Corporation
import { __extends } from "tslib";
import { MIDDLE_VISUALIZATION_HEIGHT, DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT } from "./constants";
import { BaseChartDescriptor } from "./baseChart/BaseChartDescriptor";
var BigChartDescriptor = /** @class */ (function (_super) {
    __extends(BigChartDescriptor, _super);
    function BigChartDescriptor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BigChartDescriptor.prototype.getSizeInfo = function (_insight, layoutDescriptor, settings) {
        return {
            width: {
                default: 6,
                min: 4,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getMinHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    };
    BigChartDescriptor.prototype.getMinHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MIDDLE_VISUALIZATION_HEIGHT;
    };
    return BigChartDescriptor;
}(BaseChartDescriptor));
export { BigChartDescriptor };
