// @gooddata/sdk-backend-tiger/esm/backend/workspace/settings/index.js
import { __assign, __awaiter, __generator } from "tslib";
import { DefaultUiSettings, DefaultUserSettings } from "../../uiSettings";
var TigerWorkspaceSettings = /** @class */ (function () {
    function TigerWorkspaceSettings(authCall, workspace) {
        this.authCall = authCall;
        this.workspace = workspace;
    }
    TigerWorkspaceSettings.prototype.getSettings = function () {
        var _this = this;
        return this.authCall(function (_client) { return __awaiter(_this, void 0, void 0, function () {
            var config;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, _client.organizationObjects.getEntityWorkspaces({
                            id: this.workspace,
                            metaInclude: ["config"],
                        })];
                    case 1:
                        config = (_a.sent()).data.data.meta;
                        return [2 /*return*/, __assign(__assign({ workspace: this.workspace }, DefaultUiSettings), config === null || config === void 0 ? void 0 : config.config)];
                }
            });
        }); });
    };
    TigerWorkspaceSettings.prototype.getSettingsForCurrentUser = function () {
        var _this = this;
        return this.authCall(function (_client, ctx) { return __awaiter(_this, void 0, void 0, function () {
            var principal, config;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ctx.getPrincipal()];
                    case 1:
                        principal = _a.sent();
                        return [4 /*yield*/, _client.organizationObjects.getEntityWorkspaces({
                                id: this.workspace,
                                metaInclude: ["config"],
                            })];
                    case 2:
                        config = (_a.sent()).data.data.meta;
                        return [2 /*return*/, __assign(__assign(__assign({}, DefaultUserSettings), { userId: principal.userId, workspace: this.workspace }), config === null || config === void 0 ? void 0 : config.config)];
                }
            });
        }); });
    };
    return TigerWorkspaceSettings;
}());
export { TigerWorkspaceSettings };
