// @gooddata/sdk-ui-ext/esm/internal/constants/axis.js
var _a, _b;
import { __assign } from "tslib";
// (C) 2019-2020 GoodData Corporation
import { VisualizationTypes } from "@gooddata/sdk-ui";
export var AXIS = {
    PRIMARY: "primary",
    SECONDARY: "secondary",
    DUAL: "dual",
};
export var AXIS_NAME;
(function (AXIS_NAME) {
    AXIS_NAME["X"] = "xaxis";
    AXIS_NAME["Y"] = "yaxis";
    AXIS_NAME["SECONDARY_X"] = "secondary_xaxis";
    AXIS_NAME["SECONDARY_Y"] = "secondary_yaxis";
})(AXIS_NAME || (AXIS_NAME = {}));
export var DUAL_AXES_SUPPORTED_CHARTS = [
    VisualizationTypes.COLUMN,
    VisualizationTypes.BAR,
    VisualizationTypes.LINE,
    VisualizationTypes.COMBO,
];
var BASE_X_AXIS = {
    name: AXIS_NAME.X,
    title: "properties.xaxis.title",
    subtitle: "",
    primary: false,
};
var BASE_Y_AXIS = {
    name: AXIS_NAME.Y,
    title: "properties.yaxis.title",
    subtitle: "",
    primary: true,
};
var BASE_SECONDARY_Y_AXIS = __assign(__assign({}, BASE_Y_AXIS), { name: AXIS_NAME.SECONDARY_Y });
var BAR_X_AXIS = __assign(__assign({}, BASE_X_AXIS), { primary: true });
var BAR_SECONDARY_X_AXIS = __assign(__assign({}, BASE_X_AXIS), { name: AXIS_NAME.SECONDARY_X, primary: true });
var BAR_Y_AXIS = __assign(__assign({}, BASE_Y_AXIS), { primary: false });
export var BASE_CHART_AXIS_CONFIG = (_a = {},
    _a[AXIS.PRIMARY] = [BASE_X_AXIS, BASE_Y_AXIS],
    _a[AXIS.SECONDARY] = [BASE_X_AXIS, BASE_SECONDARY_Y_AXIS],
    _a[AXIS.DUAL] = [
        BASE_X_AXIS,
        __assign(__assign({}, BASE_Y_AXIS), { subtitle: "properties.axis.left" }),
        __assign(__assign({}, BASE_SECONDARY_Y_AXIS), { subtitle: "properties.axis.right" }),
    ],
    _a);
export var BAR_CHART_AXIS_CONFIG = (_b = {},
    _b[AXIS.PRIMARY] = [BAR_X_AXIS, BAR_Y_AXIS],
    _b[AXIS.SECONDARY] = [BAR_SECONDARY_X_AXIS, BAR_Y_AXIS],
    _b[AXIS.DUAL] = [
        __assign(__assign({}, BAR_SECONDARY_X_AXIS), { subtitle: "properties.axis.top" }),
        __assign(__assign({}, BAR_X_AXIS), { subtitle: "properties.axis.bottom" }),
        BAR_Y_AXIS,
    ],
    _b);
