// @gooddata/sdk-ui-ext/esm/internal/constants/supportedProperties.js
var _a, _b, _c, _d;
import { __spreadArrays } from "tslib";
// (C) 2019-2021 GoodData Corporation
import { AXIS } from "./axis";
var BASE_PROPERTIES = ["dataLabels.visible", "grid", "legend", "colorMapping", "zoomInsight"];
var BASE_X_AXIS_PROPERTIES = [
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.name.visible",
    "xaxis.name.position",
];
var BASE_Y_AXIS_PROPERTIES = [
    "yaxis.rotation",
    "yaxis.labelsEnabled",
    "yaxis.visible",
    "yaxis.name.visible",
    "yaxis.name.position",
];
var BASE_PRIMARY_AXIS_PROPERTIES = __spreadArrays(BASE_Y_AXIS_PROPERTIES, ["yaxis.min", "yaxis.max", "yaxis.format"]);
var BAR_PRIMARY_AXIS_PROPERTIES = __spreadArrays(BASE_X_AXIS_PROPERTIES, ["xaxis.min", "xaxis.max", "xaxis.format"]);
var BASE_SECONDARY_AXIS_PROPERTIES = [
    "secondary_yaxis.rotation",
    "secondary_yaxis.labelsEnabled",
    "secondary_yaxis.visible",
    "secondary_yaxis.min",
    "secondary_yaxis.max",
    "secondary_yaxis.measures",
    "secondary_yaxis.name.visible",
    "secondary_yaxis.name.position",
    "secondary_yaxis.format",
];
var BAR_SECONDARY_AXIS_PROPERTIES = [
    "secondary_xaxis.rotation",
    "secondary_xaxis.labelsEnabled",
    "secondary_xaxis.visible",
    "secondary_xaxis.min",
    "secondary_xaxis.max",
    "secondary_xaxis.measures",
    "secondary_xaxis.name.visible",
    "secondary_xaxis.name.position",
    "secondary_xaxis.format",
];
export var OPTIONAL_STACKING_PROPERTIES = ["stackMeasures", "stackMeasuresToPercent"];
var CHART_TYPE_PROPERTIES = ["primaryChartType", "secondaryChartType", "dualAxis"];
export var BASE_CHART_SUPPORTED_PROPERTIES = __spreadArrays(BASE_PROPERTIES, BASE_X_AXIS_PROPERTIES, BASE_PRIMARY_AXIS_PROPERTIES);
export var AREA_CHART_SUPPORTED_PROPERTIES = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, OPTIONAL_STACKING_PROPERTIES, [
    "dataPoints.visible",
]);
export var COLUMN_CHART_SUPPORTED_PROPERTIES = (_a = {},
    _a[AXIS.PRIMARY] = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, OPTIONAL_STACKING_PROPERTIES),
    _a[AXIS.SECONDARY] = __spreadArrays(BASE_PROPERTIES, BASE_X_AXIS_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, OPTIONAL_STACKING_PROPERTIES),
    _a[AXIS.DUAL] = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, OPTIONAL_STACKING_PROPERTIES),
    _a);
export var LINE_CHART_SUPPORTED_PROPERTIES = (_b = {},
    _b[AXIS.PRIMARY] = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, ["dataPoints.visible"]),
    _b[AXIS.SECONDARY] = __spreadArrays(BASE_PROPERTIES, BASE_X_AXIS_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, [
        "dataPoints.visible",
    ]),
    _b[AXIS.DUAL] = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, [
        "dataPoints.visible",
    ]),
    _b);
export var BAR_CHART_SUPPORTED_PROPERTIES = (_c = {},
    _c[AXIS.PRIMARY] = __spreadArrays(BASE_PROPERTIES, BASE_Y_AXIS_PROPERTIES, BAR_PRIMARY_AXIS_PROPERTIES, OPTIONAL_STACKING_PROPERTIES),
    _c[AXIS.SECONDARY] = __spreadArrays(BASE_PROPERTIES, BASE_Y_AXIS_PROPERTIES, BAR_SECONDARY_AXIS_PROPERTIES, OPTIONAL_STACKING_PROPERTIES),
    _c[AXIS.DUAL] = __spreadArrays(BASE_PROPERTIES, BASE_Y_AXIS_PROPERTIES, BAR_PRIMARY_AXIS_PROPERTIES, BAR_SECONDARY_AXIS_PROPERTIES, OPTIONAL_STACKING_PROPERTIES),
    _c);
export var COMBO_CHART_SUPPORTED_PROPERTIES = (_d = {},
    _d[AXIS.PRIMARY] = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, CHART_TYPE_PROPERTIES, OPTIONAL_STACKING_PROPERTIES, [
        "dataPoints.visible",
    ]),
    _d[AXIS.SECONDARY] = __spreadArrays(BASE_PROPERTIES, BASE_X_AXIS_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, CHART_TYPE_PROPERTIES, OPTIONAL_STACKING_PROPERTIES, [
        "dataPoints.visible",
    ]),
    _d[AXIS.DUAL] = __spreadArrays(BASE_CHART_SUPPORTED_PROPERTIES, BASE_SECONDARY_AXIS_PROPERTIES, CHART_TYPE_PROPERTIES, OPTIONAL_STACKING_PROPERTIES, [
        "dataPoints.visible",
    ]),
    _d);
export var BUBBLE_CHART_SUPPORTED_PROPERTIES = __spreadArrays(BASE_PROPERTIES, [
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.min",
    "xaxis.max",
    "xaxis.name.visible",
    "xaxis.name.position",
    "xaxis.format",
    "yaxis.labelsEnabled",
    "yaxis.rotation",
    "yaxis.visible",
    "yaxis.min",
    "yaxis.max",
    "yaxis.name.visible",
    "yaxis.name.position",
    "yaxis.format",
]);
export var SCATTERPLOT_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "grid",
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.min",
    "xaxis.max",
    "xaxis.name.visible",
    "xaxis.name.position",
    "xaxis.format",
    "yaxis.labelsEnabled",
    "yaxis.rotation",
    "yaxis.visible",
    "yaxis.min",
    "yaxis.max",
    "yaxis.name.visible",
    "yaxis.name.position",
    "colorMapping",
    "zoomInsight",
    "yaxis.format",
];
export var PIECHART_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "legend",
    "colorMapping",
    "chart.verticalAlign",
];
export var TREEMAP_SUPPORTED_PROPERTIES = ["dataLabels.visible", "legend", "colorMapping", "zoomInsight"];
export var HEATMAP_SUPPORTED_PROPERTIES = [
    "dataLabels.visible",
    "legend",
    "yaxis.rotation",
    "yaxis.labelsEnabled",
    "yaxis.visible",
    "yaxis.name.visible",
    "yaxis.name.position",
    "xaxis.labelsEnabled",
    "xaxis.rotation",
    "xaxis.visible",
    "xaxis.name.visible",
    "xaxis.name.position",
    "colorMapping",
    "zoomInsight",
];
export var BULLET_CHART_SUPPORTED_PROPERTIES = [
    "grid",
    "xaxis.rotation",
    "xaxis.labelsEnabled",
    "xaxis.visible",
    "xaxis.min",
    "xaxis.max",
    "xaxis.name.visible",
    "xaxis.name.position",
    "xaxis.format",
    "yaxis.rotation",
    "yaxis.labelsEnabled",
    "yaxis.visible",
    "yaxis.name.visible",
    "yaxis.name.position",
    "legend",
    "colorMapping",
    "zoomInsight",
];
export var GEOPUSHPIN_SUPPORTED_PROPERTIES = [
    "tooltipText",
    "legend",
    "points.groupNearbyPoints",
    "points.minSize",
    "points.maxSize",
    "viewport.area",
    "colorMapping",
];
export var PIVOT_TABLE_SUPPORTED_PROPERTIES = ["columnWidths"];
