// @gooddata/sdk-ui-dashboard/esm/model/store/meta/metaSelectors.js
import { __assign, __rest } from "tslib";
// (C) 2021 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import { idRef, uriRef } from "@gooddata/sdk-model";
import { isTempFilterContext, } from "@gooddata/sdk-backend-spi";
import invariant from "ts-invariant";
var selectSelf = createSelector(function (state) { return state; }, function (state) { return state.meta; });
/**
 * Selects dashboard's descriptor.
 *
 * @internal
 */
export var selectDashboardDescriptor = createSelector(selectSelf, function (state) {
    invariant(state.descriptor, "attempting to access uninitialized meta state");
    return state.descriptor;
});
/**
 * Selects persisted IDashboard object - that is the IDashboard object that was used to initialize the rest
 * of the dashboard state of the dashboard component during the initial load of the dashboard.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 *
 * @internal
 */
export var selectPersistedDashboard = createSelector(selectSelf, function (state) {
    return state.persistedDashboard;
});
/**
 * Selects persisted IFilterContext/ITempFilterContext - that is the IFilterContext or ITempFilterContext that
 * was used to initialize the original filters of the dashboard component during the initial load of the
 * dashboard.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 */
export var selectPersistedDashboardFilterContext = createSelector(selectSelf, function (state) {
    var _a;
    return (_a = state.persistedDashboard) === null || _a === void 0 ? void 0 : _a.filterContext;
});
/**
 * Selects persisted IFilterContextDefinition - that is the IFilterContext or ITempFilterContext that
 * was used to initialize the original filters of the dashboard component during the initial load of the
 * dashboard but removes ref, uri and identifier, effectively creating a clone of the stored value
 * that can be used independently.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 */
export var selectPersistedDashboardFilterContextAsFilterContextDefinition = createSelector(selectPersistedDashboardFilterContext, function (filterContext) {
    if (!filterContext) {
        return undefined;
    }
    if (isTempFilterContext(filterContext)) {
        var _1 = filterContext.ref, __ = filterContext.uri, definition = __rest(filterContext, ["ref", "uri"]);
        return __assign(__assign({}, definition), { title: "filterContext", description: "" });
    }
    else {
        var _2 = filterContext.identifier, __ = filterContext.ref, ___ = filterContext.uri, definition = __rest(filterContext, ["identifier", "ref", "uri"]);
        return definition;
    }
});
/**
 * Selects ref of the persisted dashboard object that backs and is rendered-by the dashboard component.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 *
 * @public
 */
export var selectDashboardRef = createSelector(selectPersistedDashboard, function (state) {
    return state === null || state === void 0 ? void 0 : state.ref;
});
/**
 * Selects identifier of the persisted dashboard object that backs and is rendered-by the dashboard component.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 *
 * @public
 */
export var selectDashboardId = createSelector(selectPersistedDashboard, function (state) {
    return state === null || state === void 0 ? void 0 : state.identifier;
});
/**
 * Selects URI of the persisted dashboard object that backs and is rendered-by the dashboard component.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 *
 * @public
 */
export var selectDashboardUri = createSelector(selectPersistedDashboard, function (state) {
    return state === null || state === void 0 ? void 0 : state.uri;
});
/**
 * Selects idRef of the persisted dashboard object that backs and is rendered-by the dashboard component.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 *
 * @public
 */
export var selectDashboardIdRef = createSelector(selectDashboardId, function (id) {
    return id ? idRef(id, "analyticalDashboard") : undefined;
});
/**
 * Selects uriRef of the persisted dashboard object that backs and is rendered-by the dashboard component.
 *
 * Note that this may be undefined when the dashboard component works with a dashboard that has not yet
 * been persisted (typically newly created dashboard being edited).
 *
 * @public
 */
export var selectDashboardUriRef = createSelector(selectDashboardUri, function (uri) {
    return uri ? uriRef(uri) : undefined;
});
//
//
//
/**
 * Selects current dashboard title.
 *
 * @public
 */
export var selectDashboardTitle = createSelector(selectDashboardDescriptor, function (state) {
    return state.title;
});
/**
 * Selects current dashboard description.
 *
 * @public
 */
export var selectDashboardDescription = createSelector(selectDashboardDescriptor, function (state) {
    return state.description;
});
/**
 * Selects dashboard tags.
 *
 * @public
 */
export var selectDashboardTags = createSelector(selectDashboardDescriptor, function (state) {
    return state.tags;
});
/**
 * Selects dashboard share status.
 *
 * @alpha
 */
export var selectDashboardShareStatus = createSelector(selectDashboardDescriptor, function (state) {
    return state.shareStatus;
});
/**
 * Selects dashboard lock status.
 *
 * @alpha
 */
export var selectDashboardLockStatus = createSelector(selectDashboardDescriptor, function (state) {
    return state.isLocked;
});
/**
 * Selects complete dashboard share info.
 *
 * @alpha
 */
export var selectDashboardShareInfo = createSelector(selectDashboardDescriptor, function (_a) {
    var shareStatus = _a.shareStatus, isLocked = _a.isLocked;
    return ({
        shareStatus: shareStatus,
        isLocked: isLocked,
    });
});
