// @gooddata/sdk-ui-dashboard/esm/model/commandHandlers/layout/addSectionItemsHandler.js
import { __generator } from "tslib";
import { invalidArgumentsProvided } from "../../events/general";
import { selectLayout, selectStash } from "../../store/layout/layoutSelectors";
import { call, put, select } from "redux-saga/effects";
import { validateItemPlacement, validateSectionExists } from "./validation/layoutValidation";
import { validateAndResolveStashedItems } from "./validation/stashValidation";
import isEmpty from "lodash/isEmpty";
import { layoutActions } from "../../store/layout";
import { layoutSectionItemsAdded } from "../../events/layout";
import { resolveIndexOfNewItem } from "../../utils/arrayOps";
import { selectInsightsMap } from "../../store/insights/insightsSelectors";
import { batchActions } from "redux-batched-actions";
import { insightsActions } from "../../store/insights";
import { validateAndNormalizeWidgetItems, validateAndResolveItemFilterSettings, } from "./validation/itemValidation";
import { addTemporaryIdentityToWidgets } from "../../utils/dashboardItemUtils";
function validateAndResolveItems(commandCtx) {
    var ctx = commandCtx.ctx, layout = commandCtx.layout, stash = commandCtx.stash, items = commandCtx.items, _a = commandCtx.cmd.payload, sectionIndex = _a.sectionIndex, itemIndex = _a.itemIndex;
    if (!validateSectionExists(layout, sectionIndex)) {
        throw invalidArgumentsProvided(ctx, commandCtx.cmd, "Attempting to add items to non-existing layout section at index " + sectionIndex + ".");
    }
    var section = layout.sections[sectionIndex];
    if (!validateItemPlacement(section, itemIndex)) {
        throw invalidArgumentsProvided(ctx, commandCtx.cmd, "Attempting to insert new item at wrong index " + itemIndex + ". There are currently " + section.items.length + " items.");
    }
    var stashValidationResult = validateAndResolveStashedItems(stash, items);
    if (!isEmpty(stashValidationResult.missing)) {
        throw invalidArgumentsProvided(ctx, commandCtx.cmd, "Attempting to use non-existing stashes. Identifiers of missing stashes: " + stashValidationResult.missing.join(", "));
    }
    return {
        stashValidationResult: stashValidationResult,
        section: section,
    };
}
export function addSectionItemsHandler(ctx, cmd) {
    var items, commandCtx, _a, _b, stashValidationResult, section, _c, itemIndex, sectionIndex, autoResolveDateFilterDataset, normalizationResult, itemsToAdd;
    return __generator(this, function (_d) {
        switch (_d.label) {
            case 0:
                items = cmd.payload.items;
                _a = {
                    ctx: ctx,
                    cmd: cmd,
                    items: addTemporaryIdentityToWidgets(items)
                };
                return [4 /*yield*/, select(selectLayout)];
            case 1:
                _a.layout = _d.sent();
                return [4 /*yield*/, select(selectStash)];
            case 2:
                _a.stash = _d.sent();
                return [4 /*yield*/, select(selectInsightsMap)];
            case 3:
                commandCtx = (_a.availableInsights = _d.sent(),
                    _a);
                _b = validateAndResolveItems(commandCtx), stashValidationResult = _b.stashValidationResult, section = _b.section;
                _c = cmd.payload, itemIndex = _c.itemIndex, sectionIndex = _c.sectionIndex, autoResolveDateFilterDataset = _c.autoResolveDateFilterDataset;
                return [4 /*yield*/, call(validateAndNormalizeWidgetItems, ctx, stashValidationResult, cmd)];
            case 4:
                normalizationResult = _d.sent();
                return [4 /*yield*/, call(validateAndResolveItemFilterSettings, ctx, cmd, normalizationResult, autoResolveDateFilterDataset)];
            case 5:
                itemsToAdd = _d.sent();
                return [4 /*yield*/, put(batchActions([
                        insightsActions.addInsights(normalizationResult.resolvedInsights.loaded),
                        layoutActions.addSectionItems({
                            sectionIndex: sectionIndex,
                            itemIndex: itemIndex,
                            items: itemsToAdd,
                            usedStashes: stashValidationResult.existing,
                            undo: {
                                cmd: cmd,
                            },
                        }),
                    ]))];
            case 6:
                _d.sent();
                return [2 /*return*/, layoutSectionItemsAdded(ctx, sectionIndex, resolveIndexOfNewItem(section.items, itemIndex), stashValidationResult.resolved, stashValidationResult.existing, cmd.correlationId)];
        }
    });
}
