// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/geoChart/GeoPushpinChartDescriptor.js
// (C) 2021 GoodData Corporation
import { __extends } from "tslib";
import { PluggableGeoPushpinChart } from "./PluggableGeoPushpinChart";
import { DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT, MIDDLE_VISUALIZATION_HEIGHT } from "../constants";
import { BaseChartDescriptor } from "../baseChart/BaseChartDescriptor";
var GeoPushpinChartDescriptor = /** @class */ (function (_super) {
    __extends(GeoPushpinChartDescriptor, _super);
    function GeoPushpinChartDescriptor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    GeoPushpinChartDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableGeoPushpinChart(params); };
    };
    GeoPushpinChartDescriptor.prototype.getSizeInfo = function (_insight, layoutDescriptor, settings) {
        return {
            width: {
                default: 6,
                min: 6,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    };
    GeoPushpinChartDescriptor.prototype.getMinHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MIDDLE_VISUALIZATION_HEIGHT;
    };
    return GeoPushpinChartDescriptor;
}(BaseChartDescriptor));
export { GeoPushpinChartDescriptor };
