// @gooddata/sdk-backend-bear/esm/convertors/toBackend/afm/MeasureConverter.js
import { __assign } from "tslib";
// (C) 2007-2021 GoodData Corporation
import { isArithmeticMeasureDefinition, isMeasureDefinition, isPoPMeasureDefinition, isPreviousPeriodMeasureDefinition, } from "@gooddata/sdk-model";
import { convertMeasureFilter } from "./FilterConverter";
import { toBearRef } from "../ObjRefConverter";
import compact from "lodash/compact";
import { DEFAULT_INTEGER_FORMAT, DEFAULT_PERCENTAGE_FORMAT } from "./constants";
export function convertMeasure(measure) {
    var definition = measure.measure.definition;
    var convertedDefinition = convertMeasureDefinition(definition);
    var format = getFormat(measure);
    var formatProp = format ? { format: format } : {};
    var alias = measure.measure.alias ? measure.measure.alias : measure.measure.title;
    var aliasProp = alias ? { alias: alias } : {};
    return __assign(__assign({ localIdentifier: measure.measure.localIdentifier, definition: convertedDefinition }, aliasProp), formatProp);
}
export function convertAggregation(aggregation) {
    if (aggregation === "approximate_count") {
        // Bear doesn't support approximate_count so transparently fallback to exact count.
        return "count";
    }
    return aggregation;
}
function convertMeasureDefinition(definition) {
    if (isMeasureDefinition(definition)) {
        return convertSimpleMeasureDefinition(definition);
    }
    else if (isPoPMeasureDefinition(definition)) {
        return convertPopMeasureDefinition(definition);
    }
    else if (isPreviousPeriodMeasureDefinition(definition)) {
        return convertPreviousPeriodMeasureDefinition(definition);
    }
    else if (isArithmeticMeasureDefinition(definition)) {
        return convertArithmeticMeasureDefinition(definition);
    }
    else {
        throw Error("The measure definition is not supported: " + JSON.stringify(definition));
    }
}
function convertSimpleMeasureDefinition(definition) {
    var measureDefinition = definition.measureDefinition;
    var filters = measureDefinition.filters
        ? compact(measureDefinition.filters.map(convertMeasureFilter))
        : [];
    var filtersProp = filters.length ? { filters: filters } : {};
    var aggregation = convertAggregation(measureDefinition.aggregation);
    var aggregationProp = aggregation ? { aggregation: aggregation } : {};
    var computeRatio = measureDefinition.computeRatio;
    var computeRatioProp = computeRatio ? { computeRatio: computeRatio } : {};
    return {
        measure: __assign(__assign(__assign({ item: toBearRef(measureDefinition.item) }, filtersProp), aggregationProp), computeRatioProp),
    };
}
function convertPopMeasureDefinition(definition) {
    var popMeasureDefinition = definition.popMeasureDefinition;
    return {
        popMeasure: {
            measureIdentifier: popMeasureDefinition.measureIdentifier,
            popAttribute: toBearRef(popMeasureDefinition.popAttribute),
        },
    };
}
function convertPreviousPeriodMeasureDefinition(definition) {
    var previousPeriodMeasure = definition.previousPeriodMeasure;
    return {
        previousPeriodMeasure: {
            measureIdentifier: previousPeriodMeasure.measureIdentifier,
            dateDataSets: previousPeriodMeasure.dateDataSets.map(function (dateDataSet) { return ({
                dataSet: toBearRef(dateDataSet.dataSet),
                periodsAgo: dateDataSet.periodsAgo,
            }); }),
        },
    };
}
function convertArithmeticMeasureDefinition(definition) {
    var arithmeticMeasure = definition.arithmeticMeasure;
    return {
        arithmeticMeasure: {
            measureIdentifiers: arithmeticMeasure.measureIdentifiers.slice(),
            operator: arithmeticMeasure.operator,
        },
    };
}
function getFormat(measure) {
    var _a = measure.measure, definition = _a.definition, format = _a.format;
    if (format) {
        return format;
    }
    var isArithmeticMeasureChange = isArithmeticMeasureDefinition(definition) && definition.arithmeticMeasure.operator === "change";
    if (isArithmeticMeasureChange) {
        return DEFAULT_PERCENTAGE_FORMAT;
    }
    if (isMeasureDefinition(definition)) {
        var measureDefinition = definition.measureDefinition;
        if (measureDefinition.computeRatio) {
            return DEFAULT_PERCENTAGE_FORMAT;
        }
        if (measureDefinition.aggregation === "count") {
            return DEFAULT_INTEGER_FORMAT;
        }
    }
    return undefined;
}
