// @gooddata/sdk-backend-bear/esm/convertors/fromBackend/ExecutionResultConverter.js
// (C) 2019-2020 GoodData Corporation
import { __assign } from "tslib";
import { GdcExecution } from "@gooddata/api-model-bear";
import { isUri } from "@gooddata/api-client-bear";
import { uriRef } from "@gooddata/sdk-model";
var isAttributeHeader = GdcExecution.isAttributeHeader;
export function convertWarning(warning) {
    var _a;
    return {
        warningCode: warning.warningCode,
        message: warning.message,
        parameters: (_a = warning.parameters) === null || _a === void 0 ? void 0 : _a.map(function (param) { return (isUri(param) ? uriRef(param) : param); }),
    };
}
/**
 * Converts execution result's dimension headers as passed by backend into dimension descriptor. At the moment, this function
 * ensures that the 'ref' properties are correctly filled in.
 *
 * @param dims - result dimensions.
 */
export function convertDimensions(dims) {
    return dims.map(function (dim) {
        return {
            headers: dim.headers.map(function (header) {
                if (isAttributeHeader(header)) {
                    return {
                        attributeHeader: __assign(__assign({}, header.attributeHeader), { ref: uriRef(header.attributeHeader.uri), formOf: __assign(__assign({}, header.attributeHeader.formOf), { ref: uriRef(header.attributeHeader.formOf.uri) }) }),
                    };
                }
                else {
                    return {
                        measureGroupHeader: {
                            items: header.measureGroupHeader.items.map(function (measure) {
                                return {
                                    measureHeaderItem: __assign(__assign({}, measure.measureHeaderItem), { ref: measure.measureHeaderItem.uri
                                            ? uriRef(measure.measureHeaderItem.uri)
                                            : undefined }),
                                };
                            }),
                            totalItems: header.measureGroupHeader.totalItems,
                        },
                    };
                }
            }),
        };
    });
}
