// @gooddata/sdk-ui-dashboard/esm/plugins/customizationApis/customizationLogging.js
// (C) 2021-2022 GoodData Corporation
import { pluginDebugStr } from "./pluginUtils";
function addPluginInfoToMessage(plugin, message) {
    return plugin ? pluginDebugStr(plugin) + ": " + message : message;
}
/**
 * Common logger to use for all events that occur during customization. The logger is responsible for adding
 * information about plugin whose registration code triggered those events.
 */
var DashboardCustomizationLogger = /** @class */ (function () {
    function DashboardCustomizationLogger() {
        var _this = this;
        this.setCurrentPlugin = function (plugin) {
            _this.currentPlugin = plugin;
        };
        this.log = function (message) {
            var optionalParams = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                optionalParams[_i - 1] = arguments[_i];
            }
            // eslint-disable-next-line no-console
            console.log(addPluginInfoToMessage(_this.currentPlugin, message), optionalParams);
        };
        this.warn = function (message) {
            var optionalParams = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                optionalParams[_i - 1] = arguments[_i];
            }
            // eslint-disable-next-line no-console
            console.warn(addPluginInfoToMessage(_this.currentPlugin, message), optionalParams);
        };
        this.error = function (message) {
            var optionalParams = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                optionalParams[_i - 1] = arguments[_i];
            }
            // eslint-disable-next-line no-console
            console.error(addPluginInfoToMessage(_this.currentPlugin, message), optionalParams);
        };
    }
    return DashboardCustomizationLogger;
}());
export { DashboardCustomizationLogger };
