// @gooddata/sdk-ui-dashboard/esm/plugins/customizationApis/customizationBuilder.js
// (C) 2021-2022 GoodData Corporation
import { __assign } from "tslib";
import { DefaultInsightCustomizer } from "./insightCustomizer";
import { DashboardCustomizationLogger } from "./customizationLogging";
import { DefaultKpiCustomizer } from "./kpiCustomizer";
import { DefaultWidgetCustomizer } from "./widgetCustomizer";
import { DefaultLayoutCustomizer } from "./layoutCustomizer";
import { DefaultFilterBarCustomizer } from "./filterBarCustomizer";
/**
 * @internal
 */
var DashboardCustomizationBuilder = /** @class */ (function () {
    function DashboardCustomizationBuilder() {
        var _this = this;
        this.logger = new DashboardCustomizationLogger();
        this.insightCustomizer = new DefaultInsightCustomizer(this.logger);
        this.kpiCustomizer = new DefaultKpiCustomizer(this.logger);
        this.widgetCustomizer = new DefaultWidgetCustomizer(this.logger);
        this.layoutCustomizer = new DefaultLayoutCustomizer(this.logger);
        this.filterBarCustomizer = new DefaultFilterBarCustomizer(this.logger);
        this.sealCustomizers = function () {
            _this.insightCustomizer.sealCustomizer();
            _this.kpiCustomizer.sealCustomizer();
            _this.widgetCustomizer.sealCustomizer();
            _this.layoutCustomizer.sealCustomizer();
        };
        this.insightWidgets = function () {
            return _this.insightCustomizer;
        };
        this.kpiWidgets = function () {
            return _this.kpiCustomizer;
        };
        this.customWidgets = function () {
            return _this.widgetCustomizer;
        };
        this.layout = function () {
            return _this.layoutCustomizer;
        };
        this.filterBar = function () {
            return _this.filterBarCustomizer;
        };
        this.onBeforePluginRegister = function (plugin) {
            _this.logger.setCurrentPlugin(plugin);
            _this.logger.log("Starting registration.");
        };
        this.onAfterPluginRegister = function () {
            _this.logger.log("Registration finished.");
            _this.logger.setCurrentPlugin(undefined);
        };
        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
        this.onPluginRegisterError = function (error) {
            _this.logger.error("Plugin register() method threw and exception. Plugin may be partially registered.", error);
            _this.logger.setCurrentPlugin(undefined);
        };
        this.build = function () {
            var filterBarCustomizerResult = _this.filterBarCustomizer.getCustomizerResult();
            var props = __assign({ InsightComponentProvider: _this.insightCustomizer.getInsightProvider(), KpiComponentProvider: _this.kpiCustomizer.getKpiProvider(), WidgetComponentProvider: _this.widgetCustomizer.getWidgetComponentProvider(), customizationFns: {
                    existingDashboardTransformFn: _this.layoutCustomizer.getExistingDashboardTransformFn(),
                } }, (filterBarCustomizerResult.FilterBarComponent
                ? { FilterBarComponent: filterBarCustomizerResult.FilterBarComponent }
                : {}));
            _this.sealCustomizers();
            return props;
        };
    }
    return DashboardCustomizationBuilder;
}());
export { DashboardCustomizationBuilder };
