// @gooddata/sdk-backend-base/esm/index.js
// (C) 2019-2022 GoodData Corporation
/**
 * This package provides foundational reusable code useful for building new or decorating existing Analytical Backend implementations.
 *
 * @remarks
 * The package includes several composable backend decorators and metadata object builders.
 * You can use them to either build your own Analytical Backend implementation or enhance existing
 * implementations (by adding caching, for example).
 *
 * @packageDocumentation
 */
export { dummyBackend, dummyBackendEmptyData, dummyDataView } from "./dummyBackend";
export { decoratedBackend, } from "./decoratedBackend";
export { DecoratedExecutionFactory, DecoratedPreparedExecution, DecoratedExecutionResult, } from "./decoratedBackend/execution";
export { DecoratedWorkspaceDashboardsService } from "./decoratedBackend/dashboards";
export { DecoratedWorkspaceCatalogFactory, DecoratedWorkspaceCatalog, } from "./decoratedBackend/catalog";
export { DecoratedSecuritySettingsService } from "./decoratedBackend/securitySettings";
export { DecoratedWorkspaceSettingsService } from "./decoratedBackend/workspaceSettings";
export { withEventing } from "./eventingBackend";
export { withCaching, DefaultCachingConfiguration, } from "./cachingBackend";
export { withCustomWorkspaceSettings, } from "./workspaceSettingsBackend";
export { withNormalization } from "./normalizingBackend";
export { Normalizer, Denormalizer } from "./normalizingBackend/normalizer";
export { AuthProviderCallGuard, NoopAuthProvider, AnonymousAuthProvider, } from "./toolkit/auth";
export { AbstractExecutionFactory, ExecutionFactoryWithFixedFilters, ExecutionFactoryUpgradingToExecByReference, } from "./toolkit/execution";
export { InMemoryPaging, ServerPaging } from "./toolkit/paging";
export { validatePluginUrlIsSane } from "./toolkit/pluginUrlValidation";
export { customBackend } from "./customBackend";
export { Builder, builderFactory, resolveValueOrUpdateCallback, } from "./ldmFactories/builder";
export { InsightWidgetBuilder, newInsightWidget, } from "./ldmFactories/dashboard/insightWidgetFactory";
export { KpiWidgetBuilder, newKpiWidget } from "./ldmFactories/dashboard/kpiWidgetFactory";
export { WidgetBaseBuilder } from "./ldmFactories/dashboard/widgetFactory";
export { CatalogAttributeBuilder, newCatalogAttribute } from "./ldmFactories/catalog/attributeFactory";
export { CatalogDateAttributeBuilder, CatalogDateDatasetBuilder, newCatalogDateAttribute, newCatalogDateDataset, } from "./ldmFactories/catalog/dateDatasetFactory";
export { CatalogFactBuilder, newCatalogFact } from "./ldmFactories/catalog/factFactory";
export { CatalogGroupBuilder, GroupableCatalogItemBuilder, newCatalogGroup, } from "./ldmFactories/catalog/groupFactory";
export { CatalogMeasureBuilder, newCatalogMeasure } from "./ldmFactories/catalog/measureFactory";
export { AttributeMetadataObjectBuilder, newAttributeMetadataObject, } from "./ldmFactories/metadata/attributeFactory";
export { DataSetMetadataObjectBuilder, newDataSetMetadataObject, } from "./ldmFactories/metadata/dataSetFactory";
export { AttributeDisplayFormMetadataObjectBuilder, newAttributeDisplayFormMetadataObject, } from "./ldmFactories/metadata/displayFormFactory";
export { MetadataObjectBuilder } from "./ldmFactories/metadata/factory";
export { newFactMetadataObject, FactMetadataObjectBuilder } from "./ldmFactories/metadata/factFactory";
export { MeasureMetadataObjectBuilder, newMeasureMetadataObject, } from "./ldmFactories/metadata/measureFactory";
export { newVariableMetadataObject, VariableMetadataObjectBuilder, } from "./ldmFactories/metadata/variableFactory";
export { newDashboardMetadataObject, DashboardMetadataObjectBuilder, } from "./ldmFactories/metadata/dashboardFactory";
export { transformResultHeaders } from "./convertors/fromBackend/afm/result";
