// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/_chartCreators/drilldownEventing.js
import { __assign } from "tslib";
// (C) 2007-2021 GoodData Corporation
import debounce from "lodash/debounce";
import cloneDeep from "lodash/cloneDeep";
import CustomEvent from "custom-event";
import invariant from "ts-invariant";
import { getVisualizationType, VisualizationTypes, } from "@gooddata/sdk-ui";
import { isBulletChart, isComboChart, isHeatmap, isTreemap } from "../_util/common";
import { isGroupHighchartsDrillEvent } from "./isGroupHighchartsDrillEvent";
export function getClickableElementNameByChartType(type) {
    switch (type) {
        case VisualizationTypes.LINE:
        case VisualizationTypes.AREA:
        case VisualizationTypes.SCATTER:
        case VisualizationTypes.BUBBLE:
            return "point";
        case VisualizationTypes.COLUMN:
        case VisualizationTypes.BAR:
            return "bar";
        case VisualizationTypes.PIE:
        case VisualizationTypes.TREEMAP:
        case VisualizationTypes.DONUT:
        case VisualizationTypes.FUNNEL:
            return "slice";
        case VisualizationTypes.HEATMAP:
            return "cell";
        default:
            invariant(false, "Unknown visualization type: " + type);
            return null;
    }
}
function fireEvent(onDrill, data, target) {
    var returnValue = onDrill(data);
    // if user-specified onDrill fn returns false, do not fire default DOM event
    if (returnValue !== false) {
        var event_1 = new CustomEvent("drill", {
            detail: data,
            bubbles: true,
        });
        target.dispatchEvent(event_1);
    }
}
var getElementChartType = function (chartType, point) {
    var _a, _b;
    return (_b = (_a = point === null || point === void 0 ? void 0 : point.series) === null || _a === void 0 ? void 0 : _a.type) !== null && _b !== void 0 ? _b : chartType;
};
var getDrillPointCustomProps = function (point, chartType) {
    var _a, _b, _c;
    if (isComboChart(chartType)) {
        return { type: (_a = point === null || point === void 0 ? void 0 : point.series) === null || _a === void 0 ? void 0 : _a.type };
    }
    if (isBulletChart(chartType)) {
        return { type: (_c = (_b = point === null || point === void 0 ? void 0 : point.series) === null || _b === void 0 ? void 0 : _b.userOptions) === null || _c === void 0 ? void 0 : _c.bulletChartMeasureType };
    }
    return {};
};
var getYValueForBulletChartTarget = function (point) {
    if (point.isNullTarget) {
        return null;
    }
    return point.target;
};
function composeDrillContextGroup(points, chartType) {
    var sanitizedPoints = sanitizeContextPoints(chartType, points);
    var contextPoints = sanitizedPoints.map(function (point) {
        var elementChartType = getElementChartType(chartType, point);
        var customProps = getDrillPointCustomProps(point, chartType);
        return __assign({ x: point.x, y: elementChartType === "bullet" ? getYValueForBulletChartTarget(point) : point.y, intersection: point.drillIntersection }, customProps);
    });
    return {
        type: chartType,
        element: "label",
        points: contextPoints,
    };
}
function getClickableElementNameForBulletChart(point) {
    return point.series.userOptions.bulletChartMeasureType;
}
function composeDrillContextPoint(point, chartType) {
    var zProp = isNaN(point.z) ? {} : { z: point.z };
    var valueProp = isTreemap(chartType) || isHeatmap(chartType)
        ? {
            value: point.value ? point.value.toString() : "",
        }
        : {};
    var elementChartType = getElementChartType(chartType, point);
    var xyProp = isTreemap(chartType)
        ? {}
        : {
            x: point.x,
            y: elementChartType === "bullet" ? point.target : point.y,
        };
    var customProp = isComboChart(chartType)
        ? {
            elementChartType: elementChartType,
        }
        : {};
    return __assign(__assign(__assign(__assign({ type: chartType, element: isBulletChart(chartType)
            ? getClickableElementNameForBulletChart(point)
            : getClickableElementNameByChartType(elementChartType), intersection: point.drillIntersection }, xyProp), zProp), valueProp), customProp);
}
var chartClickDebounced = debounce(function (drillConfig, event, target, chartType) {
    var dataView = drillConfig.dataView, onDrill = drillConfig.onDrill;
    var type = getVisualizationType(chartType);
    var drillContext;
    if (isGroupHighchartsDrillEvent(event)) {
        var points = event.points;
        drillContext = composeDrillContextGroup(points, type);
    }
    else {
        var point = event.point;
        drillContext = composeDrillContextPoint(point, type);
    }
    var data = {
        dataView: dataView,
        drillContext: drillContext,
    };
    fireEvent(onDrill, data, target);
});
export function chartClick(drillConfig, event, target, chartType) {
    chartClickDebounced(drillConfig, event, target, chartType);
}
var tickLabelClickDebounce = debounce(function (drillConfig, points, target, chartType) {
    var dataView = drillConfig.dataView, onDrill = drillConfig.onDrill;
    var sanitizedPoints = sanitizeContextPoints(chartType, points);
    var contextPoints = sanitizedPoints.map(function (point) {
        var _a, _b;
        var customProps = isBulletChart(chartType)
            ? { type: (_b = (_a = point === null || point === void 0 ? void 0 : point.series) === null || _a === void 0 ? void 0 : _a.userOptions) === null || _b === void 0 ? void 0 : _b.bulletChartMeasureType }
            : {};
        return __assign({ x: point.x, y: point.y, intersection: cloneDeep(point.drillIntersection) }, customProps);
    });
    var drillContext = {
        type: chartType,
        element: "label",
        points: contextPoints,
    };
    var data = {
        dataView: dataView,
        drillContext: drillContext,
    };
    fireEvent(onDrill, data, target);
});
function sanitizeContextPoints(chartType, points) {
    if (isHeatmap(chartType)) {
        return points.filter(function (point) { return !point.ignoredInDrillEventContext; });
    }
    return points;
}
export function tickLabelClick(drillConfig, points, target, chartType) {
    tickLabelClickDebounce(drillConfig, points, target, chartType);
}
