// @gooddata/sdk-ui-pivot/esm/impl/base/agUtils.js
import { MEASURE_COLUMN } from "./constants";
import { agColId } from "../structure/tableDescriptorTypes";
import { ColumnEventSourceType } from "../../columnWidths";
/*
 * Assorted utility functions used in our Pivot Table -> ag-grid integration.
 */
export function getGridIndex(position, gridDistance) {
    return Math.floor(position / gridDistance);
}
export function isColumn(item) {
    return !!item.getColDef;
}
export function isMeasureColumn(item) {
    if (isColumn(item)) {
        return item.getColDef().type === MEASURE_COLUMN;
    }
    return item.type === MEASURE_COLUMN;
}
export function agColIds(columns) {
    return columns.map(agColId);
}
export function isHeaderResizer(target) {
    return target.classList.contains("ag-header-cell-resize");
}
export function isManualResizing(columnEvent) {
    return Boolean(columnEvent && columnEvent.source === ColumnEventSourceType.UI_DRAGGED && columnEvent.columns);
}
export function scrollBarExists(target) {
    var _a = target.getElementsByClassName("ag-body-horizontal-scroll-viewport")[0], scrollWidth = _a.scrollWidth, clientWidth = _a.clientWidth;
    return scrollWidth > clientWidth;
}
