// @gooddata/sdk-ui-pivot/esm/impl/base/constants.js
// (C) 2007-2021 GoodData Corporation
export var ROW_ATTRIBUTE_COLUMN = "ROW_ATTRIBUTE_COLUMN";
export var COLUMN_ATTRIBUTE_COLUMN = "COLUMN_ATTRIBUTE_COLUMN";
export var MEASURE_COLUMN = "MEASURE_COLUMN";
export var ROW_TOTAL = "rowTotal";
export var ROW_SUBTOTAL = "rowSubtotal";
export var COLUMN_GROUPING_DELIMITER = " › ";
export var AVAILABLE_TOTALS = ["sum", "max", "min", "avg", "med", "nat"];
export var COLS_PER_PAGE = 1000;
export var VALUE_CLASS = "s-value";
export var HEADER_LABEL_CLASS = "s-header-cell-label";
export var ROW_TOTAL_CLASS = "gd-row-total";
export var ROW_SUBTOTAL_CLASS = "gd-table-row-subtotal";
export var DEFAULT_HEADER_FONT = '12px / 28px gdcustomfont, avenir, "Helvetica Neue", arial, sans-serif';
export var DEFAULT_ROW_FONT = '12px / 26px gdcustomfont, avenir, "Helvetica Neue", arial, sans-serif';
export var DEFAULT_SUBTOTAL_FONT = '700 12px / 26px gdcustomfont, avenir, "Helvetica Neue", arial, sans-serif';
export var DEFAULT_TOTAL_FONT = '700 12px / 26px gdcustomfont, avenir, "Helvetica Neue", arial, sans-serif';
/**
 * DEFAULT_AUTOSIZE_PADDING needs to match real padding from styles
 */
export var DEFAULT_AUTOSIZE_PADDING = 12;
export var DEFAULT_ROW_HEIGHT = 28;
