// @gooddata/sdk-ui-dashboard/esm/model/store/ui/uiReducers.js
var openScheduleEmailDialog = function (state) {
    state.scheduleEmailDialog.open = true;
};
var closeScheduleEmailDialog = function (state) {
    state.scheduleEmailDialog.open = false;
};
var openSaveAsDialog = function (state) {
    state.saveAsDialog.open = true;
};
var closeSaveAsDialog = function (state) {
    state.saveAsDialog.open = false;
};
var openShareDialog = function (state) {
    state.shareDialog.open = true;
};
var closeShareDialog = function (state) {
    state.shareDialog.open = false;
};
var setFilterBarHeight = function (state, action) {
    state.filterBar.height = action.payload;
};
var setFilterBarExpanded = function (state, action) {
    state.filterBar.expanded = action.payload;
};
var closeKpiAlertDialog = function (state) {
    state.kpiAlerts.openedWidgetRef = undefined;
};
var openKpiAlertDialog = function (state, action) {
    state.kpiAlerts.openedWidgetRef = action.payload;
};
var highlightKpiAlert = function (state, action) {
    state.kpiAlerts.highlightedWidgetRef = action.payload;
};
var setMenuButtonItemsVisibility = function (state, action) {
    state.menuButton.itemsVisibility = action.payload;
};
export var uiReducers = {
    openScheduleEmailDialog: openScheduleEmailDialog,
    closeScheduleEmailDialog: closeScheduleEmailDialog,
    openSaveAsDialog: openSaveAsDialog,
    closeSaveAsDialog: closeSaveAsDialog,
    setFilterBarHeight: setFilterBarHeight,
    setFilterBarExpanded: setFilterBarExpanded,
    closeKpiAlertDialog: closeKpiAlertDialog,
    openKpiAlertDialog: openKpiAlertDialog,
    highlightKpiAlert: highlightKpiAlert,
    openShareDialog: openShareDialog,
    closeShareDialog: closeShareDialog,
    setMenuButtonItemsVisibility: setMenuButtonItemsVisibility,
};
