// @gooddata/sdk-ui-dashboard/esm/model/store/ui/uiSelectors.js
// (C) 2021-2022 GoodData Corporation
import { createSelector } from "@reduxjs/toolkit";
import { selectWidgetsMap } from "../layout/layoutSelectors";
import { createMemoizedSelector } from "../_infra/selectors";
var selectSelf = createSelector(function (state) { return state; }, function (state) { return state.ui; });
/**
 * @alpha
 */
export var selectIsScheduleEmailDialogOpen = createSelector(selectSelf, function (state) { return state.scheduleEmailDialog.open; });
/**
 * @alpha
 */
export var selectIsSaveAsDialogOpen = createSelector(selectSelf, function (state) { return state.saveAsDialog.open; });
/**
 * @alpha
 */
export var selectIsShareDialogOpen = createSelector(selectSelf, function (state) { return state.shareDialog.open; });
/**
 * @internal
 */
export var selectFilterBarHeight = createSelector(selectSelf, function (state) { return state.filterBar.height; });
/**
 * @alpha
 */
export var selectFilterBarExpanded = createSelector(selectSelf, function (state) { return state.filterBar.expanded; });
var selectHighlightedKpiWidgetRef = createSelector(selectSelf, function (state) { return state.kpiAlerts.highlightedWidgetRef; });
var selectOpenedKpiWidgetRef = createSelector(selectSelf, function (state) { return state.kpiAlerts.openedWidgetRef; });
/**
 * @alpha
 */
export var selectIsKpiAlertOpenedByWidgetRef = createMemoizedSelector(function (ref) {
    return createSelector(selectWidgetsMap, selectOpenedKpiWidgetRef, function (widgets, openedWidgetRef) {
        if (!ref) {
            return false;
        }
        var openedWidget = openedWidgetRef && widgets.get(openedWidgetRef);
        if (!openedWidget) {
            return false;
        }
        var targetWidget = widgets.get(ref);
        if (!targetWidget) {
            return false;
        }
        return targetWidget.identifier === openedWidget.identifier;
    });
});
/**
 * @alpha
 */
export var selectIsKpiAlertHighlightedByWidgetRef = createMemoizedSelector(function (ref) {
    return createSelector(selectWidgetsMap, selectHighlightedKpiWidgetRef, function (widgets, highlightedWidgetRef) {
        if (!ref) {
            return false;
        }
        var highlightedWidget = highlightedWidgetRef && widgets.get(highlightedWidgetRef);
        if (!highlightedWidget) {
            return false;
        }
        var targetWidget = widgets.get(ref);
        if (!targetWidget) {
            return false;
        }
        return targetWidget.identifier === highlightedWidget.identifier;
    });
});
/**
 * @alpha
 */
export var selectMenuButtonItemsVisibility = createSelector(selectSelf, function (state) { var _a; return (_a = state.menuButton.itemsVisibility) !== null && _a !== void 0 ? _a : {}; });
