// @gooddata/sdk-ui-filters/esm/DateFilter/index.js
// (C) 2007-2021 GoodData Corporation
import { granularityIntlCodes } from "./constants/i18n";
import { defaultDateFilterOptions } from "./constants/config";
import { validateFilterOption } from "./validation/OptionValidation";
import { mapOptionToAfm } from "./utils/AFMConversions";
import { applyExcludeCurrentPeriod, canExcludeCurrentPeriod } from "./utils/PeriodExclusion";
import { filterVisibleDateFilterOptions } from "./utils/OptionUtils";
import { getDateFilterTitle, getDateFilterRepresentation, getDateFilterTitleUsingTranslator, formatAbsoluteDateRange, formatRelativeDateRange, } from "./utils/Translations/DateFilterTitle";
import { DateFilter, } from "./DateFilter";
/**
 * @beta
 */
var DateFilterHelpers = {
    validateFilterOption: validateFilterOption,
    getDateFilterTitle: getDateFilterTitle,
    getDateFilterTitleUsingTranslator: getDateFilterTitleUsingTranslator,
    getDateFilterRepresentation: getDateFilterRepresentation,
    granularityIntlCodes: granularityIntlCodes,
    applyExcludeCurrentPeriod: applyExcludeCurrentPeriod,
    defaultDateFilterOptions: defaultDateFilterOptions,
    canExcludeCurrentPeriod: canExcludeCurrentPeriod,
    mapOptionToAfm: mapOptionToAfm,
    formatAbsoluteDateRange: formatAbsoluteDateRange,
    formatRelativeDateRange: formatRelativeDateRange,
    filterVisibleDateFilterOptions: filterVisibleDateFilterOptions,
};
// This is 1:1 reexported by root index.ts and is part of SDK's public API
export { DateFilter, DateFilterHelpers, filterVisibleDateFilterOptions, };
export { isAbsoluteDateFilterOption, isRelativeDateFilterOption, isUiRelativeDateFilterForm, } from "./interfaces";
export { defaultDateFilterOptions } from "./constants/config";
