// @gooddata/sdk-model/esm/execution/base/dimension.js
// (C) 2019-2020 GoodData Corporation
import { __assign } from "tslib";
import invariant from "ts-invariant";
import { isTotal } from "./totals";
import isEmpty from "lodash/isEmpty";
import findIndex from "lodash/findIndex";
import { attributeLocalId, isAttribute } from "../attribute";
/**
 * Measure Group is a pseudo-identifier which can be used in an execution dimension and indicates
 * that this dimension MUST contain all the measures.
 *
 * @public
 */
export var MeasureGroupIdentifier = "measureGroup";
var isMeasureGroupIdentifier = function (itemOrTotal) { return itemOrTotal === MeasureGroupIdentifier; };
//
// Type guards
//
/**
 * Type guard checking whether object is of IDimension type.
 *
 * @public
 */
export function isDimension(obj) {
    return !isEmpty(obj) && obj.itemIdentifiers !== undefined;
}
//
// Public functions
//
/**
 * Gets totals defined in the provided dimension
 *
 * @param dim - dimension to work with
 * @returns totals in the dimension or empty array if none
 * @public
 */
export function dimensionTotals(dim) {
    invariant(dim, "dimension must be specified");
    return dim.totals ? dim.totals : [];
}
/**
 * Creates a new dimension which has same items as the provided dimension but different totals.
 *
 * @param dim - dimension to inherit item identifiers from
 * @param totals - totals to have in the new dimension
 * @returns new dimension
 * @public
 */
export function dimensionSetTotals(dim, totals) {
    if (totals === void 0) { totals = []; }
    invariant(dim, "dimension must be specified");
    var totalsProp = !isEmpty(totals) ? { totals: totals } : {};
    return __assign({ itemIdentifiers: dim.itemIdentifiers }, totalsProp);
}
/**
 * Creates new two dimensional specification where each dimension will have the provided set of
 * identifiers.
 *
 * The 'measureGroup' identifier MAY be specified in only one of the dimensions.
 *
 * @param dim1Input - items to put into the first dimension, this can be item identifiers or totals
 * @param dim2Input - items to put into the second dimension, this can be item identifiers or totals
 * @returns array with exactly two dimensions
 * @public
 */
export function newTwoDimensional(dim1Input, dim2Input) {
    invariant(dim1Input, "input for first dimension must be specified");
    invariant(dim2Input, "input for second dimension must be specified");
    var atMostOneMeasureGroup = !(dim1Input.find(isMeasureGroupIdentifier) && dim2Input.find(isMeasureGroupIdentifier));
    invariant(atMostOneMeasureGroup, "The 'measureGroup' identifier must only be specified in one dimension.");
    return [newDimension(dim1Input), newDimension(dim2Input)];
}
/**
 * Creates new single-dimensional specification where the dimension will have the provided set of identifiers.
 *
 * @param items - allows for mix of item identifiers, attributes and total definitions to have in the new dimension
 * @param totals - additional totals to add to the dimension
 * @returns single dimension
 * @public
 */
export function newDimension(items, totals) {
    var _a;
    if (items === void 0) { items = []; }
    if (totals === void 0) { totals = []; }
    var input = items.reduce(function (acc, value) {
        if (isTotal(value)) {
            acc.totals.push(value);
        }
        else if (isAttribute(value)) {
            acc.ids.push(attributeLocalId(value));
        }
        else {
            acc.ids.push(value);
        }
        return acc;
    }, { ids: [], totals: [] });
    (_a = input.totals).push.apply(_a, totals);
    var totalsProp = !isEmpty(input.totals) ? { totals: input.totals } : {};
    return __assign({ itemIdentifiers: input.ids }, totalsProp);
}
/**
 * Looks for item with the provided local identifier among the dimensions.
 *
 * @param dims - list of dimensions to look in
 * @param localId - local identifier to find among item identifiers
 * @returns list of items in dimensions, empty if not found, may contain more than one entry if
 *  item is in multiple dimensions
 * @public
 */
export function dimensionsFindItem(dims, localId) {
    var result = [];
    for (var dimIdx = 0; dimIdx < dims.length; dimIdx++) {
        var dim = dims[dimIdx];
        var itemIdx = findIndex(dim.itemIdentifiers, function (i) { return i === localId; });
        if (itemIdx >= 0) {
            result.push({ dim: dim, dimIdx: dimIdx, itemIdx: itemIdx });
        }
    }
    return result;
}
