// @gooddata/sdk-ui-ext/esm/dashboardView/DashboardWidgetRenderer/KpiView/alertManipulationHooks/useAlertDeleteHandler.js
// (C) 2021 GoodData Corporation
import { useState } from "react";
import { useDeleteWidgetAlert } from "../../../hooks/internal";
import { useAlerts } from "../../../contexts";
export function useAlertDeleteHandler(_a) {
    var closeAlertDialog = _a.closeAlertDialog, backend = _a.backend, workspace = _a.workspace;
    var removeAlert = useAlerts().removeAlert;
    var _b = useState(undefined), alertToDelete = _b[0], setAlertToDelete = _b[1];
    var deleteAlert = function (alert) { return setAlertToDelete(alert); };
    var _c = useState("idle"), alertDeletingStatus = _c[0], setAlertDeletingStatus = _c[1];
    useDeleteWidgetAlert({
        backend: backend,
        workspace: workspace,
        widgetAlert: alertToDelete,
        onError: function () {
            setAlertDeletingStatus("error");
            setAlertToDelete(undefined);
        },
        onLoading: function () { return setAlertDeletingStatus("inProgress"); },
        onSuccess: function () {
            setAlertDeletingStatus("idle");
            removeAlert(alertToDelete);
            setAlertToDelete(undefined);
            closeAlertDialog();
        },
    });
    return {
        alertDeletingStatus: alertDeletingStatus,
        deleteAlert: deleteAlert,
    };
}
