// @gooddata/sdk-ui-ext/esm/dataLoaders/index.js
import { colorPaletteDataLoaderFactory } from "./ColorPaletteDataLoader";
import { insightDataLoaderFactory } from "./InsightDataLoader";
import { userWorkspaceSettingsDataLoaderFactory } from "./UserWorkspaceSettingsDataLoader";
import { attributesWithDrillDownDataLoaderFactory } from "./AttributesWithDrillDownDataLoader";
import { dashboardAlertsDataLoaderFactory } from "./DashboardAlertsDataLoader";
import { dashboardDataLoaderFactory } from "./DashboardDataLoader";
import { dateDatasetsDataLoaderFactory } from "./DateDatasetsDataLoader";
import { userWorkspacePermissionsDataLoaderFactory } from "./UserWorkspacePermissionsDataLoader";
/**
 * Clears all the caches used by the InsightView components.
 *
 * @public
 */
export function clearInsightViewCaches() {
    var relevantFactories = [
        colorPaletteDataLoaderFactory,
        insightDataLoaderFactory,
        userWorkspaceSettingsDataLoaderFactory,
    ];
    relevantFactories.forEach(function (factory) { return factory.reset(); });
}
/**
 * Clears all the caches used by the DashboardView components.
 *
 * @beta
 */
export function clearDashboardViewCaches() {
    var relevantFactories = [
        attributesWithDrillDownDataLoaderFactory,
        colorPaletteDataLoaderFactory,
        dashboardAlertsDataLoaderFactory,
        dashboardDataLoaderFactory,
        dateDatasetsDataLoaderFactory,
        insightDataLoaderFactory,
        userWorkspacePermissionsDataLoaderFactory,
        userWorkspaceSettingsDataLoaderFactory,
    ];
    relevantFactories.forEach(function (factory) { return factory.reset(); });
}
export { colorPaletteDataLoaderFactory, insightDataLoaderFactory, userWorkspaceSettingsDataLoaderFactory, attributesWithDrillDownDataLoaderFactory, dashboardAlertsDataLoaderFactory, dashboardDataLoaderFactory, dateDatasetsDataLoaderFactory, userWorkspacePermissionsDataLoaderFactory, };
