// @gooddata/sdk-ui-filters/esm/DateFilter/utils/AFMConversions.js
import { applyExcludeCurrentPeriod } from "./PeriodExclusion";
import { isAbsoluteDateFilterOption, isRelativeDateFilterOption, } from "../interfaces";
import { isAllTimeDateFilterOption } from "@gooddata/sdk-backend-spi";
export var mapAbsoluteFilterToAfm = function (value, dataSet) { return ({
    absoluteDateFilter: {
        dataSet: dataSet,
        from: value.from,
        to: value.to,
    },
}); };
export var mapRelativeFilterToAfm = function (value, dataSet) { return ({
    relativeDateFilter: {
        dataSet: dataSet,
        from: value.from,
        to: value.to,
        granularity: value.granularity,
    },
}); };
export var mapOptionToAfm = function (value, dateDataSet, excludeCurrentPeriod) {
    var excludeApplied = applyExcludeCurrentPeriod(value, excludeCurrentPeriod);
    if (isAllTimeDateFilterOption(excludeApplied)) {
        return null;
    }
    if (isAbsoluteDateFilterOption(excludeApplied)) {
        return mapAbsoluteFilterToAfm(excludeApplied, dateDataSet);
    }
    if (isRelativeDateFilterOption(excludeApplied)) {
        return mapRelativeFilterToAfm(excludeApplied, dateDataSet);
    }
    throw new Error("Unknown date filter value provided");
};
