// @gooddata/sdk-ui-dashboard/esm/presentation/widget/widget/DefaultDashboardInsightWidget.js
import { __assign } from "tslib";
// (C) 2020-2022 GoodData Corporation
import React, { useMemo } from "react";
import cx from "classnames";
import { injectIntl } from "react-intl";
import { insightVisualizationUrl } from "@gooddata/sdk-model";
import { widgetTitle } from "@gooddata/sdk-backend-spi";
import { useDashboardSelector, selectInsightsMap } from "../../../model";
import { DashboardItem, DashboardItemHeadline, DashboardItemVisualization, getVisTypeCssClass, } from "../../presentationComponents";
import { DashboardInsight } from "../insight/DashboardInsight";
import { useInsightExport } from "../common/useInsightExport";
import { useDashboardComponentsContext } from "../../dashboardContexts";
import { useInsightMenu } from "./useInsightMenu";
// Sometimes this component is rendered even before insights are ready, which blows up.
// Since the behavior is nearly impossible to replicate reliably, let's be defensive here and not render
// anything until the insights "catch up".
var DefaultDashboardInsightWidgetWrapper = function (props) {
    var widget = props.widget, 
    // @ts-expect-error Don't expose index prop on public interface (we need it only for css class for KD tests)
    index = props.index;
    var insights = useDashboardSelector(selectInsightsMap);
    var insight = insights.get(widget.insight);
    if (!insight) {
        // eslint-disable-next-line no-console
        console.debug("DefaultDashboardInsightWidget rendered before the insights were ready, skipping render.");
        return null;
    }
    return (React.createElement(DefaultDashboardInsightWidgetCore, __assign({}, props, { insight: insight, 
        // @ts-expect-error Don't expose index prop on public interface (we need it only for css class for KD tests)
        index: index })));
};
/**
 * @internal
 */
var DefaultDashboardInsightWidgetCore = function (_a) {
    var widget = _a.widget, insight = _a.insight, screen = _a.screen, onError = _a.onError, onExportReady = _a.onExportReady, onLoadingChanged = _a.onLoadingChanged, intl = _a.intl, 
    // @ts-expect-error Don't expose index prop on public interface (we need it only for css class for KD tests)
    index = _a.index;
    var visType = insightVisualizationUrl(insight).split(":")[1];
    var _b = useInsightExport({
        widgetRef: widget.ref,
        title: widgetTitle(widget) || intl.formatMessage({ id: "export.defaultTitle" }),
        insight: insight,
    }), exportCSVEnabled = _b.exportCSVEnabled, exportXLSXEnabled = _b.exportXLSXEnabled, onExportCSV = _b.onExportCSV, onExportXLSX = _b.onExportXLSX;
    var _c = useInsightMenu({
        insight: insight,
        widget: widget,
        exportCSVEnabled: exportCSVEnabled,
        exportXLSXEnabled: exportXLSXEnabled,
        onExportCSV: onExportCSV,
        onExportXLSX: onExportXLSX,
    }), closeMenu = _c.closeMenu, isMenuOpen = _c.isMenuOpen, menuItems = _c.menuItems, openMenu = _c.openMenu;
    var _d = useDashboardComponentsContext(), InsightMenuButtonComponentProvider = _d.InsightMenuButtonComponentProvider, InsightMenuComponentProvider = _d.InsightMenuComponentProvider, ErrorComponent = _d.ErrorComponent, LoadingComponent = _d.LoadingComponent;
    var InsightMenuButtonComponent = useMemo(function () { return InsightMenuButtonComponentProvider(insight, widget); }, [InsightMenuButtonComponentProvider, insight, widget]);
    var InsightMenuComponent = useMemo(function () { return InsightMenuComponentProvider(insight, widget); }, [InsightMenuComponentProvider, insight, widget]);
    return (React.createElement(DashboardItem, { className: cx("s-dash-item-" + index, "type-visualization", "gd-dashboard-view-widget", getVisTypeCssClass(widget.type, visType)), screen: screen },
        React.createElement(DashboardItemVisualization, { renderHeadline: function (clientHeight) { return (React.createElement(DashboardItemHeadline, { title: widget.title, clientHeight: clientHeight })); }, renderBeforeVisualization: function () { return (React.createElement(InsightMenuButtonComponent, { insight: insight, widget: widget, isOpen: isMenuOpen, onClick: openMenu, items: menuItems })); }, renderAfterContent: function () {
                if (!isMenuOpen) {
                    return null;
                }
                return (React.createElement(InsightMenuComponent, { insight: insight, widget: widget, isOpen: isMenuOpen, onClose: closeMenu, items: menuItems }));
            } }, function (_a) {
            var clientHeight = _a.clientHeight, clientWidth = _a.clientWidth;
            return (React.createElement(DashboardInsight, { clientHeight: clientHeight, clientWidth: clientWidth, insight: insight, widget: widget, onExportReady: onExportReady, onLoadingChanged: onLoadingChanged, onError: onError, ErrorComponent: ErrorComponent, LoadingComponent: LoadingComponent }));
        })));
};
export var DefaultDashboardInsightWidget = injectIntl(DefaultDashboardInsightWidgetWrapper);
