// @gooddata/sdk-ui-ext/esm/internal/components/VisualizationCatalog.js
import { __assign } from "tslib";
// (C) 2020-2021 GoodData Corporation
import last from "lodash/last";
import { insightVisualizationUrl } from "@gooddata/sdk-model";
import { UnexpectedSdkError } from "@gooddata/sdk-ui";
import { AreaChartDescriptor } from "./pluggableVisualizations/areaChart/AreaChartDescriptor";
import { BarChartDescriptor } from "./pluggableVisualizations/barChart/BarChartDescriptor";
import { BubbleChartDescriptor } from "./pluggableVisualizations/bubbleChart/BubbleChartDescriptor";
import { BulletChartDescriptor } from "./pluggableVisualizations/bulletChart/BulletChartDescriptor";
import { ColumnChartDescriptor } from "./pluggableVisualizations/columnChart/ColumnChartDescriptor";
import { ComboChartDescriptor } from "./pluggableVisualizations/comboChart/ComboChartDescriptor";
import { ComboChartDescriptorDeprecated } from "./pluggableVisualizations/comboChart/ComboChartDescriptorDeprecated";
import { DonutChartDescriptor } from "./pluggableVisualizations/donutChart/DonutChartDescriptor";
import { FunnelChartDescriptor } from "./pluggableVisualizations/funnelChart/FunnelChartDescriptor";
import { HeadlineDescriptor } from "./pluggableVisualizations/headline/HeadlineDescriptor";
import { HeatmapDescriptor } from "./pluggableVisualizations/heatMap/HeatmapDescriptor";
import { LineChartDescriptor } from "./pluggableVisualizations/lineChart/LineChartDescriptor";
import { PieChartDescriptor } from "./pluggableVisualizations/pieChart/PieChartDescriptor";
import { PivotTableDescriptor } from "./pluggableVisualizations/pivotTable/PivotTableDescriptor";
import { ScatterPlotDescriptor } from "./pluggableVisualizations/scatterPlot/ScatterPlotDescriptor";
import { TreemapDescriptor } from "./pluggableVisualizations/treeMap/TreemapDescriptor";
import { XirrDescriptor } from "./pluggableVisualizations/xirr/XirrDescriptor";
import { GeoPushpinChartDescriptor } from "./pluggableVisualizations/geoChart/GeoPushpinChartDescriptor";
/**
 * @internal
 */
var CatalogViaTypeToClassMap = /** @class */ (function () {
    function CatalogViaTypeToClassMap(mapping) {
        this.mapping = mapping;
    }
    CatalogViaTypeToClassMap.prototype.forUri = function (uri) {
        var VisType = this.findInMapping(uri);
        if (!VisType) {
            throw new UnexpectedSdkError("Unknown visualization class URI: " + uri);
        }
        return new VisType();
    };
    CatalogViaTypeToClassMap.prototype.hasDescriptorForUri = function (uri) {
        return !!this.findInMapping(uri);
    };
    CatalogViaTypeToClassMap.prototype.forInsight = function (insight) {
        /*
         * the URIs follow "local:visualizationType" format
         */
        var visClassUri = insightVisualizationUrl(insight);
        return this.forUri(visClassUri);
    };
    CatalogViaTypeToClassMap.prototype.hasDescriptorForInsight = function (insight) {
        var visClassUri = insightVisualizationUrl(insight);
        return this.hasDescriptorForUri(visClassUri);
    };
    CatalogViaTypeToClassMap.prototype.findInMapping = function (uri) {
        var split = uri.split(":");
        var key = last(split);
        return this.mapping[key];
    };
    return CatalogViaTypeToClassMap;
}());
export { CatalogViaTypeToClassMap };
//
// Default static catalog implementation
//
/**
 * Compile-time pluggable visualization 'catalog'.
 */
var DefaultVisualizations = {
    bar: BarChartDescriptor,
    bullet: BulletChartDescriptor,
    column: ColumnChartDescriptor,
    line: LineChartDescriptor,
    area: AreaChartDescriptor,
    pie: PieChartDescriptor,
    donut: DonutChartDescriptor,
    table: PivotTableDescriptor,
    headline: HeadlineDescriptor,
    scatter: ScatterPlotDescriptor,
    bubble: BubbleChartDescriptor,
    heatmap: HeatmapDescriptor,
    combo: ComboChartDescriptorDeprecated,
    combo2: ComboChartDescriptor,
    treemap: TreemapDescriptor,
    funnel: FunnelChartDescriptor,
    pushpin: GeoPushpinChartDescriptor,
};
/**
 * Default pluggable visualization catalog. This is implemented using static lookup table between vis type
 * and the actual plug vis implementation.
 *
 * @alpha
 */
export var DefaultVisualizationCatalog = new CatalogViaTypeToClassMap(DefaultVisualizations);
/**
 * Pluggable visualization catalog containing all available visualizations.
 *
 * @alpha
 */
export var FullVisualizationCatalog = new CatalogViaTypeToClassMap(__assign(__assign({}, DefaultVisualizations), { xirr: XirrDescriptor }));
