// @gooddata/sdk-backend-tiger/esm/backend/organization/index.js
// (C) 2021-2022 GoodData Corporation
import { __awaiter, __generator } from "tslib";
import { SecuritySettingsService } from "./securitySettings";
var TigerOrganization = /** @class */ (function () {
    function TigerOrganization(authCall, organizationId, organizationName) {
        this.authCall = authCall;
        this.organizationId = organizationId;
        this.organizationName = organizationName;
    }
    TigerOrganization.prototype.getDescriptor = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response, organizationData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // if we already have the data, no reason to download them again
                        if (this.organizationName) {
                            return [2 /*return*/, {
                                    id: this.organizationId,
                                    title: this.organizationName,
                                }];
                        }
                        return [4 /*yield*/, this.authCall(function (client) { return client.axios.get("/api/profile"); })];
                    case 1:
                        response = _a.sent();
                        organizationData = response.data;
                        return [2 /*return*/, {
                                id: organizationData.organizationId,
                                title: organizationData.organizationName,
                            }];
                }
            });
        });
    };
    TigerOrganization.prototype.securitySettings = function () {
        return new SecuritySettingsService(this.organizationId);
    };
    return TigerOrganization;
}());
export { TigerOrganization };
var TigerOrganizations = /** @class */ (function () {
    function TigerOrganizations(authCall) {
        this.authCall = authCall;
    }
    TigerOrganizations.prototype.getCurrentOrganization = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response, organizationData, organizationId, organizationName;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.authCall(function (client) { return client.axios.get("/api/profile"); })];
                    case 1:
                        response = _a.sent();
                        organizationData = response.data;
                        organizationId = organizationData.organizationId;
                        organizationName = organizationData.organizationName;
                        return [2 /*return*/, new TigerOrganization(this.authCall, organizationId, organizationName)];
                }
            });
        });
    };
    return TigerOrganizations;
}());
export { TigerOrganizations };
