// @gooddata/sdk-ui-kit/esm/List/LegacySingleSelectList.js
import { __assign, __extends } from "tslib";
// (C) 2007-2020 GoodData Corporation
import React, { Component } from "react";
import pick from "lodash/pick";
import cx from "classnames";
import noop from "lodash/noop";
import { LegacyList } from "./LegacyList";
import { LegacyListItem } from "./LegacyListItem";
import { LegacySingleSelectListItem } from "./LegacySingleSelectListItem";
import { guidFor } from "./guid";
/**
 * @internal
 * @deprecated This component is deprecated use SingleSelectList instead
 */
var LegacySingleSelectList = /** @class */ (function (_super) {
    __extends(LegacySingleSelectList, _super);
    function LegacySingleSelectList() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LegacySingleSelectList.prototype.getSelectableItems = function () {
        var props = this.props;
        return props.items.map(function (source) { return ({
            source: source,
            onSelect: props.onSelect,
            onMouseOver: props.onItemMouseOver,
            onMouseOut: props.onItemMouseOut,
            onMouseEnter: props.onItemMouseEnter,
            onMouseLeave: props.onItemMouseLeave,
            selected: source === props.selection,
            id: props.getItemKey(source),
        }); });
    };
    LegacySingleSelectList.prototype.getClassNames = function () {
        return cx("gd-list", this.props.className);
    };
    LegacySingleSelectList.prototype.getRowItem = function () {
        var _a = this.props, rowItem = _a.rowItem, listItemClass = _a.listItemClass;
        return rowItem || React.createElement(LegacyListItem, { listItemClass: listItemClass });
    };
    LegacySingleSelectList.prototype.getDataSource = function () {
        var selectableItems = this.getSelectableItems();
        return {
            rowsCount: this.props.itemsCount || selectableItems.length,
            getObjectAt: function (rowIndex) { return selectableItems[rowIndex]; },
        };
    };
    LegacySingleSelectList.prototype.render = function () {
        var rowItem = this.getRowItem();
        var dataSource = this.getDataSource();
        return (React.createElement(LegacyList, __assign({ className: this.getClassNames() }, pick(this.props, ["width", "height", "itemHeight"]), { dataSource: dataSource, rowItem: rowItem, onScroll: this.props.onRangeChange, onScrollStart: this.props.onScrollStart, scrollToSelected: this.props.scrollToSelected })));
    };
    LegacySingleSelectList.defaultProps = {
        className: "",
        getItemKey: guidFor,
        items: [],
        itemsCount: 0,
        listItemClass: LegacySingleSelectListItem,
        onItemMouseOut: noop,
        onItemMouseOver: noop,
        onItemMouseEnter: noop,
        onItemMouseLeave: noop,
        onRangeChange: noop,
        onScrollStart: noop,
        onSelect: noop,
        rowItem: null,
        scrollToSelected: false,
        selection: {},
    };
    return LegacySingleSelectList;
}(Component));
export { LegacySingleSelectList };
