// @gooddata/sdk-ui-kit/esm/List/MenuList.js
import { __assign, __rest } from "tslib";
// (C) 2007-2020 GoodData Corporation
import React from "react";
import classnames from "classnames";
/**
 * @internal
 */
export var ItemsWrapper = function (_a) {
    var smallItemsSpacing = _a.smallItemsSpacing, className = _a.className, children = _a.children, style = _a.style;
    return (React.createElement("div", { className: classnames({
            "gd-menu-wrapper": true,
            "gd-menu-wrapper-small-spacing": smallItemsSpacing,
        }, className), style: style }, children));
};
ItemsWrapper.defaultProps = {
    smallItemsSpacing: false,
};
/**
 * @internal
 */
export var Separator = function (props) { return (React.createElement("div", __assign({ className: "gd-list-item gd-list-item-separator" }, props))); };
/**
 * @internal
 */
export var Header = function (_a) {
    var children = _a.children, restProps = __rest(_a, ["children"]);
    return (React.createElement("div", __assign({ className: "gd-list-item gd-list-item-header" }, restProps), children));
};
/**
 * @internal
 */
export var Item = function (_a) {
    var checked = _a.checked, subMenu = _a.subMenu, disabled = _a.disabled, className = _a.className, children = _a.children, style = _a.style, onClick = _a.onClick;
    return (React.createElement("div", { className: classnames({
            "gd-list-item": true,
            "gd-menu-item": true,
            "is-checked": checked,
            "is-submenu": subMenu,
            "is-disabled": disabled,
        }, className), style: style, onClick: onClick }, children));
};
Item.defaultProps = {
    checked: false,
    subMenu: false,
    disabled: false,
};
