// @gooddata/sdk-ui-filters/esm/DateFilter/DateRangePicker/DateRangePicker.js
import { __assign, __extends } from "tslib";
// (C) 2007-2019 GoodData Corporation
import React from "react";
import cx from "classnames";
import MomentLocaleUtils from "react-day-picker/moment";
import { injectIntl } from "react-intl";
import { translationUtils } from "@gooddata/util";
import { convertDateToPlatformDateString, convertPlatformDateStringToDate } from "../utils/DateConversions";
import { DateRangePickerInputField } from "./DateRangePickerInputField";
import { mergeDayPickerProps, areRangeBoundsCrossed } from "./utils";
import { DateRangePickerError } from "./DateRangePickerError";
import { DateRangePickerInputFieldBody } from "./DateRangePickerInputFieldBody";
var DateRangePickerComponent = /** @class */ (function (_super) {
    __extends(DateRangePickerComponent, _super);
    function DateRangePickerComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.fromInputRef = React.createRef();
        _this.toInputRef = React.createRef();
        _this.focusField = function (inputRef) {
            if (inputRef.current) {
                // Focus needs to happen on the next tick otherwise the day picker is not updated
                setTimeout(function () {
                    inputRef.current.getInput().focus();
                }, 0);
            }
        };
        _this.blurField = function (inputRef) {
            if (inputRef.current) {
                inputRef.current.getInput().blur();
            }
        };
        _this.updateRange = function (from, to) {
            _this.props.onRangeChange({ from: from, to: to });
        };
        _this.handleFromDayClick = function () {
            _this.focusField(_this.toInputRef);
        };
        _this.handleToDayClick = function (to) {
            var rangeBoundsCrossed = areRangeBoundsCrossed(_this.props.range.from, to);
            if (to && !rangeBoundsCrossed) {
                _this.blurField(_this.toInputRef);
            }
            else if (rangeBoundsCrossed) {
                _this.focusField(_this.fromInputRef);
            }
        };
        _this.getSanitizedInputValue = function (inputRef, date) {
            var inputValue = inputRef.current && inputRef.current.getInput().value;
            return !date && !inputValue ? null : date;
        };
        _this.handleFromChange = function (from) {
            var useFrom = from && _this.props.range.to && areRangeBoundsCrossed(from, _this.props.range.to);
            var to = useFrom ? from : _this.props.range.to;
            var sanitizedFrom = _this.getSanitizedInputValue(_this.fromInputRef, from);
            _this.updateRange(sanitizedFrom, to);
        };
        _this.handleToChange = function (to) {
            var useTo = to && _this.props.range.from && areRangeBoundsCrossed(_this.props.range.from, to);
            var from = useTo ? to : _this.props.range.from;
            var sanitizedTo = _this.getSanitizedInputValue(_this.toInputRef, to);
            _this.updateRange(from, sanitizedTo);
        };
        return _this;
    }
    DateRangePickerComponent.prototype.render = function () {
        var _this = this;
        var _a = this.props, dateFormat = _a.dateFormat, _b = _a.range, from = _b.from, to = _b.to, dayPickerProps = _a.dayPickerProps, intl = _a.intl, isMobile = _a.isMobile, _c = _a.errors, _d = _c === void 0 ? { from: undefined, to: undefined } : _c, errorFrom = _d.from, errorTo = _d.to;
        var defaultDayPickerProps = {
            showOutsideDays: true,
            modifiers: { start: from, end: to },
            selectedDays: [from, { from: from, to: to }],
            locale: translationUtils.sanitizeLocaleForMoment(intl.locale),
            localeUtils: MomentLocaleUtils,
        };
        var dayPickerPropsWithDefaults = mergeDayPickerProps(defaultDayPickerProps, dayPickerProps);
        var FromField = isMobile ? (React.createElement(DateRangePickerInputFieldBody, { type: "date", className: cx("s-date-range-picker-from", "gd-date-range-picker-input", "gd-date-range-picker-input-native", errorFrom && "gd-date-range-picker-input-error"), placeholder: intl.formatMessage({ id: "filters.from" }), onChange: function (event) {
                return _this.handleFromChange(convertPlatformDateStringToDate(event.target.value));
            }, value: convertDateToPlatformDateString(from) })) : (React.createElement(DateRangePickerInputField, { className: cx("s-date-range-picker-from", errorFrom && "gd-date-range-picker-input-error"), classNameCalendar: "s-date-range-calendar-from", ref: this.fromInputRef, onDayChange: this.handleFromChange, value: from || "", format: dateFormat, placeholder: intl.formatMessage({ id: "filters.from" }), dayPickerProps: __assign(__assign({}, dayPickerPropsWithDefaults), { onDayClick: this.handleFromDayClick }) }));
        var ToField = isMobile ? (React.createElement(DateRangePickerInputFieldBody, { type: "date", className: cx("s-date-range-picker-to", "gd-date-range-picker-input", "gd-date-range-picker-input-native", errorTo && "gd-date-range-picker-input-error"), placeholder: intl.formatMessage({ id: "filters.to" }), onChange: function (event) { return _this.handleToChange(convertPlatformDateStringToDate(event.target.value)); }, value: convertDateToPlatformDateString(to) })) : (React.createElement(DateRangePickerInputField, { className: cx("s-date-range-picker-to", errorTo && "gd-date-range-picker-input-error"), classNameCalendar: "s-date-range-calendar-to", ref: this.toInputRef, onDayChange: this.handleToChange, value: to || "", format: dateFormat, placeholder: intl.formatMessage({ id: "filters.to" }), dayPickerProps: __assign(__assign({}, dayPickerPropsWithDefaults), { onDayClick: this.handleToDayClick }) }));
        return (React.createElement(React.Fragment, null,
            React.createElement("div", { className: "gd-date-range-picker s-date-range-picker" },
                FromField,
                React.createElement("span", { className: "gd-date-range-picker-dash" }, "\u2014"),
                ToField),
            (errorFrom || errorTo) && (React.createElement(DateRangePickerError, { dateFormat: dateFormat, errorId: 
                // This means that when both inputs are invalid, error is shown only for "from"
                errorFrom || errorTo }))));
    };
    return DateRangePickerComponent;
}(React.Component));
export var DateRangePicker = injectIntl(DateRangePickerComponent);
