// @gooddata/sdk-ui-charts/esm/highcharts/chartTypes/comboChart/comboConfiguration.js
import { __assign } from "tslib";
// (C) 2007-2021 GoodData Corporation
import { MAX_POINT_WIDTH } from "../_chartCreators/commonConfiguration";
import { LINE_WIDTH } from "../lineChart/lineConfiguration";
import { isLineChart } from "../_util/common";
import { BucketNames, VisualizationTypes } from "@gooddata/sdk-ui";
import { bucketIsEmpty, bucketsFind } from "@gooddata/sdk-model";
import { styleVariables } from "../_chartCreators/styles/variables";
var COLUMN = VisualizationTypes.COLUMN, LINE = VisualizationTypes.LINE;
function getDefaultComboTypes(config) {
    var _a, _b;
    return {
        primaryChartType: (_a = config === null || config === void 0 ? void 0 : config.primaryChartType) !== null && _a !== void 0 ? _a : COLUMN,
        secondaryChartType: (_b = config === null || config === void 0 ? void 0 : config.secondaryChartType) !== null && _b !== void 0 ? _b : LINE,
    };
}
export function getDefaultChartType(config) {
    var _a = getDefaultComboTypes(config), primaryChartType = _a.primaryChartType, secondaryChartType = _a.secondaryChartType;
    if (primaryChartType === secondaryChartType) {
        return primaryChartType;
    }
    if (primaryChartType === COLUMN || secondaryChartType === COLUMN) {
        return COLUMN;
    }
    return LINE;
}
function isOnlyLineSeries(config, definition) {
    var _a = getDefaultComboTypes(config), primaryChartType = _a.primaryChartType, secondaryChartType = _a.secondaryChartType;
    var buckets = definition ? definition.buckets : [];
    var primaryBucket = bucketsFind(buckets, BucketNames.MEASURES);
    var secondaryBucket = bucketsFind(buckets, BucketNames.SECONDARY_MEASURES);
    var isEmptyPrimaryMeasure = !primaryBucket || bucketIsEmpty(primaryBucket);
    var isEmptySecondaryMeasure = !secondaryBucket || bucketIsEmpty(secondaryBucket);
    var isLineChartOnLeftAxis = isLineChart(primaryChartType);
    var isLineChartOnRightAxis = isLineChart(secondaryChartType);
    return ((isLineChartOnLeftAxis && isLineChartOnRightAxis) ||
        (isLineChartOnLeftAxis && isEmptySecondaryMeasure) ||
        (isEmptyPrimaryMeasure && isLineChartOnRightAxis));
}
export function getComboConfiguration(config, definition, theme) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    var series = isOnlyLineSeries(config, definition)
        ? {
            series: {
                states: {
                    inactive: {
                        opacity: 1,
                    },
                },
                borderColor: "#00000000",
            },
        }
        : {
            series: {
                borderColor: "#00000000",
            },
        };
    return {
        chart: {
            type: getDefaultChartType(config),
            spacingTop: 20,
        },
        plotOptions: __assign({ column: {
                dataLabels: {
                    enabled: true,
                    crop: false,
                    overflow: "none",
                    padding: 2,
                },
                maxPointWidth: MAX_POINT_WIDTH,
                borderColor: "#00000000",
            }, line: {
                marker: {
                    symbol: "circle",
                    radius: 4.5,
                    lineColor: (_e = (_b = (_a = theme === null || theme === void 0 ? void 0 : theme.chart) === null || _a === void 0 ? void 0 : _a.backgroundColor) !== null && _b !== void 0 ? _b : (_d = (_c = theme === null || theme === void 0 ? void 0 : theme.palette) === null || _c === void 0 ? void 0 : _c.complementary) === null || _d === void 0 ? void 0 : _d.c0) !== null && _e !== void 0 ? _e : styleVariables.gdColorBackground,
                },
                lineWidth: LINE_WIDTH,
                fillOpacity: 0.3,
                stickyTracking: false,
                states: {
                    hover: {
                        lineWidth: LINE_WIDTH + 1,
                    },
                },
                dataLabels: {
                    style: {
                        fontWeight: "normal",
                    },
                },
            }, area: {
                marker: {
                    symbol: "circle",
                    radius: 4.5,
                    lineColor: (_k = (_g = (_f = theme === null || theme === void 0 ? void 0 : theme.chart) === null || _f === void 0 ? void 0 : _f.backgroundColor) !== null && _g !== void 0 ? _g : (_j = (_h = theme === null || theme === void 0 ? void 0 : theme.palette) === null || _h === void 0 ? void 0 : _h.complementary) === null || _j === void 0 ? void 0 : _j.c0) !== null && _k !== void 0 ? _k : styleVariables.gdColorBackground,
                },
                lineWidth: LINE_WIDTH,
                fillOpacity: 0.6,
                stickyTracking: false,
                states: {
                    hover: {
                        lineWidth: LINE_WIDTH + 1,
                    },
                },
            } }, series),
    };
}
