// @gooddata/api-model-bear/esm/meta/GdcMetadata.js
// (C) 2007-2021 GoodData Corporation
import isEmpty from "lodash/fp/isEmpty";
import has from "lodash/has";
/**
 * @public
 */
export var GdcMetadata;
(function (GdcMetadata) {
    function isAttribute(obj) {
        return !isEmpty(obj) && obj.meta.category === "attribute";
    }
    GdcMetadata.isAttribute = isAttribute;
    function isWrappedAttribute(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "attribute");
    }
    GdcMetadata.isWrappedAttribute = isWrappedAttribute;
    function isWrappedAttributeDisplayForm(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "attributeDisplayForm");
    }
    GdcMetadata.isWrappedAttributeDisplayForm = isWrappedAttributeDisplayForm;
    function isAttributeDisplayForm(obj) {
        return !isEmpty(obj) && obj.meta.category === "attributeDisplayForm";
    }
    GdcMetadata.isAttributeDisplayForm = isAttributeDisplayForm;
    function isWrappedMetric(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "metric");
    }
    GdcMetadata.isWrappedMetric = isWrappedMetric;
    function isMetric(obj) {
        return !isEmpty(obj) && obj.meta.category === "metric";
    }
    GdcMetadata.isMetric = isMetric;
    function isWrappedFact(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "fact");
    }
    GdcMetadata.isWrappedFact = isWrappedFact;
    function isFact(obj) {
        return !isEmpty(obj) && obj.meta.category === "fact";
    }
    GdcMetadata.isFact = isFact;
    function isKpiAlert(obj) {
        return !isEmpty(obj) && obj.meta.category === "kpiAlert";
    }
    GdcMetadata.isKpiAlert = isKpiAlert;
    function isWrappedKpiAlert(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "kpiAlert");
    }
    GdcMetadata.isWrappedKpiAlert = isWrappedKpiAlert;
    function isDataSet(obj) {
        return !isEmpty(obj) && obj.meta.category === "dataSet";
    }
    GdcMetadata.isDataSet = isDataSet;
    function isWrappedDataSet(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "dataSet");
    }
    GdcMetadata.isWrappedDataSet = isWrappedDataSet;
    function isPrompt(obj) {
        return !isEmpty(obj) && obj.meta.category === "prompt";
    }
    GdcMetadata.isPrompt = isPrompt;
    function isWrappedPrompt(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "prompt");
    }
    GdcMetadata.isWrappedPrompt = isWrappedPrompt;
    function isTheme(obj) {
        return !isEmpty(obj) && obj.meta.category === "theme";
    }
    GdcMetadata.isTheme = isTheme;
    function isWrappedTheme(obj) {
        // eslint-disable-next-line no-prototype-builtins
        return !isEmpty(obj) && has(obj, "theme");
    }
    GdcMetadata.isWrappedTheme = isWrappedTheme;
})(GdcMetadata || (GdcMetadata = {}));
