// @gooddata/sdk-ui-kit/esm/AutoSize/AutoSize.js
import { __extends } from "tslib";
// (C) 2007-2020 GoodData Corporation
import React, { Component, createRef } from "react";
import throttle from "lodash/throttle";
import { elementRegion } from "../utils/domUtilities";
/**
 * @internal
 */
var AutoSize = /** @class */ (function (_super) {
    __extends(AutoSize, _super);
    function AutoSize() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            width: 0,
            height: 0,
        };
        _this.updateSize = function () {
            var _a = elementRegion(_this.wrapperRef.current), width = _a.width, height = _a.height;
            _this.setState({
                width: width,
                height: height,
            });
        };
        _this.throttledUpdateSize = throttle(_this.updateSize, 250, { leading: false });
        _this.wrapperRef = createRef();
        return _this;
    }
    AutoSize.prototype.render = function () {
        var children = this.props.children;
        var _a = this.state, width = _a.width, height = _a.height;
        return (React.createElement("div", { ref: this.wrapperRef, style: { height: "100%", width: "100%" } }, children({ width: width, height: height })));
    };
    AutoSize.prototype.componentDidMount = function () {
        window.addEventListener("resize", this.throttledUpdateSize);
        this.throttledUpdateSize();
    };
    AutoSize.prototype.componentWillUnmount = function () {
        this.throttledUpdateSize.cancel();
        window.removeEventListener("resize", this.throttledUpdateSize);
    };
    return AutoSize;
}(Component));
export { AutoSize };
